// @browserbasehq/stagehand/dist/cjs/lib/v3/cache/CacheStorage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheStorage = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const jsonClone = (value) => {
    const serialized = JSON.stringify(value);
    if (serialized === undefined) {
        return value;
    }
    return JSON.parse(serialized);
};
class CacheStorage {
    logger;
    dir;
    memoryStore;
    constructor(logger, dir, memoryStore) {
        this.logger = logger;
        this.dir = dir;
        this.memoryStore = memoryStore;
    }
    static create(cacheDir, logger, options) {
        if (!cacheDir) {
            return new CacheStorage(logger);
        }
        const resolved = path_1.default.resolve(cacheDir);
        try {
            fs_1.default.mkdirSync(resolved, { recursive: true });
            return new CacheStorage(logger, resolved);
        }
        catch (err) {
            const label = options?.label ?? "cache directory";
            logger({
                category: "cache",
                message: `unable to initialize ${label}: ${resolved}`,
                level: 1,
                auxiliary: {
                    error: { value: String(err), type: "string" },
                },
            });
            return new CacheStorage(logger);
        }
    }
    static createMemory(logger) {
        return new CacheStorage(logger, undefined, new Map());
    }
    get directory() {
        return this.dir;
    }
    get enabled() {
        return !!this.dir || !!this.memoryStore;
    }
    resolvePath(fileName) {
        if (!this.dir)
            return null;
        return path_1.default.join(this.dir, fileName);
    }
    async readJson(fileName) {
        if (this.memoryStore) {
            if (!this.memoryStore.has(fileName)) {
                return { value: null };
            }
            const existing = this.memoryStore.get(fileName);
            return { value: jsonClone(existing) };
        }
        const filePath = this.resolvePath(fileName);
        if (!filePath) {
            return { value: null };
        }
        try {
            const raw = await fs_1.default.promises.readFile(filePath, "utf8");
            return { value: JSON.parse(raw) };
        }
        catch (err) {
            const code = err?.code;
            if (code === "ENOENT") {
                return { value: null };
            }
            return { value: null, error: err, path: filePath };
        }
    }
    async writeJson(fileName, data) {
        if (this.memoryStore) {
            this.memoryStore.set(fileName, jsonClone(data));
            return {};
        }
        const filePath = this.resolvePath(fileName);
        if (!filePath) {
            return {};
        }
        try {
            await fs_1.default.promises.mkdir(path_1.default.dirname(filePath), { recursive: true });
            await fs_1.default.promises.writeFile(filePath, JSON.stringify(data, null, 2), "utf8");
            return {};
        }
        catch (err) {
            return { error: err, path: filePath };
        }
    }
}
exports.CacheStorage = CacheStorage;
