// @browserbasehq/stagehand/dist/cjs/lib/v3/cache/serverAgentCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__internalCreateInMemoryAgentCacheHandle = __internalCreateInMemoryAgentCacheHandle;
const AgentCache_js_1 = require("./AgentCache.js");
const CacheStorage_js_1 = require("./CacheStorage.js");
// TODO (refactor-caching): this reflective access is a known temporary escape hatch.
// Once the caching internals are reworked, replace it with proper V3 helpers so
// we stop poking private fields from the outside.
function getInternalField(instance, key) {
    return instance[key];
}
function setInternalField(instance, key, value) {
    instance[key] = value;
}
function createMemoryAgentCache(stagehand) {
    const resolveLlmClient = getInternalField(stagehand, "resolveLlmClient");
    return new AgentCache_js_1.AgentCache({
        storage: CacheStorage_js_1.CacheStorage.createMemory(stagehand.logger),
        logger: stagehand.logger,
        getActHandler: () => getInternalField(stagehand, "actHandler"),
        getContext: () => getInternalField(stagehand, "ctx"),
        getDefaultLlmClient: () => resolveLlmClient.call(stagehand),
        getBaseModelName: () => getInternalField(stagehand, "modelName"),
        getSystemPrompt: () => getInternalField(stagehand, "opts").systemPrompt,
        domSettleTimeoutMs: getInternalField(stagehand, "domSettleTimeoutMs"),
        act: stagehand.act.bind(stagehand),
        bufferLatestEntry: true,
    });
}
function __internalCreateInMemoryAgentCacheHandle(stagehand) {
    const originalCache = getInternalField(stagehand, "agentCache");
    const memoryCache = createMemoryAgentCache(stagehand);
    setInternalField(stagehand, "agentCache", memoryCache);
    let restored = false;
    const restore = () => {
        if (!restored) {
            setInternalField(stagehand, "agentCache", originalCache);
            restored = true;
        }
    };
    return {
        complete: () => {
            const entry = memoryCache.consumeBufferedEntry();
            restore();
            return entry;
        },
        discard: () => {
            restore();
        },
    };
}
