// @browserbasehq/stagehand/dist/cjs/lib/v3/cache/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneForCache = cloneForCache;
exports.safeGetPageUrl = safeGetPageUrl;
exports.waitForCachedSelector = waitForCachedSelector;
const DEFAULT_WAIT_TIMEOUT_MS = 15000;
function cloneForCache(value) {
    return JSON.parse(JSON.stringify(value));
}
async function safeGetPageUrl(page) {
    try {
        return page.url();
    }
    catch {
        return "";
    }
}
/**
 * Waits for a cached action's selector to be attached to the DOM before executing.
 * Logs a warning and proceeds if the wait times out (non-blocking).
 */
async function waitForCachedSelector(params) {
    const { page, selector, timeout, logger, context } = params;
    if (!selector)
        return;
    try {
        await page.waitForSelector(selector, {
            state: "attached",
            timeout: timeout ?? DEFAULT_WAIT_TIMEOUT_MS,
        });
    }
    catch (err) {
        logger({
            category: "cache",
            message: `waitForSelector failed for ${context ?? "cached"} action selector, proceeding anyway`,
            level: 2,
            auxiliary: {
                selector: { value: selector, type: "string" },
                error: { value: String(err), type: "string" },
            },
        });
    }
}
