// @browserbasehq/stagehand/dist/cjs/lib/v3/flowlogger/EventSink.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShallowInMemoryEventSink = exports.InMemoryEventSink = exports.PrettyStderrEventSink = exports.PrettyLogFileEventSink = exports.JsonlFileEventSink = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const FlowLogger_js_1 = require("./FlowLogger.js");
const prettify_js_1 = require("./prettify.js");
// Checks whether an event matches a query used by queryable sinks. `eventId` matches both the event itself and descendants of that event.
function matchesEventStoreQuery(event, query) {
    if (query.sessionId && event.sessionId !== query.sessionId)
        return false;
    if (query.eventId) {
        const matchesEvent = event.eventId === query.eventId ||
            event.eventParentIds.includes(query.eventId);
        if (!matchesEvent) {
            return false;
        }
    }
    if (query.eventType) {
        const pattern = new RegExp(`^${query.eventType
            .replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
            .replace(/\\\*/g, ".*")}$`);
        if (!pattern.test(event.eventType)) {
            return false;
        }
    }
    return true;
}
// =============================================================================
// File Sink Helpers
// =============================================================================
// Returns true when a file sink's stream is still open and writable.
function isWritable(stream) {
    return !!(stream && !stream.destroyed && stream.writable);
}
// Writes a serialized event to a file sink and converts callback-style stream completion into a promise.
function writeToStream(stream, value) {
    return new Promise((resolve, reject) => {
        try {
            stream.write(value, (error) => {
                if (error) {
                    reject(error);
                    return;
                }
                resolve();
            });
        }
        catch (error) {
            reject(error);
        }
    });
}
// =============================================================================
// Event Sink Implementations
// =============================================================================
class FileEventSink {
    streamPromise; // Lazily opens the one file stream owned by this sink when the session directory resolves.
    // Creates a best-effort file sink bound to a single session directory.
    constructor(sessionDirPromise, fileName) {
        this.streamPromise = sessionDirPromise.then((sessionDir) => sessionDir
            ? node_fs_1.default.createWriteStream(node_path_1.default.join(sessionDir, fileName), { flags: "a" })
            : null);
    }
    // Serializes and appends a single event. File sinks are intentionally best-effort and never allowed to affect library execution flow.
    async emit(event) {
        try {
            const stream = await this.streamPromise;
            if (!isWritable(stream)) {
                return;
            }
            const serialized = await this.serialize(event);
            if (!serialized) {
                return;
            }
            await writeToStream(stream, serialized);
        }
        catch {
            // best effort only
        }
    }
    // File sinks are write-only and do not support query reads.
    async query() {
        return [];
    }
    // Closes the underlying file stream when the owning store shuts down.
    async destroy() {
        const stream = await this.streamPromise.catch(() => null);
        if (!isWritable(stream)) {
            return;
        }
        await new Promise((resolve) => {
            stream.end(resolve);
        });
    }
}
class JsonlFileEventSink extends FileEventSink {
    // Writes full verbatim events to `session_events.jsonl`.
    constructor(sessionDirPromise) {
        super(sessionDirPromise, "session_events.jsonl");
    }
    // Serializes the full event for lossless machine-readable storage.
    async serialize(event) {
        return `${JSON.stringify(event)}\n`;
    }
}
exports.JsonlFileEventSink = JsonlFileEventSink;
class PrettyLogFileEventSink extends FileEventSink {
    store;
    // Writes human-readable pretty lines to `session_events.log`.
    constructor(sessionDirPromise, store) {
        super(sessionDirPromise, "session_events.log");
        this.store = store;
    }
    // Pretty-prints the event using recent in-memory ancestry.
    async serialize(event) {
        const line = await (0, prettify_js_1.prettifyEvent)(this.store, (0, prettify_js_1.prettifySanitizeEvent)(event));
        return line ? `${line}\n` : null;
    }
}
exports.PrettyLogFileEventSink = PrettyLogFileEventSink;
class PrettyStderrEventSink {
    store;
    // Writes pretty lines to stderr for verbose local debugging. CDP events are intentionally omitted here to keep stderr high-signal.
    constructor(store) {
        this.store = store;
    } // Queried during prettification so stderr lines can include recent ancestry tags.
    // Best-effort stderr writer used only for interactive debugging output.
    async emit(event) {
        try {
            if ((0, prettify_js_1.prettifyIsCdpEvent)(event)) {
                return;
            }
            const line = await (0, prettify_js_1.prettifyEvent)(this.store, (0, prettify_js_1.prettifySanitizeEvent)(event));
            if (!line) {
                return;
            }
            await new Promise((resolve, reject) => {
                try {
                    process.stderr.write(`${(0, prettify_js_1.prettifyColorStderrLine)(line)}\n`, (error) => {
                        if (error) {
                            reject(error);
                            return;
                        }
                        resolve();
                    });
                }
                catch (error) {
                    reject(error);
                }
            });
        }
        catch {
            // best effort only
        }
    }
    // Stderr sink is write-only and does not support query reads.
    async query() {
        return [];
    }
    // No teardown is required for stderr.
    async destroy() { }
}
exports.PrettyStderrEventSink = PrettyStderrEventSink;
class InMemoryEventSink {
    limit;
    // Retains recent events for query lookups. Tests usually attach this sink explicitly when they need full historical payloads.
    constructor(limit = Infinity) {
        this.limit = limit;
    }
    events = []; // Retained history; `emit()` appends to it and trims old entries when `limit` is exceeded.
    // Gives subclasses a hook to transform events before they are retained.
    storeEvent(event) {
        return event;
    }
    // Stores a new event and trims the oldest retained entries once the sink exceeds its configured limit.
    async emit(event) {
        this.events.push(this.storeEvent(event));
        if (this.events.length > this.limit) {
            this.events.splice(0, this.events.length - this.limit);
        }
    }
    // Returns retained events that match the query, ordered by creation time.
    async query(query) {
        const filtered = this.events.filter((event) => matchesEventStoreQuery(event, query));
        filtered.sort((left, right) => {
            const createdAtOrder = left.eventCreatedAt.localeCompare(right.eventCreatedAt);
            if (createdAtOrder !== 0) {
                return createdAtOrder;
            }
            return left.eventId.localeCompare(right.eventId);
        });
        return query.limit ? filtered.slice(-query.limit) : filtered;
    }
    // Clears retained history when the owning store shuts down.
    async destroy() {
        this.events.length = 0;
    }
}
exports.InMemoryEventSink = InMemoryEventSink;
class ShallowInMemoryEventSink extends InMemoryEventSink {
    // Retains only ancestry metadata for the default query sink so verbose or long-running sessions do not hold onto large payloads such as screenshots.
    storeEvent(event) {
        return new FlowLogger_js_1.FlowEvent({
            eventType: event.eventType,
            eventId: event.eventId,
            eventCreatedAt: event.eventCreatedAt,
            sessionId: event.sessionId,
            eventParentIds: [...event.eventParentIds],
            data: {},
        });
    }
}
exports.ShallowInMemoryEventSink = ShallowInMemoryEventSink;
