// @browserbasehq/stagehand/dist/esm/tests/integration/testUtils.js
import { AISdkClient } from "../../lib/v3/llm/aisdk.js";
/**
 * Races a promise against a timeout.
 * Resolves to the promise value or "timeout" if the deadline expires.
 */
export function raceTimeout(promise, ms) {
    let timer;
    const timeout = new Promise((resolve) => {
        timer = setTimeout(() => resolve("timeout"), ms);
    });
    return Promise.race([promise, timeout]).finally(() => clearTimeout(timer));
}
const CLOSE_TIMEOUT_MS = 5_000;
async function settleWithTimeout(promise, timeoutMs) {
    let timeoutId;
    const timeout = new Promise((resolve) => {
        timeoutId = setTimeout(resolve, timeoutMs);
    });
    try {
        await Promise.race([promise.catch(() => { }), timeout]);
    }
    finally {
        if (timeoutId)
            clearTimeout(timeoutId);
    }
}
export async function closeV3(v3) {
    if (!v3)
        return;
    const isBrowserbase = v3.isBrowserbase;
    if (isBrowserbase) {
        try {
            await settleWithTimeout(v3.context.conn.send("Browser.close"), CLOSE_TIMEOUT_MS);
        }
        catch {
            // best-effort cleanup
        }
    }
    await settleWithTimeout(v3.close(), CLOSE_TIMEOUT_MS);
}
const DEFAULT_USAGE = {
    inputTokens: 1,
    outputTokens: 1,
    totalTokens: 2,
    reasoningTokens: 0,
    cachedInputTokens: 0,
};
const mergeUsage = (usage) => ({
    ...DEFAULT_USAGE,
    ...(usage ?? {}),
});
function consumeScriptValue(value, fallback) {
    if (!Array.isArray(value)) {
        return value ?? fallback;
    }
    if (value.length <= 1) {
        return value[0] ?? fallback;
    }
    return value.shift() ?? fallback;
}
function resolveJsonResponseKey(options) {
    const responseFormat = options.responseFormat;
    if (!responseFormat || responseFormat.type !== "json") {
        return "default";
    }
    const schema = responseFormat.schema;
    const properties = schema?.properties ?? {};
    if ("elementId" in properties && "twoStep" in properties) {
        return "act";
    }
    if ("elements" in properties) {
        return "Observation";
    }
    if ("completed" in properties && "progress" in properties) {
        return "Metadata";
    }
    return "Extraction";
}
export function promptToText(prompt) {
    return (prompt ?? [])
        .flatMap((message) => {
        if (typeof message.content === "string") {
            return [message.content];
        }
        return (message.content ?? [])
            .map((part) => (part.type === "text" ? part.text : ""))
            .filter((text) => text.length > 0);
    })
        .join("\n");
}
function findEncodedIds(options) {
    return [...promptToText(options.prompt).matchAll(/\b\d+-\d+\b/g)].map((match) => match[0]);
}
export function findEncodedIdForText(options, text) {
    const promptText = promptToText(options.prompt);
    const lines = promptText.split("\n");
    const line = lines.find((entry) => entry.includes(text));
    const match = line?.match(/\b\d+-\d+\b/);
    if (!match) {
        throw new Error(`Could not find encoded id for text: ${text}`);
    }
    return match[0];
}
export function findLastEncodedId(options) {
    const matches = findEncodedIds(options);
    if (matches.length === 0) {
        throw new Error("Could not find any encoded ids in the prompt.");
    }
    return matches[matches.length - 1];
}
export function toolCallResponse(toolName, input, toolCallId = `${toolName}-1`) {
    return {
        content: [
            {
                type: "tool-call",
                toolCallId,
                toolName,
                input: JSON.stringify(input),
            },
        ],
        finishReason: "tool-calls",
        usage: DEFAULT_USAGE,
    };
}
export function doneToolResponse(reasoning = "done", taskComplete = true, toolCallId = "done-1") {
    return toolCallResponse("done", { reasoning, taskComplete }, toolCallId);
}
function createGenerateResult(result) {
    return {
        content: result.content,
        finishReason: result.finishReason ?? "stop",
        usage: mergeUsage(result.usage),
        warnings: [],
    };
}
export function createScriptedAisdkTestLlmClient(options) {
    const jsonResponses = Object.fromEntries(Object.entries(options?.jsonResponses ?? {}).map(([key, value]) => [
        key,
        Array.isArray(value) ? [...value] : value,
    ]));
    const generateResponses = [...(options?.generateResponses ?? [])];
    const model = {
        provider: "mock",
        modelId: options?.modelId ?? "mock/stagehand-flow-logger",
        specificationVersion: "v2",
        supportedUrls: {},
        doGenerateCalls: [],
        doGenerate: async (callOptions) => {
            model.doGenerateCalls.push(callOptions);
            if (callOptions.responseFormat?.type === "json") {
                const key = resolveJsonResponseKey(callOptions);
                const responseScripts = consumeScriptValue(jsonResponses[key], jsonResponses.default);
                const responseScript = consumeScriptValue(responseScripts, undefined);
                const response = typeof responseScript === "function"
                    ? responseScript(callOptions)
                    : (responseScript ?? {});
                return createGenerateResult({
                    content: [{ type: "text", text: JSON.stringify(response) }],
                });
            }
            const responseScript = consumeScriptValue(generateResponses, undefined);
            if (!responseScript) {
                return createGenerateResult({
                    content: [{ type: "text", text: "done" }],
                });
            }
            const response = typeof responseScript === "function"
                ? responseScript(callOptions)
                : responseScript;
            return createGenerateResult(response);
        },
        doStream: async () => {
            throw new Error("Streaming is not implemented for this test model.");
        },
    };
    return new AISdkClient({ model });
}
