// @browserbasehq/stagehand/dist/cjs/tests/cache-variables.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const ActCache_1 = require("../lib/v3/cache/ActCache");
function createFakeStorage(entry) {
    return {
        enabled: true,
        readJson: vitest_1.vi.fn().mockResolvedValue({ value: entry }),
        writeJson: vitest_1.vi.fn().mockResolvedValue({}),
        directory: "/tmp/cache",
    };
}
(0, vitest_1.describe)("ActCache variable handling", () => {
    (0, vitest_1.it)("cache key includes variable keys but not values", async () => {
        const storage = {
            enabled: true,
            readJson: vitest_1.vi.fn(),
            writeJson: vitest_1.vi.fn().mockResolvedValue({}),
            directory: "/tmp/cache",
        };
        const cache = new ActCache_1.ActCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => null,
            getDefaultLlmClient: () => ({}),
            domSettleTimeoutMs: undefined,
        });
        const fakePage = {
            url: vitest_1.vi.fn().mockResolvedValue("https://example.com"),
        };
        // First context with username="user1@example.com"
        const context1 = await cache.prepareContext("type %username% into the email field", fakePage, { username: "user1@example.com" });
        // Second context with username="user2@example.com"
        const context2 = await cache.prepareContext("type %username% into the email field", fakePage, { username: "user2@example.com" });
        // Third context with different variable key name
        const context3 = await cache.prepareContext("type %email% into the email field", fakePage, { email: "user3@example.com" });
        // Same instruction + same variable keys = same cache key
        (0, vitest_1.expect)(context1?.cacheKey).toBe(context2?.cacheKey);
        // Different variable keys = different cache key
        (0, vitest_1.expect)(context1?.cacheKey).not.toBe(context3?.cacheKey);
        // Verify variable keys are sorted and stored
        (0, vitest_1.expect)(context1?.variableKeys).toEqual(["username"]);
        (0, vitest_1.expect)(context2?.variableKeys).toEqual(["username"]);
        (0, vitest_1.expect)(context3?.variableKeys).toEqual(["email"]);
        // Verify variable values are preserved in context
        (0, vitest_1.expect)(context1?.variables).toEqual({ username: "user1@example.com" });
        (0, vitest_1.expect)(context2?.variables).toEqual({ username: "user2@example.com" });
    });
    (0, vitest_1.it)("replays cached actions with variable substitution", async () => {
        // Cached action contains variable placeholder %username%
        const action = {
            selector: "xpath=/html/body/input[@type='email']",
            description: "type username into email field",
            method: "type",
            arguments: ["%username%"], // Variable placeholder
        };
        const entry = {
            version: 1,
            instruction: "type %username% into the email field",
            url: "https://example.com",
            variableKeys: ["username"],
            actions: [action],
            actionDescription: "type username",
            message: "done",
        };
        const storage = createFakeStorage(entry);
        // Track what variables are passed to takeDeterministicAction
        const capturedVariables = [];
        const handler = {
            takeDeterministicAction: vitest_1.vi
                .fn()
                .mockImplementation(async (_action, _page, _timeout, _client, _ensure, variables) => {
                capturedVariables.push(variables || {});
                return {
                    success: true,
                    message: "ok",
                    actionDescription: "type username",
                    actions: [action],
                };
            }),
        };
        const defaultClient = {};
        const cache = new ActCache_1.ActCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => handler,
            getDefaultLlmClient: () => defaultClient,
            domSettleTimeoutMs: undefined,
        });
        // First replay with username="user1@example.com"
        const context1 = {
            instruction: "type %username% into the email field",
            cacheKey: "test-key",
            pageUrl: "https://example.com",
            variableKeys: ["username"],
            variables: { username: "user1@example.com" },
        };
        const result1 = await cache.tryReplay(context1, {});
        (0, vitest_1.expect)(result1?.success).toBe(true);
        (0, vitest_1.expect)(handler.takeDeterministicAction).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(capturedVariables[0]).toEqual({ username: "user1@example.com" });
        // Reset
        vitest_1.vi.clearAllMocks();
        capturedVariables.length = 0;
        // Second replay with username="user2@example.com"
        const context2 = {
            instruction: "type %username% into the email field",
            cacheKey: "test-key", // Same cache key!
            pageUrl: "https://example.com",
            variableKeys: ["username"],
            variables: { username: "user2@example.com" },
        };
        const result2 = await cache.tryReplay(context2, {});
        (0, vitest_1.expect)(result2?.success).toBe(true);
        (0, vitest_1.expect)(handler.takeDeterministicAction).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(capturedVariables[0]).toEqual({ username: "user2@example.com" });
    });
    (0, vitest_1.it)("cache miss when variable keys don't match", async () => {
        const action = {
            selector: "xpath=/html/body/input",
            description: "type username",
            method: "type",
            arguments: ["%username%"],
        };
        // Cached entry expects "username" variable
        const entry = {
            version: 1,
            instruction: "type %username% into the field",
            url: "https://example.com",
            variableKeys: ["username"],
            actions: [action],
        };
        const storage = createFakeStorage(entry);
        const cache = new ActCache_1.ActCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => null,
            getDefaultLlmClient: () => ({}),
            domSettleTimeoutMs: undefined,
        });
        // Context has different variable key "email"
        const context = {
            instruction: "type %email% into the field",
            cacheKey: "test-key",
            pageUrl: "https://example.com",
            variableKeys: ["email"],
            variables: { email: "test@example.com" },
        };
        const result = await cache.tryReplay(context, {});
        // Should return null (cache miss) due to variable key mismatch
        (0, vitest_1.expect)(result).toBeNull();
    });
    (0, vitest_1.it)("cache miss when required variables are missing", async () => {
        const action = {
            selector: "xpath=/html/body/input",
            description: "type username",
            method: "type",
            arguments: ["%username%"],
        };
        const entry = {
            version: 1,
            instruction: "type %username% into the field",
            url: "https://example.com",
            variableKeys: ["username"],
            actions: [action],
        };
        const storage = createFakeStorage(entry);
        const logger = vitest_1.vi.fn();
        const cache = new ActCache_1.ActCache({
            storage,
            logger,
            getActHandler: () => null,
            getDefaultLlmClient: () => ({}),
            domSettleTimeoutMs: undefined,
        });
        // Context missing the username variable value
        const context = {
            instruction: "type %username% into the field",
            cacheKey: "test-key",
            pageUrl: "https://example.com",
            variableKeys: ["username"],
            variables: {}, // Missing username value!
        };
        const result = await cache.tryReplay(context, {});
        // Should return null (cache miss)
        (0, vitest_1.expect)(result).toBeNull();
        // Should log the miss reason
        (0, vitest_1.expect)(logger).toHaveBeenCalledWith(vitest_1.expect.objectContaining({
            category: "cache",
            message: "act cache miss: missing variables for replay",
            level: 2,
        }));
    });
    (0, vitest_1.it)("handles multiple variables correctly", async () => {
        const storage = {
            enabled: true,
            readJson: vitest_1.vi.fn(),
            writeJson: vitest_1.vi.fn().mockResolvedValue({}),
            directory: "/tmp/cache",
        };
        const cache = new ActCache_1.ActCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => null,
            getDefaultLlmClient: () => ({}),
            domSettleTimeoutMs: undefined,
        });
        const fakePage = {
            url: vitest_1.vi.fn().mockResolvedValue("https://example.com"),
        };
        // Context with multiple variables
        const context1 = await cache.prepareContext("fill %username% and %password%", fakePage, { username: "user1", password: "pass1" });
        const context2 = await cache.prepareContext("fill %username% and %password%", fakePage, { username: "user2", password: "pass2" });
        // Same cache key despite different values
        (0, vitest_1.expect)(context1?.cacheKey).toBe(context2?.cacheKey);
        // Variable keys should be sorted
        (0, vitest_1.expect)(context1?.variableKeys).toEqual(["password", "username"]);
        (0, vitest_1.expect)(context2?.variableKeys).toEqual(["password", "username"]);
    });
});
