// @browserbasehq/stagehand/dist/cjs/tests/unit/agent-captcha-hooks.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const captchaSolver_js_1 = require("../../lib/v3/agent/utils/captchaSolver.js");
const v3AgentHandler_js_1 = require("../../lib/v3/handlers/v3AgentHandler.js");
const SOLVING_STARTED = "browserbase-solving-started";
const SOLVING_FINISHED = "browserbase-solving-finished";
const SOLVING_ERRORED = "browserbase-solving-errored";
class MockPage {
    listeners = new Set();
    captchaBoxes = [];
    on(event, listener) {
        if (event === "console") {
            this.listeners.add(listener);
        }
    }
    off(event, listener) {
        if (event === "console") {
            this.listeners.delete(listener);
        }
    }
    emitConsole(text) {
        const message = { text: () => text };
        for (const listener of this.listeners) {
            listener(message);
        }
    }
    url() {
        return "https://example.com";
    }
    async screenshot() {
        return Buffer.from("fake-image");
    }
    async evaluate() {
        return this.captchaBoxes;
    }
    mainFrame() {
        return {
            evaluate: async () => ({ w: 1288, h: 711 }),
        };
    }
}
class FakeCuaClient {
    contextNotes = [];
    preStepHook;
    actionHandler;
    executeImpl = vitest_1.vi.fn(async (options) => {
        void options;
        return {
            success: true,
            message: "ok",
            actions: [],
            completed: true,
        };
    });
    captureScreenshot = vitest_1.vi.fn(async () => null);
    setViewport = vitest_1.vi.fn();
    setCurrentUrl = vitest_1.vi.fn();
    setScreenshotProvider = vitest_1.vi.fn();
    setSafetyConfirmationHandler = vitest_1.vi.fn();
    setActionHandler(handler) {
        this.actionHandler = handler;
    }
    setPreStepHook(handler) {
        this.preStepHook = handler;
    }
    addContextNote(note) {
        this.contextNotes.push(note);
    }
    async execute(options) {
        return this.executeImpl(options);
    }
}
let fakeCuaClient;
vitest_1.vi.mock("../../lib/v3/agent/AgentProvider", () => ({
    AgentProvider: class {
        constructor(logger) {
            void logger;
        }
        getClient() {
            return fakeCuaClient;
        }
    },
}));
const v3CuaAgentHandler_js_1 = require("../../lib/v3/handlers/v3CuaAgentHandler.js");
function collectUserMessages(messages) {
    return messages.filter((message) => message.role === "user" && typeof message.content === "string");
}
(0, vitest_1.describe)("agent captcha hooks", () => {
    let page;
    let logs;
    let logger;
    (0, vitest_1.beforeEach)(() => {
        page = new MockPage();
        logs = [];
        logger = (line) => {
            logs.push(line);
        };
        fakeCuaClient = new FakeCuaClient();
    });
    (0, vitest_1.it)("blocks regular agent prepareStep until the solver finishes and injects one solved message", async () => {
        const handler = new v3AgentHandler_js_1.V3AgentHandler({
            isCaptchaAutoSolveEnabled: true,
        }, logger, {});
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => page);
        const userCallback = vitest_1.vi.fn(async (options) => options);
        const prepareStep = handler.createPrepareStep(userCallback, solver);
        const options = {
            messages: [{ role: "user", content: "start" }],
        };
        await prepareStep(options);
        page.emitConsole(SOLVING_STARTED);
        const secondCall = prepareStep(options);
        await Promise.resolve();
        (0, vitest_1.expect)(userCallback).toHaveBeenCalledTimes(1);
        page.emitConsole(SOLVING_FINISHED);
        await secondCall;
        (0, vitest_1.expect)(userCallback).toHaveBeenCalledTimes(2);
        (0, vitest_1.expect)(collectUserMessages(options.messages).filter((message) => message.content.includes("automatically detected and solved"))).toHaveLength(1);
    });
    (0, vitest_1.it)("injects one error message when the regular agent solver errors", async () => {
        const handler = new v3AgentHandler_js_1.V3AgentHandler({
            isCaptchaAutoSolveEnabled: true,
        }, logger, {});
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => page);
        const prepareStep = handler.createPrepareStep(undefined, solver);
        const options = {
            messages: [{ role: "user", content: "start" }],
        };
        await prepareStep(options);
        page.emitConsole(SOLVING_STARTED);
        const pending = prepareStep(options);
        page.emitConsole(SOLVING_ERRORED);
        await pending;
        (0, vitest_1.expect)(collectUserMessages(options.messages).filter((message) => message.content.includes("automatic captcha solver failed"))).toHaveLength(1);
    });
    (0, vitest_1.it)("pauses the CUA loop at prepareStep while Browserbase solves a captcha", async () => {
        let secondPrepareStarted = false;
        fakeCuaClient.executeImpl = vitest_1.vi.fn(async () => {
            await fakeCuaClient.preStepHook?.();
            page.emitConsole(SOLVING_STARTED);
            const blockedPrepare = fakeCuaClient.preStepHook?.() ?? Promise.resolve();
            secondPrepareStarted = true;
            await blockedPrepare;
            return {
                success: true,
                message: "ok",
                actions: [],
                completed: true,
            };
        });
        const handler = new v3CuaAgentHandler_js_1.V3CuaAgentHandler({
            context: {
                awaitActivePage: async () => page,
            },
            bus: { emit: vitest_1.vi.fn() },
            isCaptchaAutoSolveEnabled: true,
            isAdvancedStealth: false,
            configuredViewport: { width: 1288, height: 711 },
            isAgentReplayActive: () => false,
            updateMetrics: vitest_1.vi.fn(),
        }, logger, {
            modelName: "anthropic/claude-haiku-4-5-20251001",
            clientOptions: { waitBetweenActions: 1 },
        });
        const execution = handler.execute({
            instruction: "Describe the page briefly.",
            highlightCursor: false,
        });
        await vitest_1.vi.waitFor(() => {
            (0, vitest_1.expect)(secondPrepareStarted).toBe(true);
            (0, vitest_1.expect)(logs.some((line) => line.message.includes("waiting for Browserbase to solve"))).toBe(true);
        });
        (0, vitest_1.expect)(logs.some((line) => line.message.includes("Captcha solved"))).toBe(false);
        page.emitConsole(SOLVING_FINISHED);
        await execution;
        (0, vitest_1.expect)(fakeCuaClient.contextNotes).toEqual([
            vitest_1.expect.stringContaining("automatically detected and solved"),
        ]);
        (0, vitest_1.expect)(logs.some((line) => line.message.includes("Captcha solved"))).toBe(true);
    });
    (0, vitest_1.it)("pauses CUA actions until the captcha solver finishes", async () => {
        let actionStarted = false;
        fakeCuaClient.executeImpl = vitest_1.vi.fn(async () => {
            await fakeCuaClient.preStepHook?.();
            page.emitConsole(SOLVING_STARTED);
            const pendingAction = fakeCuaClient.actionHandler?.({ type: "screenshot" }) ??
                Promise.resolve();
            actionStarted = true;
            await pendingAction;
            return {
                success: true,
                message: "ok",
                actions: [],
                completed: true,
            };
        });
        const handler = new v3CuaAgentHandler_js_1.V3CuaAgentHandler({
            context: {
                awaitActivePage: async () => page,
            },
            bus: { emit: vitest_1.vi.fn() },
            isCaptchaAutoSolveEnabled: true,
            isAdvancedStealth: false,
            configuredViewport: { width: 1288, height: 711 },
            isAgentReplayActive: () => false,
            updateMetrics: vitest_1.vi.fn(),
        }, logger, {
            modelName: "anthropic/claude-haiku-4-5-20251001",
            clientOptions: { waitBetweenActions: 1 },
        });
        const executeActionSpy = vitest_1.vi
            .spyOn(handler, "executeAction")
            .mockResolvedValue({ success: true });
        vitest_1.vi.spyOn(handler, "captureAndSendScreenshot").mockResolvedValue(null);
        const execution = handler.execute({
            instruction: "Describe the page briefly.",
            highlightCursor: false,
        });
        await vitest_1.vi.waitFor(() => {
            (0, vitest_1.expect)(actionStarted).toBe(true);
        });
        (0, vitest_1.expect)(executeActionSpy).not.toHaveBeenCalled();
        page.emitConsole(SOLVING_FINISHED);
        await execution;
        (0, vitest_1.expect)(executeActionSpy).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(fakeCuaClient.contextNotes).toEqual([
            vitest_1.expect.stringContaining("automatically detected and solved"),
        ]);
        (0, vitest_1.expect)(logs.some((line) => line.message.includes("Captcha solved"))).toBe(true);
    });
    (0, vitest_1.it)("skips post-solve clicks on the captcha widget and injects another note", async () => {
        page.captchaBoxes = [{ left: 0, top: 400, right: 140, bottom: 470 }];
        fakeCuaClient.executeImpl = vitest_1.vi.fn(async () => {
            await fakeCuaClient.preStepHook?.();
            page.emitConsole(SOLVING_STARTED);
            const blockedPrepare = fakeCuaClient.preStepHook?.() ?? Promise.resolve();
            page.emitConsole(SOLVING_FINISHED);
            await blockedPrepare;
            await fakeCuaClient.actionHandler?.({
                type: "click",
                button: "left",
                x: 63,
                y: 436,
            });
            return {
                success: true,
                message: "ok",
                actions: [],
                completed: true,
            };
        });
        const handler = new v3CuaAgentHandler_js_1.V3CuaAgentHandler({
            context: {
                awaitActivePage: async () => page,
            },
            bus: { emit: vitest_1.vi.fn() },
            isCaptchaAutoSolveEnabled: true,
            isAdvancedStealth: false,
            configuredViewport: { width: 1288, height: 711 },
            isAgentReplayActive: () => false,
            updateMetrics: vitest_1.vi.fn(),
        }, logger, {
            modelName: "anthropic/claude-haiku-4-5-20251001",
            clientOptions: { waitBetweenActions: 1 },
        });
        const executeActionSpy = vitest_1.vi
            .spyOn(handler, "executeAction")
            .mockResolvedValue({ success: true });
        vitest_1.vi.spyOn(handler, "captureAndSendScreenshot").mockResolvedValue(null);
        await handler.execute({
            instruction: "Describe the page briefly.",
            highlightCursor: false,
        });
        (0, vitest_1.expect)(executeActionSpy).not.toHaveBeenCalled();
        (0, vitest_1.expect)(fakeCuaClient.contextNotes).toEqual([
            vitest_1.expect.stringContaining("automatically detected and solved"),
            vitest_1.expect.stringContaining("Original task: Describe the page briefly."),
        ]);
        (0, vitest_1.expect)(logs.some((line) => line.message.includes("Skipped click on solved captcha widget"))).toBe(true);
    });
});
