// @browserbasehq/stagehand/dist/cjs/tests/unit/api-client-observe-variables.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const api_js_1 = require("../../lib/v3/api.js");
(0, vitest_1.describe)("StagehandAPIClient variable serialization", () => {
    (0, vitest_1.it)("preserves rich variables when sending the act request", async () => {
        const client = new api_js_1.StagehandAPIClient({
            apiKey: "bb-test",
            logger: vitest_1.vi.fn(),
        });
        const executeMock = vitest_1.vi.fn().mockResolvedValue({
            success: true,
            message: "ok",
            actionDescription: "typed",
            actions: [],
        });
        client.execute = executeMock;
        await client.act({
            input: "type %username% into the email field",
            options: {
                variables: {
                    username: {
                        value: "john@example.com",
                        description: "The login email",
                    },
                    password: "secret",
                },
            },
        });
        (0, vitest_1.expect)(executeMock).toHaveBeenCalledWith({
            method: "act",
            args: {
                input: "type %username% into the email field",
                options: {
                    variables: {
                        username: {
                            value: "john@example.com",
                            description: "The login email",
                        },
                        password: "secret",
                    },
                },
                frameId: undefined,
            },
            serverCache: undefined,
        });
    });
    (0, vitest_1.it)("preserves rich variables when sending the observe request", async () => {
        const client = new api_js_1.StagehandAPIClient({
            apiKey: "bb-test",
            logger: vitest_1.vi.fn(),
        });
        const executeMock = vitest_1.vi.fn().mockResolvedValue([]);
        client.execute = executeMock;
        await client.observe({
            instruction: "find the field where %username% should be entered",
            options: {
                variables: {
                    username: {
                        value: "john@example.com",
                        description: "The login email",
                    },
                    password: "secret",
                },
            },
        });
        (0, vitest_1.expect)(executeMock).toHaveBeenCalledWith({
            method: "observe",
            args: {
                instruction: "find the field where %username% should be entered",
                options: {
                    variables: {
                        username: {
                            value: "john@example.com",
                            description: "The login email",
                        },
                        password: "secret",
                    },
                },
                frameId: undefined,
            },
            serverCache: undefined,
        });
    });
});
