// @browserbasehq/stagehand/dist/cjs/tests/unit/api-optional-model-api-key.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const api_1 = require("../../lib/v3/api");
/**
 * Tests that modelApiKey is optional when calling StagehandAPIClient.init().
 *
 * Previously, init() would throw "modelApiKey is required" if the key was not
 * provided. After the fix, sessions can be started without a model API key
 * (the server may provide its own key or the user may not need one).
 * When provided, the key should still be sent via the x-model-api-key header.
 */
(0, vitest_1.describe)("StagehandAPIClient - optional modelApiKey", () => {
    const logger = vitest_1.vi.fn();
    // We mock fetch to avoid real network calls; we just need to verify
    // that init() doesn't throw when modelApiKey is omitted and that
    // the header is conditionally included.
    let originalFetch;
    function createSessionStartResponse(sessionId) {
        return new Response(JSON.stringify({
            success: true,
            data: { sessionId, available: true },
        }), {
            status: 200,
            headers: { "Content-Type": "application/json" },
        });
    }
    (0, vitest_1.beforeEach)(() => {
        originalFetch = globalThis.fetch;
    });
    (0, vitest_1.afterEach)(() => {
        globalThis.fetch = originalFetch;
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("should NOT throw when modelApiKey is omitted", async () => {
        globalThis.fetch = vitest_1.vi
            .fn()
            .mockResolvedValue(createSessionStartResponse("sess-123"));
        const client = new api_1.StagehandAPIClient({
            apiKey: "test-api-key",
            logger,
        });
        // Should not throw "modelApiKey is required"
        await (0, vitest_1.expect)(client.init({
            modelName: "openai/gpt-4.1-mini",
        })).resolves.toBeDefined();
    });
    (0, vitest_1.it)("should NOT throw when modelApiKey is undefined", async () => {
        globalThis.fetch = vitest_1.vi
            .fn()
            .mockResolvedValue(createSessionStartResponse("sess-456"));
        const client = new api_1.StagehandAPIClient({
            apiKey: "test-api-key",
            logger,
        });
        await (0, vitest_1.expect)(client.init({
            modelName: "openai/gpt-4.1-mini",
            modelApiKey: undefined,
        })).resolves.toBeDefined();
    });
    (0, vitest_1.it)("should send x-model-api-key header when modelApiKey IS provided", async () => {
        const fetchSpy = vitest_1.vi
            .fn()
            .mockResolvedValue(createSessionStartResponse("sess-789"));
        globalThis.fetch = fetchSpy;
        const client = new api_1.StagehandAPIClient({
            apiKey: "test-api-key",
            logger,
        });
        await client.init({
            modelName: "openai/gpt-4.1-mini",
            modelApiKey: "my-model-key",
        });
        // Verify the fetch was called with x-model-api-key header
        const [, requestInit] = fetchSpy.mock.calls[0];
        (0, vitest_1.expect)(requestInit.headers["x-model-api-key"]).toBe("my-model-key");
    });
    (0, vitest_1.it)("should NOT send x-model-api-key header when modelApiKey is omitted", async () => {
        const fetchSpy = vitest_1.vi
            .fn()
            .mockResolvedValue(createSessionStartResponse("sess-012"));
        globalThis.fetch = fetchSpy;
        const client = new api_1.StagehandAPIClient({
            apiKey: "test-api-key",
            logger,
        });
        await client.init({
            modelName: "openai/gpt-4.1-mini",
        });
        // Verify x-model-api-key header is NOT present
        const [, requestInit] = fetchSpy.mock.calls[0];
        (0, vitest_1.expect)(requestInit.headers["x-model-api-key"]).toBeUndefined();
    });
});
