// @browserbasehq/stagehand/dist/cjs/tests/unit/api-variables-schema.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const index_js_1 = require("../../lib/v3/types/public/index.js");
(0, vitest_1.describe)("API variable schemas", () => {
    (0, vitest_1.it)("accepts rich variables for act requests", () => {
        const result = index_js_1.Api.ActRequestSchema.safeParse({
            input: "type %username% into the email field",
            options: {
                variables: {
                    username: {
                        value: "john@example.com",
                        description: "The login email",
                    },
                    rememberMe: true,
                },
            },
        });
        (0, vitest_1.expect)(result.success).toBe(true);
    });
    (0, vitest_1.it)("accepts rich variables for observe requests", () => {
        const result = index_js_1.Api.ObserveRequestSchema.safeParse({
            instruction: "find the field where %username% should be entered",
            options: {
                variables: {
                    username: {
                        value: "john@example.com",
                        description: "The login email",
                    },
                    rememberMe: true,
                },
            },
        });
        (0, vitest_1.expect)(result.success).toBe(true);
    });
});
