// @browserbasehq/stagehand/dist/cjs/tests/unit/cache-llm-resolution.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const ActCache_js_1 = require("../../lib/v3/cache/ActCache.js");
const AgentCache_js_1 = require("../../lib/v3/cache/AgentCache.js");
function createFakeStorage(entry) {
    return {
        enabled: true,
        readJson: vitest_1.vi.fn().mockResolvedValue({ value: entry }),
        writeJson: vitest_1.vi.fn().mockResolvedValue({}),
        directory: "/tmp/cache",
    };
}
(0, vitest_1.describe)("Cache LLM client selection", () => {
    (0, vitest_1.it)("ActCache uses provided override client during replay", async () => {
        const action = {
            selector: "xpath=/html/body/button",
            description: "click button",
            method: "click",
            arguments: [],
        };
        const entry = {
            version: 1,
            instruction: "click button",
            url: "https://example.com",
            variableKeys: [],
            actions: [action],
            actionDescription: "click button",
            message: "done",
        };
        const storage = createFakeStorage(entry);
        const handler = {
            takeDeterministicAction: vitest_1.vi.fn().mockResolvedValue({
                success: true,
                message: "ok",
                actionDescription: "click button",
                actions: [action],
            }),
        };
        const defaultClient = { id: "default" };
        const overrideClient = { id: "override" };
        const cache = new ActCache_js_1.ActCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => handler,
            getDefaultLlmClient: () => defaultClient,
            domSettleTimeoutMs: undefined,
        });
        const context = {
            instruction: "click button",
            cacheKey: "abc",
            pageUrl: "https://example.com",
            variableKeys: [],
            variables: undefined,
        };
        const result = await cache.tryReplay(context, {}, undefined, overrideClient);
        (0, vitest_1.expect)(result?.success).toBe(true);
        (0, vitest_1.expect)(handler.takeDeterministicAction).toHaveBeenCalledTimes(1);
        const call = vitest_1.vi.mocked(handler.takeDeterministicAction).mock.calls[0];
        (0, vitest_1.expect)(call?.[3]).toBe(overrideClient);
    });
    (0, vitest_1.it)("AgentCache uses provided override client during replay", async () => {
        const action = {
            selector: "xpath=/html/body/input",
            description: "type email",
            method: "type",
            arguments: ["test@example.com"],
        };
        const agentStep = {
            type: "act",
            instruction: "type email",
            actions: [action],
        };
        const entry = {
            version: 1,
            instruction: "fill form",
            startUrl: "https://example.com",
            options: {},
            configSignature: "sig",
            steps: [agentStep],
            result: { success: true, actions: [] },
            timestamp: new Date().toISOString(),
        };
        const storage = {
            enabled: true,
            readJson: vitest_1.vi.fn().mockImplementation(async () => ({ value: entry })),
            writeJson: vitest_1.vi.fn().mockResolvedValue({}),
            directory: "/tmp/cache",
        };
        const handler = {
            takeDeterministicAction: vitest_1.vi.fn().mockResolvedValue({
                success: true,
                message: "ok",
                actionDescription: "type email",
                actions: [action],
            }),
        };
        const fakePage = {};
        const ctx = {
            awaitActivePage: vitest_1.vi.fn().mockResolvedValue(fakePage),
        };
        const defaultClient = { id: "default-agent" };
        const overrideClient = { id: "override-agent" };
        const cache = new AgentCache_js_1.AgentCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => handler,
            getContext: () => ctx,
            getDefaultLlmClient: () => defaultClient,
            getBaseModelName: () => "openai/gpt-4.1-mini",
            getSystemPrompt: () => undefined,
            domSettleTimeoutMs: undefined,
            act: vitest_1.vi.fn(),
        });
        const context = {
            instruction: "fill form",
            startUrl: "https://example.com",
            options: {},
            configSignature: "sig",
            cacheKey: "agent-key",
            variableKeys: [],
        };
        const result = await cache.tryReplay(context, overrideClient);
        (0, vitest_1.expect)(result?.success).toBe(true);
        (0, vitest_1.expect)(handler.takeDeterministicAction).toHaveBeenCalledTimes(1);
        const call = vitest_1.vi.mocked(handler.takeDeterministicAction).mock.calls[0];
        (0, vitest_1.expect)(call?.[3]).toBe(overrideClient);
    });
    (0, vitest_1.it)("AgentCache replays non-act steps without requiring an override client", async () => {
        const gotoEntry = {
            version: 1,
            instruction: "navigate home",
            startUrl: "https://example.com/source",
            options: {},
            configSignature: "sig",
            steps: [
                {
                    type: "goto",
                    url: "https://example.com/target",
                    waitUntil: "load",
                },
            ],
            result: { success: true, actions: [] },
            timestamp: new Date().toISOString(),
        };
        const storage = {
            enabled: true,
            readJson: vitest_1.vi.fn().mockResolvedValue({ value: gotoEntry }),
            writeJson: vitest_1.vi.fn().mockResolvedValue({}),
            directory: "/tmp/cache",
        };
        const handler = {
            takeDeterministicAction: vitest_1.vi.fn(),
        };
        const fakePage = { goto: vitest_1.vi.fn() };
        const ctx = {
            awaitActivePage: vitest_1.vi.fn().mockResolvedValue(fakePage),
        };
        const cache = new AgentCache_js_1.AgentCache({
            storage,
            logger: vitest_1.vi.fn(),
            getActHandler: () => handler,
            getContext: () => ctx,
            getDefaultLlmClient: () => ({ id: "default" }),
            getBaseModelName: () => "openai/gpt-4.1-mini",
            getSystemPrompt: () => undefined,
            domSettleTimeoutMs: undefined,
            act: vitest_1.vi.fn(),
        });
        const context = {
            instruction: "navigate home",
            startUrl: "https://example.com/source",
            options: {},
            configSignature: "sig",
            cacheKey: "agent-goto",
            variableKeys: [],
        };
        const result = await cache.tryReplay(context);
        (0, vitest_1.expect)(result?.success).toBe(true);
        (0, vitest_1.expect)(handler.takeDeterministicAction).not.toHaveBeenCalled();
        (0, vitest_1.expect)(fakePage.goto).toHaveBeenCalledWith("https://example.com/target", {
            waitUntil: "load",
        });
    });
});
