// @browserbasehq/stagehand/dist/cjs/tests/unit/captcha-solver.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const captchaSolver_js_1 = require("../../lib/v3/agent/utils/captchaSolver.js");
const SOLVING_STARTED = "browserbase-solving-started";
const SOLVING_FINISHED = "browserbase-solving-finished";
const SOLVING_ERRORED = "browserbase-solving-errored";
class MockPage {
    listeners = new Set();
    onCalls = 0;
    offCalls = 0;
    on(event, listener) {
        if (event !== "console")
            return;
        this.onCalls++;
        this.listeners.add(listener);
    }
    off(event, listener) {
        if (event !== "console")
            return;
        this.offCalls++;
        this.listeners.delete(listener);
    }
    emitConsole(text) {
        const message = { text: () => text };
        for (const listener of this.listeners) {
            listener(message);
        }
    }
    listenerCount() {
        return this.listeners.size;
    }
}
(0, vitest_1.describe)("CaptchaSolver", () => {
    (0, vitest_1.it)("resolves all concurrent waiters when a solve finishes", async () => {
        const page = new MockPage();
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => page);
        await solver.ensureAttached();
        page.emitConsole(SOLVING_STARTED);
        const firstWait = solver.waitIfSolving();
        const secondWait = solver.waitIfSolving();
        await new Promise((resolve) => setTimeout(resolve, 0));
        const sharedWaitPromise = solver.waitPromise;
        (0, vitest_1.expect)(sharedWaitPromise).not.toBeNull();
        (0, vitest_1.expect)(solver.waitPromise).toBe(sharedWaitPromise);
        let firstResolved = false;
        let secondResolved = false;
        void firstWait.then(() => {
            firstResolved = true;
        });
        void secondWait.then(() => {
            secondResolved = true;
        });
        await Promise.resolve();
        (0, vitest_1.expect)(firstResolved).toBe(false);
        (0, vitest_1.expect)(secondResolved).toBe(false);
        page.emitConsole(SOLVING_FINISHED);
        await Promise.all([firstWait, secondWait]);
        (0, vitest_1.expect)(firstResolved).toBe(true);
        (0, vitest_1.expect)(secondResolved).toBe(true);
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: true,
            errored: false,
        });
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: false,
            errored: false,
        });
    });
    (0, vitest_1.it)("re-attaches to a new page and settles stale waiters when the active page changes", async () => {
        const firstPage = new MockPage();
        const secondPage = new MockPage();
        let activePage = firstPage;
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => activePage);
        await solver.ensureAttached();
        firstPage.emitConsole(SOLVING_STARTED);
        const pendingWait = solver.waitIfSolving();
        let settled = false;
        void pendingWait.then(() => {
            settled = true;
        });
        activePage = secondPage;
        await solver.waitIfSolving();
        await pendingWait;
        (0, vitest_1.expect)(settled).toBe(true);
        (0, vitest_1.expect)(firstPage.offCalls).toBe(1);
        (0, vitest_1.expect)(firstPage.listenerCount()).toBe(0);
        (0, vitest_1.expect)(secondPage.onCalls).toBe(1);
        (0, vitest_1.expect)(secondPage.listenerCount()).toBe(1);
        (0, vitest_1.expect)(solver.isSolving()).toBe(false);
    });
    (0, vitest_1.it)("surfaces solver errors exactly once per consume", async () => {
        const page = new MockPage();
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => page);
        await solver.ensureAttached();
        page.emitConsole(SOLVING_STARTED);
        const wait = solver.waitIfSolving();
        page.emitConsole(SOLVING_ERRORED);
        await wait;
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: false,
            errored: true,
        });
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: false,
            errored: false,
        });
    });
    (0, vitest_1.it)("disposes cleanly while a solve is in progress", async () => {
        const page = new MockPage();
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => page);
        await solver.ensureAttached();
        page.emitConsole(SOLVING_STARTED);
        const wait = solver.waitIfSolving();
        await new Promise((resolve) => setTimeout(resolve, 0));
        let settled = false;
        void wait.then(() => {
            settled = true;
        });
        solver.dispose();
        await wait;
        (0, vitest_1.expect)(settled).toBe(true);
        (0, vitest_1.expect)(solver.isSolving()).toBe(false);
        (0, vitest_1.expect)(page.listenerCount()).toBe(0);
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: false,
            errored: false,
        });
    });
    (0, vitest_1.it)("marks errored when detached mid-solve due to page change", async () => {
        const firstPage = new MockPage();
        const secondPage = new MockPage();
        let activePage = firstPage;
        const solver = new captchaSolver_js_1.CaptchaSolver();
        solver.init(async () => activePage);
        await solver.ensureAttached();
        firstPage.emitConsole(SOLVING_STARTED);
        const wait = solver.waitIfSolving();
        // Switch to a new page while the solve is in progress
        activePage = secondPage;
        await solver.waitIfSolving();
        await wait;
        // The interrupted solve should be reported as errored
        (0, vitest_1.expect)(solver.consumeSolveResult()).toEqual({
            solved: false,
            errored: true,
        });
    });
});
