// @browserbasehq/stagehand/dist/cjs/tests/unit/flowlogger-capturing-llm.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const FlowLogger_js_1 = require("../../lib/v3/flowlogger/FlowLogger.js");
(0, vitest_1.describe)("flow logger llm logging", () => {
    (0, vitest_1.it)("no-ops direct llm logging calls when no flow context is active", () => {
        // These helpers are called from multiple model adapters, so they must stay
        // safe even when a test or utility invokes them outside any ALS flow scope.
        (0, vitest_1.expect)(() => FlowLogger_js_1.FlowLogger.logLlmRequest({
            requestId: "req-1",
            model: "mock-model",
            prompt: "hello",
        })).not.toThrow();
        (0, vitest_1.expect)(() => FlowLogger_js_1.FlowLogger.logLlmResponse({
            requestId: "req-1",
            model: "mock-model",
            output: "world",
            inputTokens: 1,
            outputTokens: 1,
        })).not.toThrow();
    });
    (0, vitest_1.it)("does not throw from llm middleware when no flow context is active", async () => {
        const middleware = FlowLogger_js_1.FlowLogger.createLlmLoggingMiddleware("mock-model");
        // Missing flow context should degrade to a silent no-op and preserve the
        // underlying model result.
        await (0, vitest_1.expect)(middleware.wrapGenerate({
            doGenerate: async () => ({
                text: "done",
                usage: {
                    inputTokens: 1,
                    outputTokens: 1,
                    totalTokens: 2,
                },
            }),
            params: {
                prompt: [],
            },
        })).resolves.toMatchObject({
            text: "done",
        });
    });
});
