// @browserbasehq/stagehand/dist/cjs/tests/unit/flowlogger-eventstore.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const EventStore_js_1 = require("../../lib/v3/flowlogger/EventStore.js");
const EventEmitter_js_1 = require("../../lib/v3/flowlogger/EventEmitter.js");
const FlowLogger_js_1 = require("../../lib/v3/flowlogger/FlowLogger.js");
function attachEventStoreToBus(store, bus) {
    const onFlowEvent = (event) => {
        if (event instanceof FlowLogger_js_1.FlowEvent) {
            void store.emit(event);
        }
    };
    bus.on("*", onFlowEvent);
    return () => {
        bus.off("*", onFlowEvent);
    };
}
function createVerboseStoreHarness() {
    const writes = [];
    process.stderr.write = ((chunk, cb) => {
        writes.push(String(chunk));
        cb?.(null);
        return true;
    });
    const store = new EventStore_js_1.EventStore("session-test");
    const bus = new EventEmitter_js_1.EventEmitterWithWildcardSupport();
    const detachBus = attachEventStoreToBus(store, bus);
    return { writes, store, bus, detachBus };
}
(0, vitest_1.describe)("flow logger event store", () => {
    const stderrWrite = process.stderr.write.bind(process.stderr);
    (0, vitest_1.afterEach)(() => {
        process.stderr.write = stderrWrite;
    });
    (0, vitest_1.it)("queries recent events from the default in-memory sink", async () => {
        const store = new EventStore_js_1.EventStore("session-test");
        await store.emit(new FlowLogger_js_1.FlowEvent({
            eventType: "StagehandExtractEvent",
            sessionId: "session-test",
            eventId: "stagehand-1234",
            eventCreatedAt: "2026-03-16T21:45:00.000Z",
            data: { params: ["grab title"] },
        }));
        const events = await store.query({});
        (0, vitest_1.expect)(events).toHaveLength(1);
        (0, vitest_1.expect)(events[0].eventType).toBe("StagehandExtractEvent");
        await store.destroy();
    });
    (0, vitest_1.it)("drops payloads from the default in-memory sink", async () => {
        const store = new EventStore_js_1.EventStore("session-test");
        await store.emit(new FlowLogger_js_1.FlowEvent({
            eventType: "LlmRequestEvent",
            sessionId: "session-test",
            eventId: "llm-1234",
            eventCreatedAt: "2026-03-16T21:45:00.000Z",
            data: {
                prompt: [{ type: "image_url", image_url: { url: "huge" } }],
                output: "huge",
            },
        }));
        const [event] = await store.query({});
        (0, vitest_1.expect)(event.eventType).toBe("LlmRequestEvent");
        (0, vitest_1.expect)(event.eventId).toBe("llm-1234");
        (0, vitest_1.expect)(event.data).toEqual({});
        await store.destroy();
    });
    (0, vitest_1.it)("renders semantic hierarchy tags for non-cdp stderr events only", async () => {
        // Intercept stderr so the pretty sink can be asserted without polluting the
        // real test runner output.
        const { writes, store, bus, detachBus } = createVerboseStoreHarness();
        const stepEvent = new FlowLogger_js_1.FlowEvent({
            eventType: "StagehandExtractEvent",
            sessionId: "session-test",
            eventId: "stagehand-1234",
            eventCreatedAt: "2026-03-16T21:45:00.000Z",
            data: { params: ["grab title"] },
        });
        const cdpEvent = new FlowLogger_js_1.FlowEvent({
            eventType: "CdpCallEvent",
            sessionId: "session-test",
            eventId: "cdp-call-5678",
            eventCreatedAt: "2026-03-16T21:45:00.100Z",
            eventParentIds: [stepEvent.eventId],
            data: {
                method: "Runtime.evaluate",
                params: { expression: "2 + 2" },
                targetId: "1234567890ABCDEF1234567890ABCDEF",
            },
        });
        // The stderr sink intentionally suppresses CDP noise even though the event
        // still exists for in-memory and file-backed sinks.
        bus.emit(stepEvent.eventType, stepEvent);
        bus.emit(cdpEvent.eventType, cdpEvent);
        await new Promise((resolve) => setTimeout(resolve, 0));
        (0, vitest_1.expect)(writes).toHaveLength(1);
        (0, vitest_1.expect)(writes[0]).toContain("[🆂 #1234 EXTRACT]");
        (0, vitest_1.expect)(writes[0]).toContain("Stagehand.extract");
        (0, vitest_1.expect)(writes[0]).not.toContain("Runtime.evaluate");
        detachBus();
        await store.destroy();
    });
    (0, vitest_1.it)("renders generic stagehand events without crashing the stderr sink", async () => {
        const { writes, store, bus, detachBus } = createVerboseStoreHarness();
        // `StagehandEvent` has no action suffix, so this guards the formatter path
        // that cannot assume a method name exists.
        bus.emit("StagehandEvent", new FlowLogger_js_1.FlowEvent({
            eventType: "StagehandEvent",
            sessionId: "session-test",
            eventId: "stagehand-0001",
            eventCreatedAt: "2026-03-16T21:45:00.000Z",
            data: { params: ["noop"] },
        }));
        await new Promise((resolve) => setTimeout(resolve, 0));
        (0, vitest_1.expect)(writes).toHaveLength(1);
        (0, vitest_1.expect)(writes[0]).toContain("[🆂 #0001");
        (0, vitest_1.expect)(writes[0]).toContain("Stagehand(");
        detachBus();
        await store.destroy();
    });
    (0, vitest_1.it)("colorizes pretty stderr output with ansi escapes when enabled", async () => {
        const previousForceColor = process.env.FORCE_COLOR;
        const previousNoColor = process.env.NO_COLOR;
        delete process.env.NO_COLOR;
        process.env.FORCE_COLOR = "1";
        const { writes, store, bus, detachBus } = createVerboseStoreHarness();
        try {
            bus.emit("StagehandActEvent", new FlowLogger_js_1.FlowEvent({
                eventType: "StagehandActEvent",
                sessionId: "session-test",
                eventId: "stagehand-0002",
                eventCreatedAt: "2026-03-16T21:45:00.000Z",
                data: { params: ["click submit"] },
            }));
            await new Promise((resolve) => setTimeout(resolve, 0));
            (0, vitest_1.expect)(writes).toHaveLength(1);
            (0, vitest_1.expect)(writes[0]).toContain("\u001B[");
        }
        finally {
            if (previousNoColor === undefined) {
                delete process.env.NO_COLOR;
            }
            else {
                process.env.NO_COLOR = previousNoColor;
            }
            if (previousForceColor === undefined) {
                delete process.env.FORCE_COLOR;
            }
            else {
                process.env.FORCE_COLOR = previousForceColor;
            }
            detachBus();
            await store.destroy();
        }
    });
    (0, vitest_1.it)("keeps agent ancestry and start ids for completion events after many child events", async () => {
        const { writes, store, bus, detachBus } = createVerboseStoreHarness();
        const agentEvent = new FlowLogger_js_1.FlowEvent({
            eventType: "AgentExecuteEvent",
            sessionId: "session-test",
            eventId: "agent-1234",
            eventCreatedAt: "2026-03-16T21:45:00.000Z",
            data: { params: [{ instruction: "click the button" }] },
        });
        const actEvent = new FlowLogger_js_1.FlowEvent({
            eventType: "StagehandActEvent",
            sessionId: "session-test",
            eventId: "stagehand-2222",
            eventCreatedAt: "2026-03-16T21:45:00.001Z",
            eventParentIds: [agentEvent.eventId],
            data: { params: ["click the button"] },
        });
        const clickEvent = new FlowLogger_js_1.FlowEvent({
            eventType: "UnderstudyClickEvent",
            sessionId: "session-test",
            eventId: "action-3333",
            eventCreatedAt: "2026-03-16T21:45:00.002Z",
            eventParentIds: [agentEvent.eventId, actEvent.eventId],
            data: { target: "xpath=/button[1]" },
        });
        bus.emit(agentEvent.eventType, agentEvent);
        bus.emit(actEvent.eventType, actEvent);
        bus.emit(clickEvent.eventType, clickEvent);
        // Flood the retained history with child events so the completion lines have
        // to recover their displayed ancestry from the queryable sink.
        for (let index = 0; index < 150; index += 1) {
            bus.emit("CdpCallEvent", new FlowLogger_js_1.FlowEvent({
                eventType: "CdpCallEvent",
                sessionId: "session-test",
                eventId: `cdp-${String(index).padStart(4, "0")}`,
                eventCreatedAt: `2026-03-16T21:45:00.${String(index + 10).padStart(3, "0")}Z`,
                eventParentIds: [
                    agentEvent.eventId,
                    actEvent.eventId,
                    clickEvent.eventId,
                ],
                data: {
                    method: "Runtime.evaluate",
                    params: { expression: `${index}` },
                    targetId: "1234567890ABCDEF1234567890ABCDEF",
                },
            }));
        }
        bus.emit("UnderstudyClickCompletedEvent", new FlowLogger_js_1.FlowEvent({
            eventType: "UnderstudyClickCompletedEvent",
            sessionId: "session-test",
            eventId: "done-4444",
            eventCreatedAt: "2026-03-16T21:45:01.000Z",
            eventParentIds: [
                agentEvent.eventId,
                actEvent.eventId,
                clickEvent.eventId,
            ],
            data: { durationMs: 250 },
        }));
        bus.emit("StagehandActCompletedEvent", new FlowLogger_js_1.FlowEvent({
            eventType: "StagehandActCompletedEvent",
            sessionId: "session-test",
            eventId: "done-5555",
            eventCreatedAt: "2026-03-16T21:45:01.001Z",
            eventParentIds: [agentEvent.eventId, actEvent.eventId],
            data: { durationMs: 500 },
        }));
        bus.emit("AgentExecuteCompletedEvent", new FlowLogger_js_1.FlowEvent({
            eventType: "AgentExecuteCompletedEvent",
            sessionId: "session-test",
            eventId: "done-6666",
            eventCreatedAt: "2026-03-16T21:45:01.002Z",
            eventParentIds: [agentEvent.eventId],
            data: { durationMs: 750 },
        }));
        await new Promise((resolve) => setTimeout(resolve, 0));
        // Completion lines should reference the original started-event ids, not the
        // synthetic completed-event ids emitted at the end of the lifecycle.
        const clickCompletedLine = writes.find((line) => line.includes("CLICK completed"));
        const actCompletedLine = writes.find((line) => line.includes("ACT completed"));
        const agentCompletedLine = writes.find((line) => line.includes("Agent.execute() completed"));
        (0, vitest_1.expect)(clickCompletedLine).toContain("[🅰 #1234]");
        (0, vitest_1.expect)(clickCompletedLine).toContain("[🆂 #2222 ACT]");
        (0, vitest_1.expect)(clickCompletedLine).toContain("[🆄 #3333 CLICK]");
        (0, vitest_1.expect)(clickCompletedLine).not.toContain("#4444");
        (0, vitest_1.expect)(actCompletedLine).toContain("[🅰 #1234]");
        (0, vitest_1.expect)(actCompletedLine).toContain("[🆂 #2222 ACT]");
        (0, vitest_1.expect)(actCompletedLine).not.toContain("#5555");
        (0, vitest_1.expect)(agentCompletedLine).toContain("[🅰 #1234]");
        (0, vitest_1.expect)(agentCompletedLine).not.toContain("#6666");
        detachBus();
        await store.destroy();
    });
});
