// @browserbasehq/stagehand/dist/cjs/tests/unit/flowlogger-capturing-cdp.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_events_1 = require("node:events");
const vitest_1 = require("vitest");
const cdp_js_1 = require("../../lib/v3/understudy/cdp.js");
const EventSink_js_1 = require("../../lib/v3/flowlogger/EventSink.js");
const EventEmitter_js_1 = require("../../lib/v3/flowlogger/EventEmitter.js");
const EventStore_js_1 = require("../../lib/v3/flowlogger/EventStore.js");
const FlowLogger_js_1 = require("../../lib/v3/flowlogger/FlowLogger.js");
function attachEventStoreToBus(store, bus) {
    const onFlowEvent = (event) => {
        if (event instanceof FlowLogger_js_1.FlowEvent) {
            void store.emit(event);
        }
    };
    bus.on("*", onFlowEvent);
    return () => {
        bus.off("*", onFlowEvent);
    };
}
class FakeSocket extends node_events_1.EventEmitter {
    sentPayloads = [];
    readyState = 1;
    send(payload) {
        this.sentPayloads.push(payload);
    }
    close() {
        this.readyState = 3;
        this.emit("close", 1000, "");
    }
}
function createConnection(socket) {
    // The production constructor is private; tests instantiate it directly so
    // they can drive raw websocket messages without a real browser.
    const ConnectionCtor = cdp_js_1.CdpConnection;
    return new ConnectionCtor(socket);
}
function requireEvent(events, predicate, description) {
    const match = events.find(predicate);
    (0, vitest_1.expect)(match, `missing ${description}`).toBeDefined();
    return match;
}
(0, vitest_1.describe)("flow logger cdp context", () => {
    (0, vitest_1.it)("preserves the active parent chain when a session event handler issues a nested CDP call", async () => {
        const sessionId = "session-test";
        const socket = new FakeSocket();
        const eventBus = new EventEmitter_js_1.EventEmitterWithWildcardSupport();
        const sink = new EventSink_js_1.InMemoryEventSink();
        const eventStore = new EventStore_js_1.EventStore(sessionId, undefined, sink);
        const detachBus = attachEventStoreToBus(eventStore, eventBus);
        const conn = createConnection(socket);
        conn.flowLoggerContext = FlowLogger_js_1.FlowLogger.init(sessionId, eventBus);
        // Seed the target/session mapping the same way a real attach flow would
        // before any session-scoped messages are dispatched.
        conn.onMessage(JSON.stringify({
            method: "Target.attachedToTarget",
            params: {
                sessionId: "target-session",
                targetInfo: { targetId: "target-1" },
            },
        }));
        const session = conn.getSession("target-session");
        (0, vitest_1.expect)(session).toBeDefined();
        session.on("Runtime.consoleAPICalled", () => {
            // This nested send used to lose its parent chain because the callback ran
            // after the original ALS scope had already unwound.
            void session.send("Runtime.evaluate", {
                expression: "2 + 2",
            });
        });
        await FlowLogger_js_1.FlowLogger.runWithLogging({
            context: conn.flowLoggerContext,
            eventType: "SyntheticParentEvent",
        }, async () => {
            void session.send("Page.navigate", {
                url: "https://example.com",
            });
        }, []);
        conn.onMessage(JSON.stringify({
            method: "Runtime.consoleAPICalled",
            sessionId: "target-session",
            params: { type: "log" },
        }));
        // The nested Runtime.evaluate call should still attach under the synthetic
        // parent event even though it was triggered by a later session callback.
        const events = await eventStore.query({});
        const parentEvent = requireEvent(events, (event) => event.eventType === "SyntheticParentEvent", "SyntheticParentEvent");
        const nestedCallEvent = requireEvent(events, (event) => event.eventType === "CdpCallEvent" &&
            String(event.data.method) === "Runtime.evaluate", "nested Runtime.evaluate CdpCallEvent");
        (0, vitest_1.expect)(nestedCallEvent.eventParentIds).toEqual([parentEvent.eventId]);
        detachBus();
        await eventStore.destroy();
    });
});
