// @browserbasehq/stagehand/dist/cjs/tests/unit/llm-middleware.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const LLMProvider_js_1 = require("../../lib/v3/llm/LLMProvider.js");
const v3_js_1 = require("../../lib/v3/v3.js");
/**
 * Creates a recording middleware that captures every doGenerate / doStream
 * invocation along with the model identity and returned usage data.
 */
function createRecordingMiddleware() {
    const calls = [];
    const middleware = {
        wrapGenerate: async ({ doGenerate, model }) => {
            const result = await doGenerate();
            calls.push({
                type: "generate",
                modelId: model.modelId,
                provider: model.provider,
                usage: {
                    inputTokens: result.usage.inputTokens ?? undefined,
                    outputTokens: result.usage.outputTokens ?? undefined,
                },
            });
            return result;
        },
        wrapStream: async ({ doStream, model }) => {
            const result = await doStream();
            calls.push({
                type: "stream",
                modelId: model.modelId,
                provider: model.provider,
            });
            return result;
        },
    };
    return { middleware, calls };
}
/**
 * Creates a minimal mock LanguageModelV2 that returns canned results
 * without hitting any real provider. Useful for testing the wrapping
 * mechanics in isolation.
 */
function createMockLanguageModel(modelId, provider) {
    return {
        specificationVersion: "v2",
        provider,
        modelId,
        defaultObjectGenerationMode: "json",
        doGenerate: vitest_1.vi.fn().mockResolvedValue({
            content: [{ type: "text", text: "mock response" }],
            finishReason: "stop",
            usage: { inputTokens: 10, outputTokens: 5, totalTokens: 15 },
            warnings: [],
        }),
        doStream: vitest_1.vi.fn().mockResolvedValue({
            stream: new ReadableStream({
                start(controller) {
                    controller.enqueue({
                        type: "finish",
                        finishReason: "stop",
                        usage: { inputTokens: 8, outputTokens: 3, totalTokens: 11 },
                    });
                    controller.close();
                },
            }),
        }),
    };
}
// ---------------------------------------------------------------------------
// getAISDKLanguageModel with middleware
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("getAISDKLanguageModel with middleware", () => {
    (0, vitest_1.it)("returns a model when no middleware is provided", () => {
        const model = (0, LLMProvider_js_1.getAISDKLanguageModel)("ollama", "llama3.2");
        (0, vitest_1.expect)(model).toBeDefined();
        (0, vitest_1.expect)(model.modelId).toBe("llama3.2");
    });
    (0, vitest_1.it)("returns a wrapped model when middleware is provided", () => {
        const { middleware } = createRecordingMiddleware();
        const model = (0, LLMProvider_js_1.getAISDKLanguageModel)("ollama", "llama3.2", undefined, middleware);
        (0, vitest_1.expect)(model).toBeDefined();
        // wrapLanguageModel preserves modelId
        (0, vitest_1.expect)(model.modelId).toBe("llama3.2");
    });
    (0, vitest_1.it)("wrapped model preserves doGenerate and doStream methods", () => {
        const { middleware } = createRecordingMiddleware();
        const model = (0, LLMProvider_js_1.getAISDKLanguageModel)("ollama", "llama3.2", undefined, middleware);
        (0, vitest_1.expect)(typeof model.doGenerate).toBe("function");
        (0, vitest_1.expect)(typeof model.doStream).toBe("function");
        (0, vitest_1.expect)(model.provider).toContain("ollama");
    });
});
// ---------------------------------------------------------------------------
// LLMProvider with middleware
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("LLMProvider with middleware", () => {
    const noop = () => { };
    (0, vitest_1.it)("creates an AISdkClient without middleware", () => {
        const provider = new LLMProvider_js_1.LLMProvider(noop);
        const client = provider.getClient("openai/gpt-4o");
        (0, vitest_1.expect)(client).toBeDefined();
        (0, vitest_1.expect)(client.type).toBe("aisdk");
    });
    (0, vitest_1.it)("creates an AISdkClient that carries middleware-wrapped model", () => {
        const { middleware } = createRecordingMiddleware();
        const provider = new LLMProvider_js_1.LLMProvider(noop, middleware);
        const client = provider.getClient("ollama/llama3.2");
        (0, vitest_1.expect)(client).toBeDefined();
        (0, vitest_1.expect)(client.type).toBe("aisdk");
        const languageModel = client.getLanguageModel();
        (0, vitest_1.expect)(languageModel).toBeDefined();
        // Wrapped models should still expose the original modelId
        (0, vitest_1.expect)(languageModel.modelId).toBe("llama3.2");
    });
    (0, vitest_1.it)("applies the same middleware to different models from getClient", () => {
        const { middleware } = createRecordingMiddleware();
        const provider = new LLMProvider_js_1.LLMProvider(noop, middleware);
        const clientA = provider.getClient("ollama/llama3.2");
        const clientB = provider.getClient("ollama/mistral");
        (0, vitest_1.expect)(clientA.getLanguageModel()).toBeDefined();
        (0, vitest_1.expect)(clientB.getLanguageModel()).toBeDefined();
        (0, vitest_1.expect)(clientA.getLanguageModel().modelId).toBe("llama3.2");
        (0, vitest_1.expect)(clientB.getLanguageModel().modelId).toBe("mistral");
    });
});
// ---------------------------------------------------------------------------
// Middleware captures usage across act/extract/observe/agent code paths
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("middleware captures usage from doGenerate and doStream", () => {
    (0, vitest_1.it)("wrapGenerate fires and captures usage on doGenerate", async () => {
        const { middleware, calls } = createRecordingMiddleware();
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        // Simulate what wrapLanguageModel does: import and wrap
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({ model: mockModel, middleware });
        await wrapped.doGenerate({
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "act: click button" }],
                },
            ],
        });
        (0, vitest_1.expect)(calls).toHaveLength(1);
        (0, vitest_1.expect)(calls[0].type).toBe("generate");
        (0, vitest_1.expect)(calls[0].modelId).toBe("gpt-4o");
        (0, vitest_1.expect)(calls[0].usage).toEqual({ inputTokens: 10, outputTokens: 5 });
    });
    (0, vitest_1.it)("wrapStream fires on doStream", async () => {
        const { middleware, calls } = createRecordingMiddleware();
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({ model: mockModel, middleware });
        const result = await wrapped.doStream({
            prompt: [
                { role: "user", content: [{ type: "text", text: "stream this" }] },
            ],
        });
        // Consume the stream to trigger the finish chunk
        const reader = result.stream.getReader();
        while (true) {
            const { done } = await reader.read();
            if (done)
                break;
        }
        (0, vitest_1.expect)(calls).toHaveLength(1);
        (0, vitest_1.expect)(calls[0].type).toBe("stream");
        (0, vitest_1.expect)(calls[0].modelId).toBe("gpt-4o");
    });
    (0, vitest_1.it)("middleware fires for each separate doGenerate call (simulates act + extract + observe)", async () => {
        const { middleware, calls } = createRecordingMiddleware();
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({ model: mockModel, middleware });
        const callOpts = {
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "test" }],
                },
            ],
        };
        // Simulate act call
        await wrapped.doGenerate(callOpts);
        // Simulate extract call
        await wrapped.doGenerate(callOpts);
        // Simulate observe call
        await wrapped.doGenerate(callOpts);
        (0, vitest_1.expect)(calls).toHaveLength(3);
        (0, vitest_1.expect)(calls.every((c) => c.type === "generate")).toBe(true);
        (0, vitest_1.expect)(calls.every((c) => c.modelId === "gpt-4o")).toBe(true);
    });
    (0, vitest_1.it)("middleware fires for agent multi-step calls (sequential doGenerate)", async () => {
        const { middleware, calls } = createRecordingMiddleware();
        const mockModel = createMockLanguageModel("openai/gpt-4.1", "gateway.openai");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({ model: mockModel, middleware });
        const callOpts = {
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "agent step" }],
                },
            ],
        };
        // Simulate multiple agent reasoning steps
        await wrapped.doGenerate(callOpts);
        await wrapped.doGenerate(callOpts);
        await wrapped.doGenerate(callOpts);
        await wrapped.doGenerate(callOpts);
        await wrapped.doGenerate(callOpts);
        (0, vitest_1.expect)(calls).toHaveLength(5);
        calls.forEach((c) => {
            (0, vitest_1.expect)(c.usage).toEqual({ inputTokens: 10, outputTokens: 5 });
        });
    });
});
// ---------------------------------------------------------------------------
// ModelConfiguration-level middleware (the user-facing shape)
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("middleware inside ModelConfiguration", () => {
    (0, vitest_1.it)("resolveModelConfiguration extracts middleware from object config", () => {
        const { middleware: mw } = createRecordingMiddleware();
        const result = (0, v3_js_1.resolveModelConfiguration)({
            modelName: "openai/gpt-4o",
            apiKey: "sk-test",
            middleware: mw,
        });
        (0, vitest_1.expect)(result.modelName).toBe("openai/gpt-4o");
        (0, vitest_1.expect)(result.middleware).toBe(mw);
        (0, vitest_1.expect)(result.clientOptions).toEqual({ apiKey: "sk-test" });
        (0, vitest_1.expect)(result.clientOptions && "middleware" in result.clientOptions).toBe(false);
    });
    (0, vitest_1.it)("string ModelConfiguration has no middleware", () => {
        const result = (0, v3_js_1.resolveModelConfiguration)("openai/gpt-4o");
        (0, vitest_1.expect)(result.modelName).toBe("openai/gpt-4o");
        (0, vitest_1.expect)(result.middleware).toBeUndefined();
        (0, vitest_1.expect)(result.clientOptions).toBeUndefined();
    });
    (0, vitest_1.it)("middleware is separated from clientOptions when resolving per-method overrides", () => {
        const { middleware: mw } = createRecordingMiddleware();
        const result = (0, v3_js_1.resolveModelConfiguration)({
            modelName: "anthropic/claude-sonnet-4-20250514",
            apiKey: "sk-ant-test",
            middleware: mw,
        });
        (0, vitest_1.expect)(result.modelName).toBe("anthropic/claude-sonnet-4-20250514");
        (0, vitest_1.expect)(result.middleware).toBe(mw);
        (0, vitest_1.expect)(result.clientOptions).toEqual({ apiKey: "sk-ant-test" });
        (0, vitest_1.expect)(result.clientOptions && "middleware" in result.clientOptions).toBe(false);
    });
});
// ---------------------------------------------------------------------------
// Practical middleware behaviors
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("middleware that tracks duration", () => {
    (0, vitest_1.it)("measures wall-clock time of doGenerate", async () => {
        const durations = [];
        const timingMiddleware = {
            wrapGenerate: async ({ doGenerate }) => {
                const start = performance.now();
                const result = await doGenerate();
                durations.push(performance.now() - start);
                return result;
            },
        };
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({
            model: mockModel,
            middleware: timingMiddleware,
        });
        await wrapped.doGenerate({
            prompt: [{ role: "user", content: [{ type: "text", text: "time me" }] }],
        });
        (0, vitest_1.expect)(durations).toHaveLength(1);
        (0, vitest_1.expect)(durations[0]).toBeGreaterThanOrEqual(0);
    });
});
(0, vitest_1.describe)("middleware that aggregates token usage", () => {
    (0, vitest_1.it)("sums tokens across multiple calls like a billing tracker", async () => {
        let totalInput = 0;
        let totalOutput = 0;
        const billingMiddleware = {
            wrapGenerate: async ({ doGenerate }) => {
                const result = await doGenerate();
                totalInput += result.usage.inputTokens ?? 0;
                totalOutput += result.usage.outputTokens ?? 0;
                return result;
            },
        };
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({
            model: mockModel,
            middleware: billingMiddleware,
        });
        const callOpts = {
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "go" }],
                },
            ],
        };
        // act
        await wrapped.doGenerate(callOpts);
        // extract
        await wrapped.doGenerate(callOpts);
        // observe
        await wrapped.doGenerate(callOpts);
        (0, vitest_1.expect)(totalInput).toBe(30);
        (0, vitest_1.expect)(totalOutput).toBe(15);
    });
});
(0, vitest_1.describe)("middleware that logs per-model call counts", () => {
    (0, vitest_1.it)("tracks which models were called and how many times", async () => {
        const modelCallCounts = new Map();
        const countingMiddleware = {
            wrapGenerate: async ({ doGenerate, model }) => {
                const key = `${model.provider}/${model.modelId}`;
                modelCallCounts.set(key, (modelCallCounts.get(key) ?? 0) + 1);
                return doGenerate();
            },
        };
        const modelA = createMockLanguageModel("gpt-4o", "openai.chat");
        const modelB = createMockLanguageModel("claude-sonnet-4-20250514", "anthropic.messages");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrappedA = wrapLanguageModel({
            model: modelA,
            middleware: countingMiddleware,
        });
        const wrappedB = wrapLanguageModel({
            model: modelB,
            middleware: countingMiddleware,
        });
        const callOpts = {
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "go" }],
                },
            ],
        };
        await wrappedA.doGenerate(callOpts);
        await wrappedA.doGenerate(callOpts);
        await wrappedB.doGenerate(callOpts);
        (0, vitest_1.expect)(modelCallCounts.get("openai.chat/gpt-4o")).toBe(2);
        (0, vitest_1.expect)(modelCallCounts.get("anthropic.messages/claude-sonnet-4-20250514")).toBe(1);
        (0, vitest_1.expect)(modelCallCounts.size).toBe(2);
    });
});
(0, vitest_1.describe)("middleware that detects errors", () => {
    (0, vitest_1.it)("catches and re-throws errors from doGenerate while still recording the failure", async () => {
        const errors = [];
        const errorTrackingMiddleware = {
            wrapGenerate: async ({ doGenerate }) => {
                try {
                    return await doGenerate();
                }
                catch (err) {
                    errors.push(err);
                    throw err;
                }
            },
        };
        const failingModel = {
            specificationVersion: "v2",
            provider: "openai.chat",
            modelId: "gpt-4o",
            defaultObjectGenerationMode: "json",
            doGenerate: vitest_1.vi.fn().mockRejectedValue(new Error("rate limit exceeded")),
            doStream: vitest_1.vi.fn(),
        };
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({
            model: failingModel,
            middleware: errorTrackingMiddleware,
        });
        await (0, vitest_1.expect)(wrapped.doGenerate({
            prompt: [{ role: "user", content: [{ type: "text", text: "fail" }] }],
        })).rejects.toThrow("rate limit exceeded");
        (0, vitest_1.expect)(errors).toHaveLength(1);
        (0, vitest_1.expect)(errors[0].message).toBe("rate limit exceeded");
    });
});
(0, vitest_1.describe)("chaining middleware across multiple wrapped models", () => {
    (0, vitest_1.it)("same middleware instance sees calls from different models", async () => {
        const seen = [];
        const spyMiddleware = {
            wrapGenerate: async ({ doGenerate, model }) => {
                seen.push(model.modelId);
                return doGenerate();
            },
        };
        const mockA = createMockLanguageModel("gpt-4o", "openai.chat");
        const mockB = createMockLanguageModel("claude-sonnet-4-20250514", "anthropic.messages");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrappedA = wrapLanguageModel({
            model: mockA,
            middleware: spyMiddleware,
        });
        const wrappedB = wrapLanguageModel({
            model: mockB,
            middleware: spyMiddleware,
        });
        const callOpts = {
            prompt: [
                {
                    role: "user",
                    content: [{ type: "text", text: "go" }],
                },
            ],
        };
        await wrappedA.doGenerate(callOpts);
        await wrappedB.doGenerate(callOpts);
        await wrappedA.doGenerate(callOpts);
        (0, vitest_1.expect)(seen).toEqual(["gpt-4o", "claude-sonnet-4-20250514", "gpt-4o"]);
    });
});
// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------
(0, vitest_1.describe)("middleware edge cases", () => {
    (0, vitest_1.it)("middleware that only implements wrapGenerate still works for doStream", async () => {
        const generateCalls = [];
        const partialMiddleware = {
            wrapGenerate: async ({ doGenerate, model }) => {
                generateCalls.push(model.modelId);
                return doGenerate();
            },
        };
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({
            model: mockModel,
            middleware: partialMiddleware,
        });
        const result = await wrapped.doStream({
            prompt: [{ role: "user", content: [{ type: "text", text: "stream" }] }],
        });
        (0, vitest_1.expect)(result.stream).toBeDefined();
        (0, vitest_1.expect)(generateCalls).toHaveLength(0);
    });
    (0, vitest_1.it)("middleware does not alter the response data", async () => {
        const { middleware } = createRecordingMiddleware();
        const mockModel = createMockLanguageModel("gpt-4o", "openai.chat");
        const { wrapLanguageModel } = await Promise.resolve().then(() => __importStar(require("ai")));
        const wrapped = wrapLanguageModel({ model: mockModel, middleware });
        const result = await wrapped.doGenerate({
            prompt: [{ role: "user", content: [{ type: "text", text: "test" }] }],
        });
        (0, vitest_1.expect)(result.content).toEqual([{ type: "text", text: "mock response" }]);
        (0, vitest_1.expect)(result.usage).toEqual({
            inputTokens: 10,
            outputTokens: 5,
            totalTokens: 15,
        });
        (0, vitest_1.expect)(result.finishReason).toBe("stop");
    });
});
