// @browserbasehq/stagehand/dist/cjs/tests/unit/openai-cua-client.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const OpenAICUAClient_js_1 = require("../../lib/v3/agent/OpenAICUAClient.js");
function createClient() {
    return new OpenAICUAClient_js_1.OpenAICUAClient("openai", "computer-use-preview-2025-03-11", undefined, { apiKey: "test-key" });
}
(0, vitest_1.describe)("OpenAICUAClient", () => {
    (0, vitest_1.it)("exposes captchaSolvedProceed tool after a captcha context note", () => {
        const client = createClient();
        // Before captcha note — tool should not be active
        (0, vitest_1.expect)(client
            .captchaSolvedToolActive).toBe(false);
        // Simulate a captcha context note being added (as the CUA handler does)
        client.addContextNote("A captcha was automatically detected and solved — no further interaction needed.");
        (0, vitest_1.expect)(client
            .captchaSolvedToolActive).toBe(true);
    });
    (0, vitest_1.it)("does NOT activate captcha tool for non-captcha context notes", () => {
        const client = createClient();
        client.addContextNote("The page has finished loading.");
        (0, vitest_1.expect)(client
            .captchaSolvedToolActive).toBe(false);
    });
    (0, vitest_1.it)("deactivates captcha tool after takeAction handles the function call", async () => {
        const client = createClient();
        client.addContextNote("A captcha was solved.");
        (0, vitest_1.expect)(client
            .captchaSolvedToolActive).toBe(true);
        // Simulate the model calling the captchaSolvedProceed tool
        const result = await client.takeAction([
            {
                type: "function_call",
                name: "captchaSolvedProceed",
                call_id: "call-1",
                arguments: "{}",
            },
        ], vitest_1.vi.fn());
        // Tool should be deactivated
        (0, vitest_1.expect)(client
            .captchaSolvedToolActive).toBe(false);
        // Result should contain a function_call_output confirming proceed
        (0, vitest_1.expect)(result).toEqual([
            {
                type: "function_call_output",
                call_id: "call-1",
                output: vitest_1.expect.stringContaining("Continue completing"),
            },
        ]);
    });
    (0, vitest_1.it)("does NOT auto-continue follow-up questions without a captcha context", async () => {
        const client = createClient();
        const executeStepSpy = vitest_1.vi.spyOn(client, "executeStep");
        executeStepSpy.mockResolvedValueOnce({
            actions: [],
            message: "I've located the Submit button. Should I go ahead and submit it?",
            completed: true,
            nextInputItems: [],
            responseId: "response-1",
            usage: { input_tokens: 1, output_tokens: 1, inference_time_ms: 1 },
        });
        const result = await client.execute({
            options: { instruction: "Submit the form.", maxSteps: 10 },
            logger: vitest_1.vi.fn(),
        });
        // Should NOT have continued — the model's follow-up is treated as completion
        (0, vitest_1.expect)(executeStepSpy).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(result.completed).toBe(true);
    });
});
