// @browserbasehq/stagehand/dist/cjs/tests/unit/page-snapshot.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const fs_1 = require("fs");
const page_js_1 = require("../../lib/v3/understudy/page.js");
const snapshotModule = __importStar(require("../../lib/v3/understudy/a11y/snapshot/index.js"));
const baseSnapshot = {
    combinedTree: "tree",
    combinedXpathMap: {},
    combinedUrlMap: {},
    perFrame: [],
};
(0, vitest_1.describe)("Page.snapshot", () => {
    (0, vitest_1.afterEach)(() => {
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("forwards the includeIframes flag to captureHybridSnapshot", async () => {
        vitest_1.vi.spyOn(fs_1.promises, "writeFile").mockResolvedValue();
        const captureSpy = vitest_1.vi
            .spyOn(snapshotModule, "captureHybridSnapshot")
            .mockResolvedValue(baseSnapshot);
        const fakePage = {};
        await page_js_1.Page.prototype.snapshot.call(fakePage, { includeIframes: false });
        (0, vitest_1.expect)(captureSpy).toHaveBeenCalledWith(fakePage, {
            pierceShadow: true,
            includeIframes: false,
        });
    });
    (0, vitest_1.it)("falls back to default iframe inclusion when option is omitted", async () => {
        vitest_1.vi.spyOn(fs_1.promises, "writeFile").mockResolvedValue();
        const captureSpy = vitest_1.vi
            .spyOn(snapshotModule, "captureHybridSnapshot")
            .mockResolvedValue(baseSnapshot);
        const fakePage = {};
        await page_js_1.Page.prototype.snapshot.call(fakePage);
        (0, vitest_1.expect)(captureSpy).toHaveBeenCalledWith(fakePage, {
            pierceShadow: true,
            includeIframes: undefined,
        });
    });
});
