// @browserbasehq/stagehand/dist/cjs/tests/unit/safety-confirmation.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const OpenAICUAClient_js_1 = require("../../lib/v3/agent/OpenAICUAClient.js");
const GoogleCUAClient_js_1 = require("../../lib/v3/agent/GoogleCUAClient.js");
const openAISafetyInvoker = OpenAICUAClient_js_1.OpenAICUAClient.prototype.handleSafetyConfirmation;
const googleSafetyInvoker = GoogleCUAClient_js_1.GoogleCUAClient.prototype.handleSafetyConfirmation;
function createOpenAIClient() {
    return new OpenAICUAClient_js_1.OpenAICUAClient("openai", "openai/computer-use-preview", "test instructions", { apiKey: "test" });
}
function createGoogleClient() {
    return new GoogleCUAClient_js_1.GoogleCUAClient("google", "google/gemini-2.5-computer-use-preview-10-2025", "test instructions", { apiKey: "test" });
}
(0, vitest_1.describe)("Safety Confirmation Handler", () => {
    (0, vitest_1.describe)("OpenAI-style (pending_safety_checks)", () => {
        const mockChecks = [
            {
                id: "check-1",
                code: "malicious_instructions",
                message: "Potentially harmful action detected",
            },
        ];
        (0, vitest_1.it)("returns checks when handler acknowledges", async () => {
            const client = createOpenAIClient();
            const handler = vitest_1.vi.fn(async () => ({
                acknowledged: true,
            }));
            client.setSafetyConfirmationHandler(handler);
            const logger = vitest_1.vi.fn();
            const result = await openAISafetyInvoker.call(client, mockChecks, logger);
            (0, vitest_1.expect)(handler).toHaveBeenCalledWith(mockChecks);
            (0, vitest_1.expect)(result).toEqual(mockChecks);
        });
        (0, vitest_1.it)("returns undefined when handler rejects", async () => {
            const client = createOpenAIClient();
            const handler = vitest_1.vi.fn(async () => ({
                acknowledged: false,
            }));
            client.setSafetyConfirmationHandler(handler);
            const logger = vitest_1.vi.fn();
            const result = await openAISafetyInvoker.call(client, mockChecks, logger);
            (0, vitest_1.expect)(handler).toHaveBeenCalledWith(mockChecks);
            (0, vitest_1.expect)(result).toBeUndefined();
        });
        (0, vitest_1.it)("auto-acknowledges when no handler is set", async () => {
            const client = createOpenAIClient();
            const logger = vitest_1.vi.fn();
            const result = await openAISafetyInvoker.call(client, mockChecks, logger);
            (0, vitest_1.expect)(result).toEqual(mockChecks);
        });
    });
    (0, vitest_1.describe)("Google-style (safety_decision)", () => {
        const mockDecision = {
            decision: "require_confirmation",
            explanation: "Cookie consent dialog detected",
        };
        (0, vitest_1.it)("returns 'true' when handler acknowledges", async () => {
            const client = createGoogleClient();
            const handler = vitest_1.vi.fn(async () => ({
                acknowledged: true,
            }));
            client.setSafetyConfirmationHandler(handler);
            const logger = vitest_1.vi.fn();
            const result = await googleSafetyInvoker.call(client, mockDecision, logger);
            (0, vitest_1.expect)(handler).toHaveBeenCalledWith([
                {
                    id: "google-safety-decision",
                    code: "safety_decision",
                    message: JSON.stringify(mockDecision, null, 2),
                },
            ]);
            (0, vitest_1.expect)(result).toBe("true");
        });
        (0, vitest_1.it)("returns undefined when handler rejects", async () => {
            const client = createGoogleClient();
            const handler = vitest_1.vi.fn(async () => ({
                acknowledged: false,
            }));
            client.setSafetyConfirmationHandler(handler);
            const logger = vitest_1.vi.fn();
            const result = await googleSafetyInvoker.call(client, mockDecision, logger);
            (0, vitest_1.expect)(handler).toHaveBeenCalled();
            (0, vitest_1.expect)(result).toBeUndefined();
        });
        (0, vitest_1.it)("auto-acknowledges when no handler is set", async () => {
            const client = createGoogleClient();
            const logger = vitest_1.vi.fn();
            const result = await googleSafetyInvoker.call(client, mockDecision, logger);
            (0, vitest_1.expect)(result).toBe("true");
        });
        (0, vitest_1.it)("handles string safety decisions", async () => {
            const client = createGoogleClient();
            const handler = vitest_1.vi.fn(async () => ({
                acknowledged: true,
            }));
            client.setSafetyConfirmationHandler(handler);
            const logger = vitest_1.vi.fn();
            const result = await googleSafetyInvoker.call(client, "Simple string decision", logger);
            (0, vitest_1.expect)(handler).toHaveBeenCalledWith([
                {
                    id: "google-safety-decision",
                    code: "safety_decision",
                    message: "Simple string decision",
                },
            ]);
            (0, vitest_1.expect)(result).toBe("true");
        });
    });
});
