// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-capture-orchestration.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const capture = __importStar(require("../../lib/v3/understudy/a11y/snapshot/capture.js"));
const a11yTree = __importStar(require("../../lib/v3/understudy/a11y/snapshot/a11yTree.js"));
const domTree = __importStar(require("../../lib/v3/understudy/a11y/snapshot/domTree.js"));
const focusSelectors = __importStar(require("../../lib/v3/understudy/a11y/snapshot/focusSelectors.js"));
const mockCDPSession_js_1 = require("./helpers/mockCDPSession.js");
const makeProtocolFrame = (id) => ({
    id,
    loaderId: `${id}-loader`,
    url: "https://example.com",
    securityOrigin: "https://example.com",
    mimeType: "text/html",
});
const makeFrameTree = (id, children = []) => ({
    frame: makeProtocolFrame(id),
    childFrames: children,
});
const makePage = (overrides = {}) => {
    const defaultSession = new mockCDPSession_js_1.MockCDPSession({}, "default-session");
    const base = {
        mainFrameId: () => "frame-1",
        asProtocolFrameTree: () => makeFrameTree("frame-1"),
        listAllFrameIds: () => ["frame-1"],
        getSessionForFrame: () => defaultSession,
        getOrdinal: () => 0,
    };
    return { ...base, ...overrides };
};
const makeSessionIndex = () => ({
    rootBackend: 100,
    absByBe: new Map([
        [100, "/"],
        [101, "/html[1]"],
        [102, "/html[1]/body[1]"],
        [150, "/html[1]/body[1]/iframe[1]"],
        [200, "/html[1]/body[1]/iframe[1]"],
        [201, "/html[1]/body[1]/iframe[1]/div[1]"],
    ]),
    tagByBe: new Map([
        [100, "#document"],
        [101, "html"],
        [102, "body"],
        [150, "iframe"],
        [200, "#document"],
        [201, "div"],
    ]),
    scrollByBe: new Map([[201, true]]),
    docRootOf: new Map([
        [100, 100],
        [101, 100],
        [102, 100],
        [150, 100],
        [200, 200],
        [201, 200],
    ]),
    contentDocRootByIframe: new Map([[150, 200]]),
});
(0, vitest_1.beforeEach)(() => {
    vitest_1.vi.restoreAllMocks();
});
(0, vitest_1.describe)("buildFrameContext", () => {
    (0, vitest_1.it)("indexes parent relationships from the frame tree", () => {
        const frameTree = makeFrameTree("frame-1", [
            makeFrameTree("frame-2", [makeFrameTree("frame-3")]),
            makeFrameTree("frame-4"),
        ]);
        const page = makePage({
            asProtocolFrameTree: () => frameTree,
            listAllFrameIds: () => ["frame-1", "frame-2", "frame-3", "frame-4"],
        });
        const context = capture.buildFrameContext(page);
        (0, vitest_1.expect)(context.rootId).toBe("frame-1");
        (0, vitest_1.expect)(context.frames).toEqual([
            "frame-1",
            "frame-2",
            "frame-3",
            "frame-4",
        ]);
        (0, vitest_1.expect)(context.parentByFrame.get("frame-1")).toBeNull();
        (0, vitest_1.expect)(context.parentByFrame.get("frame-2")).toBe("frame-1");
        (0, vitest_1.expect)(context.parentByFrame.get("frame-3")).toBe("frame-2");
        (0, vitest_1.expect)(context.parentByFrame.get("frame-4")).toBe("frame-1");
    });
});
(0, vitest_1.describe)("buildSessionIndexes", () => {
    (0, vitest_1.it)("deduplicates frames that share the same CDP session id", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({}, "session-a");
        const page = makePage({
            // Every frame lookup returns the same session instance, so buildSessionIndexes
            // should call buildSessionDomIndex only once and reuse the result.
            getSessionForFrame: () => session,
        });
        const idx = makeSessionIndex();
        const spy = vitest_1.vi
            .spyOn(domTree, "buildSessionDomIndex")
            .mockResolvedValue(idx);
        const result = await capture.buildSessionIndexes(page, ["frame-1", "frame-2"], true);
        (0, vitest_1.expect)(spy).toHaveBeenCalledTimes(1); // only one DOM.getDocument per session id
        (0, vitest_1.expect)(spy).toHaveBeenCalledWith(session, true);
        (0, vitest_1.expect)(result.get("session-a")).toBe(idx);
    });
    (0, vitest_1.it)("builds indexes for sessions without ids using the 'root' key", async () => {
        const sessionWithoutId = {
            id: undefined,
            async send(_method, _params) {
                void _method;
                void _params;
                return {};
            },
            on() { },
            off() { },
            async close() { },
        };
        const sessionWithId = new mockCDPSession_js_1.MockCDPSession({}, "child-session");
        const page = makePage({
            getSessionForFrame: (frameId) => frameId === "frame-1" ? sessionWithoutId : sessionWithId,
        });
        const idxA = makeSessionIndex();
        const idxB = makeSessionIndex();
        const spy = vitest_1.vi
            .spyOn(domTree, "buildSessionDomIndex")
            .mockResolvedValueOnce(idxA)
            .mockResolvedValueOnce(idxB);
        const result = await capture.buildSessionIndexes(page, ["frame-1", "frame-2"], false);
        // Verifies the helper invokes buildSessionDomIndex once for each unique session,
        // keying anonymous sessions as "root" so downstream lookups remain stable.
        (0, vitest_1.expect)(spy).toHaveBeenNthCalledWith(1, sessionWithoutId, false);
        (0, vitest_1.expect)(spy).toHaveBeenNthCalledWith(2, sessionWithId, false);
        (0, vitest_1.expect)(result.get("root")).toBe(idxA);
        (0, vitest_1.expect)(result.get("child-session")).toBe(idxB);
    });
});
(0, vitest_1.describe)("collectPerFrameMaps", () => {
    (0, vitest_1.it)("builds per-frame xpath/tag maps and outlines from a shared session index", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 150 }),
        }, "session-a");
        const page = makePage({
            getSessionForFrame: () => session,
            getOrdinal: (frameId) => (frameId === "frame-1" ? 0 : 1),
        });
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const sessionIndex = makeSessionIndex();
        const sessionToIndex = new Map([[session.id, sessionIndex]]);
        vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockImplementation(async (_sess, frameId) => ({
            outline: `outline-${frameId}`,
            urlMap: { [`url-${frameId}`]: `https://${frameId}.test` },
            scopeApplied: false,
        }));
        const result = await capture.collectPerFrameMaps(page, context, sessionToIndex, { experimental: true }, true, context.frames);
        (0, vitest_1.expect)(result.perFrameOutlines).toEqual([
            { frameId: "frame-1", outline: "outline-frame-1" },
            { frameId: "frame-2", outline: "outline-frame-2" },
        ]);
        const rootMaps = result.perFrameMaps.get("frame-1");
        (0, vitest_1.expect)(rootMaps?.xpathMap["0-100"]).toBe("/");
        (0, vitest_1.expect)(rootMaps?.xpathMap["0-101"]).toBe("/html[1]");
        (0, vitest_1.expect)(rootMaps?.xpathMap["0-102"]).toBe("/html[1]/body[1]");
        const childMaps = result.perFrameMaps.get("frame-2");
        (0, vitest_1.expect)(childMaps?.xpathMap["1-200"]).toBe("/");
        (0, vitest_1.expect)(childMaps?.xpathMap["1-201"]).toBe("/div[1]");
        (0, vitest_1.expect)(childMaps?.scrollableMap["1-201"]).toBe(true);
        (0, vitest_1.expect)(childMaps?.urlMap).toEqual({
            "url-frame-2": "https://frame-2.test",
        });
        (0, vitest_1.expect)(session.callsFor("DOM.getFrameOwner")).toHaveLength(1);
    });
    (0, vitest_1.it)("builds a missing session index on demand and memoizes it", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({}, "new-session");
        const page = makePage({
            getSessionForFrame: () => session,
            getOrdinal: () => 2,
        });
        const context = {
            rootId: "frame-9",
            frames: ["frame-9"],
            parentByFrame: new Map([["frame-9", null]]),
        };
        const idx = makeSessionIndex();
        const buildSpy = vitest_1.vi
            .spyOn(domTree, "buildSessionDomIndex")
            .mockResolvedValue(idx);
        vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "outline",
            urlMap: {},
            scopeApplied: false,
        });
        const sessionToIndex = new Map();
        const result = await capture.collectPerFrameMaps(page, context, sessionToIndex, undefined, false, context.frames);
        (0, vitest_1.expect)(buildSpy).toHaveBeenCalledWith(session, false);
        (0, vitest_1.expect)(sessionToIndex.get("new-session")).toBe(idx);
        (0, vitest_1.expect)(result.perFrameMaps.get("frame-9")?.xpathMap["2-100"]).toBe("/");
    });
    (0, vitest_1.it)("skips frames that are not listed in the frameIds argument", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({}, "session-a");
        const page = makePage({
            getSessionForFrame: () => session,
            getOrdinal: (frameId) => (frameId === "frame-1" ? 0 : 1),
        });
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const sessionIndex = makeSessionIndex();
        const sessionToIndex = new Map([[session.id, sessionIndex]]);
        const a11ySpy = vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "outline",
            urlMap: {},
            scopeApplied: false,
        });
        const result = await capture.collectPerFrameMaps(page, context, sessionToIndex, undefined, true, ["frame-1"]);
        (0, vitest_1.expect)(a11ySpy).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(result.perFrameMaps.has("frame-2")).toBe(false);
        (0, vitest_1.expect)(result.perFrameOutlines.map((o) => o.frameId)).toEqual(["frame-1"]);
    });
});
(0, vitest_1.describe)("captureHybridSnapshot", () => {
    (0, vitest_1.it)("returns early when the scoped snapshot path succeeds", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({}, "session-a");
        const page = makePage({
            getSessionForFrame: () => session,
        });
        const options = { focusSelector: "/html" };
        vitest_1.vi.spyOn(focusSelectors, "resolveFocusFrameAndTail").mockResolvedValue({
            targetFrameId: "frame-1",
            tailXPath: "",
            absPrefix: "",
        });
        const domMapsSpy = vitest_1.vi
            .spyOn(domTree, "domMapsForSession")
            .mockResolvedValue({
            tagNameMap: { "0-100": "#document" },
            xpathMap: { "0-100": "/" },
            scrollableMap: {},
        });
        const a11ySpy = vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "scoped outline",
            urlMap: { "0-100": "https://frame-1.test" },
            scopeApplied: true,
        });
        const buildIndexSpy = vitest_1.vi
            .spyOn(domTree, "buildSessionDomIndex")
            .mockImplementation(() => {
            throw new Error("should not build session index when scoped");
        });
        const result = await capture.captureHybridSnapshot(page, options);
        (0, vitest_1.expect)(result.combinedTree).toBe("scoped outline");
        (0, vitest_1.expect)(result.combinedUrlMap["0-100"]).toBe("https://frame-1.test");
        (0, vitest_1.expect)(domMapsSpy).toHaveBeenCalled();
        (0, vitest_1.expect)(a11ySpy).toHaveBeenCalled();
        (0, vitest_1.expect)(buildIndexSpy).not.toHaveBeenCalled();
    });
    (0, vitest_1.it)("scoped snapshot still succeeds when iframe inclusion is disabled", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({}, "session-a");
        const page = makePage({
            getSessionForFrame: () => session,
        });
        const options = { focusSelector: "/html", includeIframes: false };
        vitest_1.vi.spyOn(focusSelectors, "resolveFocusFrameAndTail").mockResolvedValue({
            targetFrameId: "frame-1",
            tailXPath: "",
            absPrefix: "",
        });
        const domMapsSpy = vitest_1.vi
            .spyOn(domTree, "domMapsForSession")
            .mockResolvedValue({
            tagNameMap: { "0-100": "#document" },
            xpathMap: { "0-100": "/" },
            scrollableMap: {},
        });
        const a11ySpy = vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "scoped outline",
            urlMap: { "0-100": "https://frame-1.test" },
            scopeApplied: true,
        });
        const buildIndexSpy = vitest_1.vi
            .spyOn(domTree, "buildSessionDomIndex")
            .mockImplementation(() => {
            throw new Error("should not build session index when scoped");
        });
        const result = await capture.captureHybridSnapshot(page, options);
        (0, vitest_1.expect)(result.combinedTree).toBe("scoped outline");
        (0, vitest_1.expect)(result.combinedUrlMap["0-100"]).toBe("https://frame-1.test");
        (0, vitest_1.expect)(domMapsSpy).toHaveBeenCalled();
        (0, vitest_1.expect)(a11ySpy).toHaveBeenCalled();
        (0, vitest_1.expect)(buildIndexSpy).not.toHaveBeenCalled();
    });
    (0, vitest_1.it)("collects per-frame data and merges it when no scoped snapshot is available", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 150 }),
        }, "session-a");
        const page = makePage({
            asProtocolFrameTree: () => makeFrameTree("frame-1", [makeFrameTree("frame-2")]),
            listAllFrameIds: () => ["frame-1", "frame-2"],
            getSessionForFrame: () => session,
            getOrdinal: (frameId) => (frameId === "frame-1" ? 0 : 1),
        });
        const idx = makeSessionIndex();
        vitest_1.vi.spyOn(domTree, "buildSessionDomIndex").mockResolvedValue(idx);
        vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockImplementation(async (_sess, frameId) => ({
            outline: frameId === "frame-1"
                ? "[0-150] iframe host"
                : "[1-200] child subtree",
            urlMap: { [`url-${frameId}`]: `https://${frameId}.test` },
            scopeApplied: false,
        }));
        const snapshot = await capture.captureHybridSnapshot(page);
        (0, vitest_1.expect)(snapshot.combinedTree).toContain("[1-200] child subtree");
        (0, vitest_1.expect)(snapshot.combinedXpathMap["0-100"]).toBe("/");
        (0, vitest_1.expect)(snapshot.combinedXpathMap["1-201"]).toBe("/html[1]/body[1]/iframe[1]/div[1]");
        (0, vitest_1.expect)(snapshot.combinedUrlMap["url-frame-2"]).toBe("https://frame-2.test");
        (0, vitest_1.expect)(snapshot.perFrame?.map((pf) => pf.frameId)).toEqual([
            "frame-1",
            "frame-2",
        ]);
    });
    (0, vitest_1.it)("omits iframe frames when includeIframes is false", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 150 }),
        }, "session-a");
        const page = makePage({
            asProtocolFrameTree: () => makeFrameTree("frame-1", [makeFrameTree("frame-2")]),
            listAllFrameIds: () => ["frame-1", "frame-2"],
            getSessionForFrame: () => session,
            getOrdinal: (frameId) => (frameId === "frame-1" ? 0 : 1),
        });
        const idx = makeSessionIndex();
        vitest_1.vi.spyOn(domTree, "buildSessionDomIndex").mockResolvedValue(idx);
        const a11ySpy = vitest_1.vi
            .spyOn(a11yTree, "a11yForFrame")
            .mockImplementation(async (_sess, frameId) => ({
            outline: frameId === "frame-1"
                ? "[0-150] iframe host"
                : "[1-200] child subtree",
            urlMap: { [`url-${frameId}`]: `https://${frameId}.test` },
            scopeApplied: false,
        }));
        const snapshot = await capture.captureHybridSnapshot(page, {
            includeIframes: false,
        });
        (0, vitest_1.expect)(a11ySpy).toHaveBeenCalledTimes(1);
        (0, vitest_1.expect)(session.callsFor("DOM.getFrameOwner")).toHaveLength(0);
        (0, vitest_1.expect)(snapshot.perFrame?.map((pf) => pf.frameId)).toEqual(["frame-1"]);
        (0, vitest_1.expect)(snapshot.combinedXpathMap["1-201"]).toBeUndefined();
        (0, vitest_1.expect)(snapshot.combinedTree).not.toContain("[1-200] child subtree");
    });
});
