// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-frame-merge.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const mockCDPSession_js_1 = require("./helpers/mockCDPSession.js");
const capture_js_1 = require("../../lib/v3/understudy/a11y/snapshot/capture.js");
const makePage = (sessions) => ({
    getSessionForFrame: (frameId) => sessions[frameId] ?? sessions.root,
    getOrdinal: (frameId) => frameId === "frame-1" ? 0 : frameId === "frame-2" ? 1 : 2,
});
(0, vitest_1.describe)("computeFramePrefixes", () => {
    (0, vitest_1.it)("derives prefixes from parent iframe xpaths within the same session", async () => {
        const parentSession = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 200 }),
        });
        const page = makePage({
            "frame-1": parentSession,
            "frame-2": parentSession,
            root: parentSession,
        });
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: { "0-200": "/html[1]/body[1]/iframe[1]" },
                },
            ],
        ]);
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const { absPrefix, iframeHostEncByChild } = await (0, capture_js_1.computeFramePrefixes)(page, context, perFrameMaps, context.frames);
        (0, vitest_1.expect)(absPrefix.get("frame-1")).toBe("");
        (0, vitest_1.expect)(absPrefix.get("frame-2")).toBe("/html[1]/body[1]/iframe[1]");
        (0, vitest_1.expect)(iframeHostEncByChild.get("frame-2")).toBe("0-200");
    });
    (0, vitest_1.it)("inherits the parent prefix when frame owner lookups fail (OOPIF)", async () => {
        const parentSession = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async (params) => {
                if (params?.frameId === "frame-2")
                    return { backendNodeId: 200 };
                if (params?.frameId === "frame-3")
                    throw new Error("unavailable");
                return {};
            },
        });
        const page = makePage({
            "frame-1": parentSession,
            "frame-2": parentSession,
            "frame-3": parentSession,
            root: parentSession,
        });
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: { "0-200": "/iframe[1]" },
                },
            ],
            [
                "frame-2",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: { "1-300": "/div[1]/iframe[1]" },
                },
            ],
        ]);
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2", "frame-3"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
                ["frame-3", "frame-2"],
            ]),
        };
        const maps = await (0, capture_js_1.computeFramePrefixes)(page, context, perFrameMaps, context.frames);
        (0, vitest_1.expect)(maps.absPrefix.get("frame-2")).toBe("/iframe[1]");
        (0, vitest_1.expect)(maps.absPrefix.get("frame-3")).toBe("/iframe[1]");
    });
    (0, vitest_1.it)("inherits parent prefix when iframe xpath mapping is missing", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 999 }),
        });
        const page = makePage({
            "frame-1": session,
            "frame-2": session,
            root: session,
        });
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: {},
                },
            ],
        ]);
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const result = await (0, capture_js_1.computeFramePrefixes)(page, context, perFrameMaps, context.frames);
        (0, vitest_1.expect)(result.absPrefix.get("frame-2")).toBe("");
    });
    (0, vitest_1.it)("does not compute prefixes for frames excluded from the scope", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getFrameOwner": async () => ({ backendNodeId: 200 }),
        });
        const page = makePage({
            "frame-1": session,
            "frame-2": session,
            root: session,
        });
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: { "0-200": "/iframe[1]" },
                },
            ],
        ]);
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const { absPrefix, iframeHostEncByChild } = await (0, capture_js_1.computeFramePrefixes)(page, context, perFrameMaps, ["frame-1"]);
        (0, vitest_1.expect)(absPrefix.has("frame-2")).toBe(false);
        (0, vitest_1.expect)(iframeHostEncByChild.has("frame-2")).toBe(false);
    });
});
(0, vitest_1.describe)("mergeFramesIntoSnapshot", () => {
    (0, vitest_1.it)("merges root and child maps, prefixing child xpaths and injecting subtrees", () => {
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: { "0-10": "https://example.com" },
                    xpathMap: { "0-10": "/html[1]/body[1]" },
                },
            ],
            [
                "frame-2",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: { "1-20": "https://child.com" },
                    xpathMap: { "1-20": "/div[1]/span[1]" },
                },
            ],
        ]);
        const perFrameOutlines = [
            { frameId: "frame-1", outline: "[0-10] body\n  [0-200] iframe" },
            { frameId: "frame-2", outline: "[1-20] child" },
        ];
        const absPrefix = new Map([
            ["frame-1", ""],
            ["frame-2", "/html[1]/body[1]/iframe[1]"],
        ]);
        const iframeHostEncByChild = new Map([
            ["frame-2", "0-200"],
        ]);
        const snapshot = (0, capture_js_1.mergeFramesIntoSnapshot)(context, perFrameMaps, perFrameOutlines, absPrefix, iframeHostEncByChild, context.frames);
        (0, vitest_1.expect)(snapshot.combinedXpathMap["0-10"]).toBe("/html[1]/body[1]");
        (0, vitest_1.expect)(snapshot.combinedXpathMap["1-20"]).toBe("/html[1]/body[1]/iframe[1]/div[1]/span[1]");
        (0, vitest_1.expect)(snapshot.combinedUrlMap["1-20"]).toBe("https://child.com");
        (0, vitest_1.expect)(snapshot.combinedTree).toContain("[1-20] child");
    });
    (0, vitest_1.it)("skips frames without maps and handles missing iframe mappings", () => {
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: { "0-10": "/html[1]" },
                },
            ],
        ]);
        const perFrameOutlines = [
            { frameId: "frame-1", outline: "[0-10] html" },
            { frameId: "frame-2", outline: "[1-20] orphan" },
        ];
        const absPrefix = new Map([
            ["frame-1", ""],
            ["frame-2", "/missing"],
        ]);
        const snapshot = (0, capture_js_1.mergeFramesIntoSnapshot)(context, perFrameMaps, perFrameOutlines, absPrefix, new Map(), context.frames);
        (0, vitest_1.expect)(snapshot.combinedXpathMap["1-20"]).toBeUndefined();
        (0, vitest_1.expect)(snapshot.combinedTree).toBe("[0-10] html");
    });
    (0, vitest_1.it)("falls back to first outline when root frame outline is missing", () => {
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const perFrameMaps = new Map([
            [
                "frame-2",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: {},
                },
            ],
        ]);
        const perFrameOutlines = [
            { frameId: "frame-2", outline: "[child] frame2" },
        ];
        const snapshot = (0, capture_js_1.mergeFramesIntoSnapshot)(context, perFrameMaps, perFrameOutlines, new Map([["frame-2", "/iframe[1]"]]), new Map(), context.frames);
        (0, vitest_1.expect)(snapshot.combinedTree).toBe("[child] frame2");
    });
    (0, vitest_1.it)("overwrites duplicate iframe host entries when multiple children map to the same parent", () => {
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2", "frame-3"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
                ["frame-3", "frame-1"],
            ]),
        };
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: {},
                    xpathMap: {},
                },
            ],
        ]);
        const perFrameOutlines = [
            { frameId: "frame-1", outline: "[root] frame1\n  [0-200] iframe slot" },
            { frameId: "frame-2", outline: "[child] frame2" },
            { frameId: "frame-3", outline: "[child] frame3" },
        ];
        const snapshot = (0, capture_js_1.mergeFramesIntoSnapshot)(context, perFrameMaps, perFrameOutlines, new Map([
            ["frame-1", ""],
            ["frame-2", ""],
            ["frame-3", ""],
        ]), new Map([
            ["frame-2", "0-200"],
            ["frame-3", "0-200"],
        ]), context.frames);
        (0, vitest_1.expect)(snapshot.combinedTree).toContain("[child] frame3");
        (0, vitest_1.expect)(snapshot.combinedTree).not.toContain("[child] frame2");
    });
    (0, vitest_1.it)("only merges xpath and url maps for frames included in frameIds", () => {
        const context = {
            rootId: "frame-1",
            frames: ["frame-1", "frame-2"],
            parentByFrame: new Map([
                ["frame-1", null],
                ["frame-2", "frame-1"],
            ]),
        };
        const perFrameMaps = new Map([
            [
                "frame-1",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: { "0-10": "https://root.test" },
                    xpathMap: { "0-10": "/html[1]" },
                },
            ],
            [
                "frame-2",
                {
                    tagNameMap: {},
                    scrollableMap: {},
                    urlMap: { "1-20": "https://child.test" },
                    xpathMap: { "1-20": "/div[1]" },
                },
            ],
        ]);
        const perFrameOutlines = [{ frameId: "frame-1", outline: "[root] doc" }];
        const snapshot = (0, capture_js_1.mergeFramesIntoSnapshot)(context, perFrameMaps, perFrameOutlines, new Map([["frame-1", ""]]), new Map(), ["frame-1"]);
        (0, vitest_1.expect)(snapshot.combinedXpathMap["0-10"]).toBe("/html[1]");
        (0, vitest_1.expect)(snapshot.combinedXpathMap["1-20"]).toBeUndefined();
        (0, vitest_1.expect)(snapshot.perFrame?.map((pf) => pf.frameId)).toEqual(["frame-1"]);
    });
});
