// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-dom-session-builders.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const domTree_js_1 = require("../../lib/v3/understudy/a11y/snapshot/domTree.js");
const sdkErrors_js_1 = require("../../lib/v3/types/public/sdkErrors.js");
const mockCDPSession_js_1 = require("./helpers/mockCDPSession.js");
let nextNodeId = 1;
const makeDomNode = (overrides = {}) => {
    const nodeId = overrides.nodeId ?? nextNodeId++;
    const backendNodeId = overrides.backendNodeId ?? nextNodeId++;
    const nodeName = overrides.nodeName ?? "DIV";
    const nodeType = overrides.nodeType ?? 1;
    const children = overrides.children ?? [];
    return {
        nodeId,
        backendNodeId,
        nodeName,
        nodeType,
        localName: overrides.localName ?? nodeName.toLowerCase(),
        nodeValue: overrides.nodeValue ?? "",
        childNodeCount: overrides.childNodeCount ?? children.length,
        children,
        shadowRoots: overrides.shadowRoots,
        contentDocument: overrides.contentDocument,
        isScrollable: overrides.isScrollable,
    };
};
const buildSampleDomTree = () => {
    const iframeChild = makeDomNode({ nodeName: "P" });
    const iframeBody = makeDomNode({
        nodeName: "BODY",
        children: [iframeChild],
        isScrollable: true,
    });
    const iframeHtml = makeDomNode({ nodeName: "HTML", children: [iframeBody] });
    const iframeDoc = makeDomNode({
        nodeName: "#document",
        nodeType: 9,
        children: [iframeHtml],
    });
    const iframeElement = makeDomNode({
        nodeName: "IFRAME",
        contentDocument: iframeDoc,
    });
    const scrollDiv = makeDomNode({
        nodeName: "DIV",
        isScrollable: true,
    });
    const body = makeDomNode({
        nodeName: "BODY",
        children: [scrollDiv, iframeElement],
    });
    const html = makeDomNode({ nodeName: "HTML", children: [body] });
    const root = makeDomNode({
        nodeName: "#document",
        nodeType: 9,
        children: [html],
    });
    return {
        root,
        html,
        body,
        scrollDiv,
        iframeElement,
        iframeDoc,
        iframeHtml,
        iframeBody,
        iframeChild,
    };
};
(0, vitest_1.describe)("hydrateDomTree", () => {
    (0, vitest_1.it)("expands truncated nodes by calling DOM.describeNode", async () => {
        const child = makeDomNode({ nodeName: "DIV" });
        const root = makeDomNode({
            nodeName: "HTML",
            childNodeCount: 1,
            children: [],
        });
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.describeNode": async () => ({
                node: {
                    ...root,
                    children: [child],
                    childNodeCount: 1,
                },
            }),
        });
        await (0, domTree_js_1.hydrateDomTree)(session, root, true);
        (0, vitest_1.expect)(root.children).toEqual([child]);
    });
    (0, vitest_1.it)("retries describeNode when CBOR errors occur before succeeding", async () => {
        const child = makeDomNode({ nodeName: "DIV" });
        const root = makeDomNode({
            nodeName: "HTML",
            childNodeCount: 1,
            children: [],
        });
        let attempts = 0;
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.describeNode": async () => {
                attempts++;
                if (attempts === 1)
                    throw new Error("CBOR: stack limit exceeded");
                return { node: { ...root, children: [child], childNodeCount: 1 } };
            },
        });
        await (0, domTree_js_1.hydrateDomTree)(session, root, true);
        (0, vitest_1.expect)(attempts).toBe(2);
        (0, vitest_1.expect)(root.children).toEqual([child]);
    });
    (0, vitest_1.it)("throws StagehandDomProcessError after exhausting describeNode retries", async () => {
        const root = makeDomNode({
            nodeName: "HTML",
            childNodeCount: 1,
            children: [],
        });
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.describeNode": async () => {
                throw new Error("CBOR: stack limit exceeded");
            },
        });
        await (0, vitest_1.expect)((0, domTree_js_1.hydrateDomTree)(session, root, true)).rejects.toBeInstanceOf(sdkErrors_js_1.StagehandDomProcessError);
    });
});
(0, vitest_1.describe)("getDomTreeWithFallback", () => {
    (0, vitest_1.it)("retries DOM.getDocument after CBOR errors and returns the hydrated root", async () => {
        const root = makeDomNode({
            nodeName: "#document",
            nodeType: 9,
            children: [],
        });
        const depths = [];
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getDocument": async (params) => {
                const depth = (params?.depth ?? 0);
                depths.push(depth);
                if (depth === -1)
                    throw new Error("CBOR: stack limit exceeded");
                return { root };
            },
            "DOM.describeNode": async () => ({ node: root }),
        });
        const result = await (0, domTree_js_1.getDomTreeWithFallback)(session, true);
        (0, vitest_1.expect)(result).toBe(root);
        (0, vitest_1.expect)(depths).toEqual([-1, 256]);
    });
    (0, vitest_1.it)("propagates non-CBOR DOM.getDocument errors", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getDocument": async () => {
                throw new Error("network fail");
            },
        });
        await (0, vitest_1.expect)((0, domTree_js_1.getDomTreeWithFallback)(session, false)).rejects.toThrow("network fail");
    });
    (0, vitest_1.it)("throws StagehandDomProcessError when all depth attempts hit CBOR limits", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.getDocument": async () => {
                throw new Error("CBOR: stack limit exceeded");
            },
        });
        await (0, vitest_1.expect)((0, domTree_js_1.getDomTreeWithFallback)(session, false)).rejects.toBeInstanceOf(sdkErrors_js_1.StagehandDomProcessError);
    });
});
(0, vitest_1.describe)("buildSessionDomIndex", () => {
    (0, vitest_1.it)("collects absolute paths, scrollability, and content-document metadata", async () => {
        const tree = buildSampleDomTree();
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.enable": async () => ({}),
            "DOM.getDocument": async () => ({ root: tree.root }),
            "DOM.describeNode": async () => ({ node: tree.root }),
        });
        const index = await (0, domTree_js_1.buildSessionDomIndex)(session, true);
        (0, vitest_1.expect)(index.rootBackend).toBe(tree.root.backendNodeId);
        (0, vitest_1.expect)(index.absByBe.get(tree.body.backendNodeId)).toBe("/html[1]/body[1]");
        (0, vitest_1.expect)(index.absByBe.get(tree.scrollDiv.backendNodeId)).toBe("/html[1]/body[1]/div[1]");
        (0, vitest_1.expect)(index.scrollByBe.get(tree.scrollDiv.backendNodeId)).toBe(true);
        (0, vitest_1.expect)(index.docRootOf.get(tree.iframeHtml.backendNodeId)).toBe(tree.iframeDoc.backendNodeId);
        (0, vitest_1.expect)(index.contentDocRootByIframe.get(tree.iframeElement.backendNodeId)).toBe(tree.iframeDoc.backendNodeId);
    });
});
(0, vitest_1.describe)("domMapsForSession", () => {
    (0, vitest_1.it)("derives frame-relative xpath/tag/scrollable maps for a frame's document root", async () => {
        const tree = buildSampleDomTree();
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.enable": async () => ({}),
            "DOM.getDocument": async () => ({ root: tree.root }),
            "DOM.getFrameOwner": async () => ({
                backendNodeId: tree.iframeElement.backendNodeId,
            }),
            "DOM.describeNode": async () => ({ node: tree.root }),
        });
        const encode = (frameId, backendNodeId) => `${frameId}-${backendNodeId}`;
        const maps = await (0, domTree_js_1.domMapsForSession)(session, "frame-A", true, encode, true);
        const iframeDocKey = `frame-A-${tree.iframeDoc.backendNodeId}`;
        const iframeBodyKey = `frame-A-${tree.iframeBody.backendNodeId}`;
        const iframeChildKey = `frame-A-${tree.iframeChild.backendNodeId}`;
        (0, vitest_1.expect)(maps.tagNameMap[iframeDocKey]).toBe("#document");
        (0, vitest_1.expect)(maps.xpathMap[iframeDocKey]).toBe("/");
        (0, vitest_1.expect)(maps.xpathMap[iframeBodyKey]).toBe("/html[1]/body[1]");
        (0, vitest_1.expect)(maps.xpathMap[iframeChildKey]).toBe("/html[1]/body[1]/p[1]");
        (0, vitest_1.expect)(maps.scrollableMap[iframeBodyKey]).toBe(true);
        (0, vitest_1.expect)(Object.keys(maps.tagNameMap)).not.toContain(`frame-A-${tree.html.backendNodeId}`);
    });
    (0, vitest_1.it)("falls back to the root document when frame owner lookup fails", async () => {
        const tree = buildSampleDomTree();
        const session = new mockCDPSession_js_1.MockCDPSession({
            "DOM.enable": async () => ({}),
            "DOM.getDocument": async () => ({ root: tree.root }),
            "DOM.getFrameOwner": async () => {
                throw new Error("owner lookup failed");
            },
            "DOM.describeNode": async () => ({ node: tree.root }),
        });
        const encode = (frameId, backendNodeId) => `${frameId}-${backendNodeId}`;
        const maps = await (0, domTree_js_1.domMapsForSession)(session, "frame-B", false, encode, true);
        (0, vitest_1.expect)(maps.xpathMap[`frame-B-${tree.html.backendNodeId}`]).toBe("/html[1]");
        (0, vitest_1.expect)(maps.xpathMap[`frame-B-${tree.scrollDiv.backendNodeId}`]).toBe("/html[1]/body[1]/div[1]");
    });
});
