// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-a11y-resolvers.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const a11yTree_js_1 = require("../../lib/v3/understudy/a11y/snapshot/a11yTree.js");
const focusSelectors = __importStar(require("../../lib/v3/understudy/a11y/snapshot/focusSelectors.js"));
const mockCDPSession_js_1 = require("./helpers/mockCDPSession.js");
const executionContextRegistry_js_1 = require("../../lib/v3/understudy/executionContextRegistry.js");
const capture_js_1 = require("../../lib/v3/understudy/a11y/snapshot/capture.js");
const domTree = __importStar(require("../../lib/v3/understudy/a11y/snapshot/domTree.js"));
const a11yTree = __importStar(require("../../lib/v3/understudy/a11y/snapshot/a11yTree.js"));
const logger = __importStar(require("../../lib/v3/logger.js"));
const stringType = "string";
const baseAxNodes = () => [
    {
        nodeId: "1",
        role: { type: stringType, value: "RootWebArea" },
        backendDOMNodeId: 100,
        childIds: ["2"],
        ignored: false,
    },
    {
        nodeId: "2",
        role: { type: stringType, value: "link" },
        name: { type: stringType, value: "Docs" },
        backendDOMNodeId: 101,
        parentId: "1",
        childIds: [],
        properties: [
            {
                name: "url",
                value: { type: stringType, value: "https://example.com" },
            },
        ],
        ignored: false,
    },
];
const baseHandlers = {
    "Accessibility.enable": async () => ({}),
    "Runtime.enable": async () => ({}),
    "DOM.enable": async () => ({}),
};
(0, vitest_1.describe)("a11yForFrame", () => {
    (0, vitest_1.beforeEach)(() => {
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("returns full outline and url map when no focus selector is provided", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "Accessibility.getFullAXTree": async () => ({ nodes: baseAxNodes() }),
        });
        const opts = {
            focusSelector: undefined,
            experimental: false,
            tagNameMap: { "enc-100": "#document", "enc-101": "a" },
            scrollableMap: {},
            encode: (backend) => `enc-${backend}`,
        };
        const result = await (0, a11yTree_js_1.a11yForFrame)(session, undefined, opts);
        (0, vitest_1.expect)(result.scopeApplied).toBe(false);
        (0, vitest_1.expect)(result.urlMap["enc-101"]).toBe("https://example.com");
        (0, vitest_1.expect)(result.outline).toContain("Docs");
    });
    (0, vitest_1.it)("scopes the tree to the resolved focus selector target", async () => {
        const nodes = baseAxNodes().map((n) => n.nodeId === "2"
            ? {
                ...n,
                childIds: ["3"],
            }
            : n);
        nodes.push({
            nodeId: "3",
            parentId: "2",
            childIds: [],
            role: { type: stringType, value: "StaticText" },
            backendDOMNodeId: 102,
            ignored: false,
        });
        let scopedOnce = false;
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "Accessibility.getFullAXTree": async (params) => {
                if (params?.frameId && !scopedOnce) {
                    scopedOnce = true;
                    throw new Error("does not belong to the target");
                }
                return { nodes };
            },
            "DOM.describeNode": async () => ({
                node: { backendNodeId: 101 },
            }),
        });
        const resolveSpy = vitest_1.vi
            .spyOn(focusSelectors, "resolveObjectIdForXPath")
            .mockResolvedValue("object-1");
        const opts = {
            focusSelector: "xpath=//a",
            experimental: false,
            tagNameMap: { "enc-101": "a" },
            scrollableMap: {},
            encode: (backend) => `enc-${backend}`,
        };
        const result = await (0, a11yTree_js_1.a11yForFrame)(session, "frame-1", opts);
        (0, vitest_1.expect)(result.scopeApplied).toBe(true);
        (0, vitest_1.expect)(result.outline).not.toContain("RootWebArea");
        (0, vitest_1.expect)(resolveSpy).toHaveBeenCalled();
        resolveSpy.mockRestore();
    });
    (0, vitest_1.it)("falls back to full tree when resolveObjectId throws", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "Accessibility.getFullAXTree": async () => ({ nodes: baseAxNodes() }),
        });
        vitest_1.vi.spyOn(focusSelectors, "resolveObjectIdForCss").mockRejectedValue(new Error("fail"));
        const opts = {
            focusSelector: ".btn",
            experimental: false,
            tagNameMap: {},
            scrollableMap: {},
            encode: (backend) => `enc-${backend}`,
        };
        const result = await (0, a11yTree_js_1.a11yForFrame)(session, "frame-1", opts);
        (0, vitest_1.expect)(result.scopeApplied).toBe(false);
    });
});
(0, vitest_1.describe)("resolveObjectIdForXPath", () => {
    (0, vitest_1.beforeEach)(() => {
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("evaluates in the target frame's main world when available", async () => {
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "waitForMainWorld").mockResolvedValue(42);
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "getMainWorld").mockReturnValue(undefined);
        const session = new mockCDPSession_js_1.MockCDPSession({
            "Runtime.evaluate": async (params) => {
                (0, vitest_1.expect)(params?.contextId).toBe(42);
                return { result: { objectId: "node-obj" } };
            },
        });
        const objectId = await focusSelectors.resolveObjectIdForXPath(session, "//div", "frame-1");
        (0, vitest_1.expect)(objectId).toBe("node-obj");
    });
    (0, vitest_1.it)("returns null when evaluation throws or reports exception details", async () => {
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "waitForMainWorld").mockRejectedValue(new Error("missing"));
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "getMainWorld").mockReturnValue(undefined);
        const session = new mockCDPSession_js_1.MockCDPSession({
            "Runtime.evaluate": async () => ({
                result: {},
                exceptionDetails: { exception: { description: "bad" } },
            }),
        });
        const objectId = await focusSelectors.resolveObjectIdForXPath(session, "//div", "frame-2");
        (0, vitest_1.expect)(objectId).toBeNull();
    });
});
(0, vitest_1.describe)("resolveObjectIdForCss", () => {
    (0, vitest_1.beforeEach)(() => {
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("returns primary evaluation result when available", async () => {
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "waitForMainWorld").mockResolvedValue(7);
        const session = new mockCDPSession_js_1.MockCDPSession({
            "Runtime.evaluate": async () => ({
                result: { objectId: "primary-obj" },
            }),
        });
        const objectId = await focusSelectors.resolveObjectIdForCss(session, ".btn", "frame-1");
        (0, vitest_1.expect)(objectId).toBe("primary-obj");
    });
    (0, vitest_1.it)("falls back to the pierce selector when the primary lookup fails", async () => {
        let call = 0;
        const session = new mockCDPSession_js_1.MockCDPSession({
            "Runtime.evaluate": async (params) => {
                call++;
                if (call === 1) {
                    (0, vitest_1.expect)(String(params?.expression)).toContain("resolveCssSelector");
                    return { result: {} };
                }
                (0, vitest_1.expect)(String(params?.expression)).toContain("resolveCssSelectorPierce");
                return { result: { objectId: "css-obj" } };
            },
        });
        const objectId = await focusSelectors.resolveObjectIdForCss(session, ".btn", undefined);
        (0, vitest_1.expect)(objectId).toBe("css-obj");
    });
    (0, vitest_1.it)("returns null when both primary and fallback evaluations throw", async () => {
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "waitForMainWorld").mockResolvedValue(11);
        vitest_1.vi.spyOn(executionContextRegistry_js_1.executionContexts, "getMainWorld").mockReturnValue(undefined);
        const session = new mockCDPSession_js_1.MockCDPSession({
            "Runtime.evaluate": async () => ({
                result: {},
                exceptionDetails: { exception: { description: "fail" } },
            }),
        });
        const objectId = await focusSelectors.resolveObjectIdForCss(session, ".missing", "frame-1");
        (0, vitest_1.expect)(objectId).toBeNull();
    });
});
(0, vitest_1.describe)("tryScopedSnapshot", () => {
    const ordinal = (frameId) => (frameId === "frame-1" ? 0 : 1);
    const context = {
        rootId: "frame-1",
        frames: ["frame-1", "frame-2"],
        parentByFrame: new Map([
            ["frame-1", null],
            ["frame-2", "frame-1"],
        ]),
    };
    const makePage = (session, overrides) => ({
        mainFrameId: () => "frame-1",
        asProtocolFrameTree: () => ({
            frame: { id: "frame-1" },
            childFrames: [{ frame: { id: "frame-2" } }],
        }),
        listAllFrameIds: () => ["frame-1", "frame-2"],
        getSessionForFrame: () => session,
        getOrdinal: (fid) => ordinal(fid),
        ...overrides,
    });
    (0, vitest_1.beforeEach)(() => {
        vitest_1.vi.restoreAllMocks();
    });
    (0, vitest_1.it)("returns scoped snapshot when focus selector resolves via CSS hops", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({});
        const domMapsSpy = vitest_1.vi
            .spyOn(domTree, "domMapsForSession")
            .mockResolvedValue({
            tagNameMap: { "1-10": "div" },
            xpathMap: { "1-10": "/div[1]" },
            scrollableMap: {},
        });
        const a11ySpy = vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "[1-10] div",
            urlMap: { "1-10": "https://example.com" },
            scopeApplied: true,
        });
        vitest_1.vi.spyOn(focusSelectors, "resolveCssFocusFrameAndTail").mockResolvedValue({
            targetFrameId: "frame-2",
            tailSelector: ".btn-inner",
            absPrefix: "/html/body/iframe[1]",
        });
        const result = await (0, capture_js_1.tryScopedSnapshot)(makePage(session), { focusSelector: ".btn" }, context, true);
        (0, vitest_1.expect)(result).not.toBeNull();
        (0, vitest_1.expect)(result?.combinedXpathMap["1-10"]).toBe("/html/body/iframe[1]/div[1]");
        (0, vitest_1.expect)(domMapsSpy).toHaveBeenCalled();
        (0, vitest_1.expect)(a11ySpy).toHaveBeenCalled();
    });
    (0, vitest_1.it)("returns null and logs fallback when scope is not applied", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({});
        vitest_1.vi.spyOn(domTree, "domMapsForSession").mockResolvedValue({
            tagNameMap: { "1-10": "div" },
            xpathMap: { "1-10": "/div[1]" },
            scrollableMap: {},
        });
        vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "ignored",
            urlMap: {},
            scopeApplied: false,
        });
        const loggerSpy = vitest_1.vi.spyOn(logger, "v3Logger").mockImplementation(() => { });
        const result = await (0, capture_js_1.tryScopedSnapshot)(makePage(session), { focusSelector: ".btn" }, context, false);
        (0, vitest_1.expect)(result).toBeNull();
        (0, vitest_1.expect)(loggerSpy).toHaveBeenCalled();
    });
    (0, vitest_1.it)("returns null immediately when no focus selector is provided", async () => {
        const result = await (0, capture_js_1.tryScopedSnapshot)(makePage(new mockCDPSession_js_1.MockCDPSession({})), {}, context, true);
        (0, vitest_1.expect)(result).toBeNull();
    });
    (0, vitest_1.it)("supports XPath focus resolution branch", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({});
        vitest_1.vi.spyOn(domTree, "domMapsForSession").mockResolvedValue({
            tagNameMap: { "1-10": "div" },
            xpathMap: { "1-10": "/div[1]" },
            scrollableMap: {},
        });
        vitest_1.vi.spyOn(a11yTree, "a11yForFrame").mockResolvedValue({
            outline: "[1-10] div",
            urlMap: {},
            scopeApplied: true,
        });
        vitest_1.vi.spyOn(focusSelectors, "resolveFocusFrameAndTail").mockResolvedValue({
            targetFrameId: "frame-1",
            tailXPath: "//div[1]",
            absPrefix: "",
        });
        const result = await (0, capture_js_1.tryScopedSnapshot)(makePage(session), { focusSelector: "xpath=//div" }, context, true);
        (0, vitest_1.expect)(result).not.toBeNull();
        (0, vitest_1.expect)(result?.combinedXpathMap["1-10"]).toBe("/div[1]");
    });
    (0, vitest_1.it)("logs and returns null when resolver throws", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({});
        vitest_1.vi.spyOn(focusSelectors, "resolveCssFocusFrameAndTail").mockRejectedValue(new Error("bad selector"));
        const loggerSpy = vitest_1.vi.spyOn(logger, "v3Logger").mockImplementation(() => { });
        const result = await (0, capture_js_1.tryScopedSnapshot)(makePage(session), { focusSelector: ".bad" }, context, true);
        (0, vitest_1.expect)(result).toBeNull();
        (0, vitest_1.expect)(loggerSpy).toHaveBeenCalled();
    });
});
