// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-a11y-tree-utils.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const a11yTree_js_1 = require("../../lib/v3/understudy/a11y/snapshot/a11yTree.js");
const axString = (value) => ({
    type: "string",
    value,
});
const defaultOpts = {
    focusSelector: undefined,
    experimental: false,
    tagNameMap: {},
    scrollableMap: {},
    encode: (backendNodeId) => `enc-${backendNodeId}`,
};
const makeAxNode = (overrides = {}) => ({
    nodeId: overrides.nodeId ?? String(Math.random()),
    backendDOMNodeId: overrides.backendDOMNodeId ?? Math.floor(Math.random() * 1e6),
    role: overrides.role ?? axString("generic"),
    childIds: overrides.childIds ?? [],
    parentId: overrides.parentId,
    properties: overrides.properties ?? [],
    name: overrides.name,
    description: overrides.description,
    value: overrides.value,
    ignored: overrides.ignored ?? false,
});
(0, vitest_1.describe)("decorateRoles", () => {
    (0, vitest_1.it)("marks scrollable DOM nodes with tag labels and encoded ids", () => {
        const opts = {
            ...defaultOpts,
            tagNameMap: {
                "enc-1": "div",
                "enc-2": "html",
                "enc-3": "#document",
                "enc-4": "#svg",
            },
            scrollableMap: { "enc-1": true, "enc-4": true },
        };
        const nodes = [
            makeAxNode({
                backendDOMNodeId: 1,
                role: { type: "string", value: "region" },
            }),
            makeAxNode({
                backendDOMNodeId: 2,
                role: { type: "string", value: "generic" },
            }),
            makeAxNode({
                backendDOMNodeId: 3,
                role: { type: "string", value: "generic" },
            }),
            makeAxNode({
                backendDOMNodeId: 4,
                role: { type: "string", value: "generic" },
            }),
        ];
        const decorated = (0, a11yTree_js_1.decorateRoles)(nodes, opts);
        (0, vitest_1.expect)(decorated).toMatchObject([
            { encodedId: "enc-1", role: "scrollable, div" },
            { encodedId: "enc-2", role: "scrollable, html" },
            { encodedId: "enc-3", role: "generic" },
            { encodedId: "enc-4", role: "scrollable, svg" },
        ]);
    });
    (0, vitest_1.it)("falls back when encoding fails", () => {
        const opts = {
            ...defaultOpts,
            encode: () => {
                throw new Error("boom");
            },
        };
        const nodes = [makeAxNode({ backendDOMNodeId: 4 })];
        const decorated = (0, a11yTree_js_1.decorateRoles)(nodes, opts);
        (0, vitest_1.expect)(decorated[0]?.encodedId).toBeUndefined();
    });
});
(0, vitest_1.describe)("buildHierarchicalTree", () => {
    const opts = {
        ...defaultOpts,
        tagNameMap: { root: "div", child: "span" },
    };
    (0, vitest_1.it)("drops structural nodes without children or names", async () => {
        const nodes = [
            {
                role: "generic",
                name: "",
                nodeId: "root",
                encodedId: "root",
                parentId: undefined,
                childIds: ["child"],
            },
            {
                role: "generic",
                name: "",
                nodeId: "child",
                encodedId: "child",
                parentId: "root",
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, opts);
        (0, vitest_1.expect)(tree).toEqual([]);
    });
    (0, vitest_1.it)("promotes select/combobox tag names for structural nodes", async () => {
        const nodes = [
            {
                role: "combobox",
                name: "Select",
                nodeId: "root",
                encodedId: "root",
                parentId: undefined,
                childIds: ["child"],
            },
            {
                role: "StaticText",
                name: "Option",
                nodeId: "child",
                encodedId: "child",
                parentId: "root",
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, {
            ...opts,
            tagNameMap: { root: "select" },
        });
        (0, vitest_1.expect)(tree[0]?.role).toBe("select");
    });
    (0, vitest_1.it)("drops structural parents with a single cleaned child while keeping it in place", async () => {
        const nodes = [
            {
                role: "generic",
                name: "",
                nodeId: "root",
                encodedId: "root",
                parentId: undefined,
                childIds: ["child"],
            },
            {
                role: "StaticText",
                name: "Ok",
                nodeId: "child",
                encodedId: "child",
                parentId: "root",
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, opts);
        (0, vitest_1.expect)(tree[0]?.role).toBe("StaticText");
    });
    (0, vitest_1.it)("drops structural parents entirely when all descendants are pruned", async () => {
        const nodes = [
            {
                role: "generic",
                name: "",
                nodeId: "root",
                encodedId: "root",
                parentId: undefined,
                childIds: ["child"],
            },
            {
                role: "generic",
                name: "",
                nodeId: "child",
                encodedId: "child",
                parentId: "root",
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, opts);
        (0, vitest_1.expect)(tree).toEqual([]);
    });
    (0, vitest_1.it)("renames structural nodes to their tag names when not combobox", async () => {
        const nodes = [
            {
                role: "generic",
                name: "Container",
                nodeId: "root",
                encodedId: "root",
                parentId: undefined,
                childIds: ["child-a", "child-b"],
            },
            {
                role: "StaticText",
                name: "A",
                nodeId: "child-a",
                encodedId: "child-a",
                parentId: "root",
                childIds: [],
            },
            {
                role: "StaticText",
                name: "B",
                nodeId: "child-b",
                encodedId: "child-b",
                parentId: "root",
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, {
            ...opts,
            tagNameMap: { root: "section" },
        });
        (0, vitest_1.expect)(tree[0]?.role).toBe("section");
    });
    (0, vitest_1.it)("skips nodes with negative node ids early", async () => {
        const nodes = [
            {
                role: "button",
                name: "Hidden",
                nodeId: "-1",
                encodedId: "hidden",
                parentId: undefined,
                childIds: [],
            },
        ];
        const { tree } = await (0, a11yTree_js_1.buildHierarchicalTree)(nodes, opts);
        (0, vitest_1.expect)(tree).toEqual([]);
    });
});
(0, vitest_1.describe)("isStructural", () => {
    (0, vitest_1.it)("marks generic/none/InlineTextBox roles as structural", () => {
        (0, vitest_1.expect)((0, a11yTree_js_1.isStructural)("generic")).toBe(true);
        (0, vitest_1.expect)((0, a11yTree_js_1.isStructural)("none")).toBe(true);
        (0, vitest_1.expect)((0, a11yTree_js_1.isStructural)("InlineTextBox")).toBe(true);
        (0, vitest_1.expect)((0, a11yTree_js_1.isStructural)("button")).toBe(false);
    });
});
(0, vitest_1.describe)("removeRedundantStaticTextChildren", () => {
    (0, vitest_1.it)("removes static text children whose concatenated text equals the parent name", () => {
        const parent = {
            role: "button",
            name: "HelloWorld",
            nodeId: "root",
        };
        const children = [
            { role: "StaticText", name: "Hello", nodeId: "c1" },
            { role: "StaticText", name: "World", nodeId: "c2" },
            { role: "button", name: "Child", nodeId: "c3" },
        ];
        const pruned = (0, a11yTree_js_1.removeRedundantStaticTextChildren)(parent, children);
        (0, vitest_1.expect)(pruned).toEqual([{ role: "button", name: "Child", nodeId: "c3" }]);
    });
    (0, vitest_1.it)("keeps static text when combined text differs", () => {
        const parent = {
            role: "button",
            name: "Hello World",
            nodeId: "root",
        };
        const children = [
            { role: "StaticText", name: "Hello", nodeId: "c1" },
            { role: "StaticText", name: "Mars", nodeId: "c2" },
        ];
        (0, vitest_1.expect)((0, a11yTree_js_1.removeRedundantStaticTextChildren)(parent, children)).toEqual(children);
    });
    (0, vitest_1.it)("returns original children when parent name is empty", () => {
        const parent = {
            role: "button",
            nodeId: "root",
        };
        const children = [
            { role: "StaticText", name: "Hello", nodeId: "c1" },
            { role: "StaticText", name: "World", nodeId: "c2" },
        ];
        (0, vitest_1.expect)((0, a11yTree_js_1.removeRedundantStaticTextChildren)(parent, children)).toEqual(children);
    });
});
(0, vitest_1.describe)("extractUrlFromAXNode", () => {
    (0, vitest_1.it)("returns trimmed URL string from node properties", () => {
        const node = makeAxNode({
            properties: [
                { name: "busy", value: axString("bar") },
                { name: "url", value: axString(" https://example.com ") },
            ],
        });
        (0, vitest_1.expect)((0, a11yTree_js_1.extractUrlFromAXNode)(node)).toBe("https://example.com");
    });
    (0, vitest_1.it)("returns undefined when url property missing or invalid", () => {
        (0, vitest_1.expect)((0, a11yTree_js_1.extractUrlFromAXNode)(makeAxNode({ properties: [] }))).toBeUndefined();
        (0, vitest_1.expect)((0, a11yTree_js_1.extractUrlFromAXNode)(makeAxNode({
            properties: [{ name: "url", value: { type: "number", value: 123 } }],
        }))).toBeUndefined();
    });
});
