// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-dom-tree-utils.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const domTree_js_1 = require("../../lib/v3/understudy/a11y/snapshot/domTree.js");
let nextNodeId = 1;
const makeNode = (overrides = {}) => {
    const base = {
        nodeId: nextNodeId++,
        backendNodeId: nextNodeId++,
        nodeType: 1,
        nodeName: "DIV",
        localName: "div",
        nodeValue: "",
        childNodeCount: overrides.childNodeCount ??
            (overrides.children ? overrides.children.length : 0),
    };
    return { ...base, ...overrides };
};
(0, vitest_1.describe)("shouldExpandNode", () => {
    (0, vitest_1.it)("returns true when declared children exceed realized children", () => {
        const node = makeNode({
            childNodeCount: 2,
            children: [makeNode()],
        });
        (0, vitest_1.expect)((0, domTree_js_1.shouldExpandNode)(node)).toBe(true);
    });
    (0, vitest_1.it)("returns false when all declared children are realized", () => {
        const child = makeNode();
        const node = makeNode({
            childNodeCount: 1,
            children: [child],
        });
        (0, vitest_1.expect)((0, domTree_js_1.shouldExpandNode)(node)).toBe(false);
    });
});
(0, vitest_1.describe)("mergeDomNodes", () => {
    (0, vitest_1.it)("overrides structural fields with expanded node data", () => {
        const originalChildren = [makeNode({ nodeName: "SPAN" })];
        const target = makeNode({
            childNodeCount: 1,
            children: originalChildren,
            shadowRoots: [makeNode({ nodeName: "shadow-old" })],
            contentDocument: makeNode({ nodeName: "doc-old" }),
        });
        const source = makeNode({
            childNodeCount: 3,
            children: [makeNode({ nodeName: "DIV" })],
            shadowRoots: [],
            contentDocument: makeNode({ nodeName: "doc-new" }),
        });
        (0, domTree_js_1.mergeDomNodes)(target, source);
        (0, vitest_1.expect)(target.childNodeCount).toBe(3);
        (0, vitest_1.expect)(target.children).toEqual(source.children);
        (0, vitest_1.expect)(target.shadowRoots).toEqual([]);
        (0, vitest_1.expect)(target.contentDocument?.nodeName).toBe("doc-new");
    });
    (0, vitest_1.it)("preserves original structures when source omits them", () => {
        const child = makeNode();
        const target = makeNode({
            childNodeCount: 1,
            children: [child],
        });
        const source = makeNode({
            childNodeCount: 5,
        });
        (0, domTree_js_1.mergeDomNodes)(target, source);
        (0, vitest_1.expect)(target.childNodeCount).toBe(5);
        (0, vitest_1.expect)(target.children).toEqual([child]);
    });
});
(0, vitest_1.describe)("collectDomTraversalTargets", () => {
    (0, vitest_1.it)("returns children, shadow roots, and content document in order", () => {
        const childA = makeNode({ nodeName: "CHILD-A" });
        const childB = makeNode({ nodeName: "CHILD-B" });
        const shadow = makeNode({ nodeName: "SHADOW" });
        const content = makeNode({ nodeName: "CONTENT" });
        const node = makeNode({
            children: [childA, childB],
            shadowRoots: [shadow],
            contentDocument: content,
        });
        const targets = (0, domTree_js_1.collectDomTraversalTargets)(node);
        (0, vitest_1.expect)(targets).toEqual([childA, childB, shadow, content]);
    });
});
(0, vitest_1.describe)("findNodeByBackendId", () => {
    (0, vitest_1.it)("finds nodes nested within children and shadow roots", () => {
        const target = makeNode({ backendNodeId: 999, nodeName: "TARGET" });
        const root = makeNode({
            children: [
                makeNode({
                    children: [makeNode(), target],
                }),
            ],
            shadowRoots: [makeNode()],
        });
        (0, vitest_1.expect)((0, domTree_js_1.findNodeByBackendId)(root, 999)).toBe(target);
    });
    (0, vitest_1.it)("returns undefined when no node matches the backend id", () => {
        const root = makeNode({
            children: [makeNode()],
            shadowRoots: [makeNode()],
        });
        (0, vitest_1.expect)((0, domTree_js_1.findNodeByBackendId)(root, 123456)).toBeUndefined();
    });
});
