// @browserbasehq/stagehand/dist/cjs/tests/unit/snapshot-cbor.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const index_js_1 = require("../../lib/v3/understudy/a11y/snapshot/index.js");
const mockCDPSession_js_1 = require("./helpers/mockCDPSession.js");
const sdkErrors_js_1 = require("../../lib/v3/types/public/sdkErrors.js");
function createFakePage(session) {
    const frameTree = {
        frame: {
            id: "root",
            loaderId: "root-loader",
            url: "http://fake",
            domainAndRegistry: "fake",
            securityOrigin: "http://fake",
            mimeType: "text/html",
            secureContextType: "Secure",
            crossOriginIsolatedContextType: "NotIsolated",
            gatedAPIFeatures: [],
        },
        childFrames: [],
    };
    return {
        mainFrameId: () => "root",
        asProtocolFrameTree: () => frameTree,
        listAllFrameIds: () => ["root"],
        getSessionForFrame: () => session,
        getOrdinal: () => 0,
    };
}
function completeDomTree() {
    return {
        nodeId: 1,
        backendNodeId: 1,
        nodeType: 9,
        nodeName: "#document",
        childNodeCount: 1,
        children: [
            {
                nodeId: 2,
                backendNodeId: 2,
                nodeType: 1,
                nodeName: "HTML",
                childNodeCount: 1,
                children: [
                    {
                        nodeId: 3,
                        backendNodeId: 3,
                        nodeType: 1,
                        nodeName: "BODY",
                        childNodeCount: 1,
                        children: [
                            {
                                nodeId: 4,
                                backendNodeId: 4,
                                nodeType: 1,
                                nodeName: "DIV",
                                childNodeCount: 0,
                                children: [],
                            },
                        ],
                    },
                ],
            },
        ],
    };
}
function truncatedDomTree() {
    return {
        nodeId: 1,
        backendNodeId: 1,
        nodeType: 9,
        nodeName: "#document",
        childNodeCount: 1,
        children: [
            {
                nodeId: 2,
                backendNodeId: 2,
                nodeType: 1,
                nodeName: "HTML",
                childNodeCount: 1,
                children: [],
            },
        ],
    };
}
function htmlWithChildren() {
    return {
        nodeId: 2,
        backendNodeId: 2,
        nodeType: 1,
        nodeName: "HTML",
        childNodeCount: 1,
        children: [
            {
                nodeId: 3,
                backendNodeId: 3,
                nodeType: 1,
                nodeName: "BODY",
                childNodeCount: 1,
                children: [
                    {
                        nodeId: 4,
                        backendNodeId: 4,
                        nodeType: 1,
                        nodeName: "DIV",
                        childNodeCount: 0,
                        children: [],
                    },
                ],
            },
        ],
    };
}
function simpleAxNodes() {
    const stringType = "string";
    return [
        {
            nodeId: "1",
            role: { type: stringType, value: "RootWebArea" },
            backendDOMNodeId: 2,
            childIds: ["2"],
            ignored: false,
        },
        {
            nodeId: "2",
            role: { type: stringType, value: "generic" },
            name: { type: stringType, value: "Content" },
            backendDOMNodeId: 4,
            parentId: "1",
            childIds: [],
            ignored: false,
        },
    ];
}
const baseHandlers = {
    "DOM.enable": async () => ({}),
    "Runtime.enable": async () => ({}),
    "Accessibility.enable": async () => ({}),
    "Accessibility.getFullAXTree": async () => ({ nodes: simpleAxNodes() }),
};
function makeCborError() {
    return new Error("CBOR: stack limit exceeded");
}
(0, vitest_1.describe)("captureHybridSnapshot CBOR fallbacks", () => {
    (0, vitest_1.it)("retries DOM.getDocument with reduced depths before succeeding", async () => {
        let domCalls = 0;
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "DOM.getDocument": async (params) => {
                domCalls += 1;
                if (domCalls === 1)
                    throw makeCborError();
                (0, vitest_1.expect)(params?.depth).toBe(256);
                return { root: completeDomTree() };
            },
        });
        const page = createFakePage(session);
        const snapshot = await (0, index_js_1.captureHybridSnapshot)(page);
        (0, vitest_1.expect)(snapshot.combinedTree).toContain("html");
        const depths = session
            .callsFor("DOM.getDocument")
            .map((c) => c.params?.depth);
        (0, vitest_1.expect)(depths).toEqual([-1, 256]);
    });
    (0, vitest_1.it)("throws StagehandDomProcessError after all DOM.getDocument attempts fail", async () => {
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "DOM.getDocument": async () => {
                throw makeCborError();
            },
        });
        const page = createFakePage(session);
        await (0, vitest_1.expect)((0, index_js_1.captureHybridSnapshot)(page)).rejects.toThrow(sdkErrors_js_1.StagehandDomProcessError);
    });
    (0, vitest_1.it)("hydrates truncated nodes by retrying DOM.describeNode depths", async () => {
        let domAttempts = 0;
        let describeAttempts = 0;
        const session = new mockCDPSession_js_1.MockCDPSession({
            ...baseHandlers,
            "DOM.getDocument": async (params) => {
                domAttempts += 1;
                if (domAttempts === 1)
                    throw makeCborError();
                (0, vitest_1.expect)(params?.depth).toBe(256);
                return { root: truncatedDomTree() };
            },
            "DOM.describeNode": async (params) => {
                describeAttempts += 1;
                if (describeAttempts === 1)
                    throw makeCborError();
                (0, vitest_1.expect)(params?.depth).toBe(64);
                return { node: htmlWithChildren() };
            },
        });
        const page = createFakePage(session);
        const snapshot = await (0, index_js_1.captureHybridSnapshot)(page);
        const describeDepths = session
            .callsFor("DOM.describeNode")
            .map((c) => c.params?.depth);
        (0, vitest_1.expect)(describeDepths).toEqual([-1, 64]);
        (0, vitest_1.expect)(snapshot.combinedXpathMap["0-4"]).toBe("/html[1]/body[1]/div[1]");
    });
});
