// @browserbasehq/stagehand/dist/cjs/tests/unit/xpath-resolver.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsdom_1 = require("jsdom");
const vitest_1 = require("vitest");
const xpathResolver_js_1 = require("../../lib/v3/dom/locatorScripts/xpathResolver.js");
const globalRef = globalThis;
const originalGlobals = {
    window: globalRef.window,
    document: globalRef.document,
    Node: globalRef.Node,
    NodeFilter: globalRef.NodeFilter,
    Element: globalRef.Element,
    HTMLElement: globalRef.HTMLElement,
    Document: globalRef.Document,
    DocumentFragment: globalRef.DocumentFragment,
    ShadowRoot: globalRef.ShadowRoot,
    XPathResult: globalRef.XPathResult,
};
let dom;
const installDomGlobals = () => {
    const win = dom.window;
    globalRef.window = win;
    globalRef.document = win.document;
    globalRef.Node = win.Node;
    globalRef.NodeFilter = win.NodeFilter;
    globalRef.Element = win.Element;
    globalRef.HTMLElement = win.HTMLElement;
    globalRef.Document = win.Document;
    globalRef.DocumentFragment =
        win.DocumentFragment;
    globalRef.ShadowRoot = win.ShadowRoot;
    globalRef.XPathResult = win.XPathResult;
};
const restoreDomGlobals = () => {
    for (const [key, value] of Object.entries(originalGlobals)) {
        if (value === undefined) {
            delete globalRef[key];
        }
        else {
            globalRef[key] = value;
        }
    }
};
(0, vitest_1.describe)("xpathResolver composed traversal", () => {
    (0, vitest_1.beforeAll)(() => {
        dom = new jsdom_1.JSDOM("<!doctype html><html><body></body></html>");
        installDomGlobals();
    });
    (0, vitest_1.afterAll)(() => {
        dom.window.close();
        restoreDomGlobals();
    });
    (0, vitest_1.beforeEach)(() => {
        document.body.innerHTML = "";
    });
    (0, vitest_1.it)("counts matches across light + shadow DOM without double counting", () => {
        document.body.innerHTML =
            '<div id="light-1"></div>' +
                '<shadow-host id="host"></shadow-host>' +
                '<div id="light-2"></div>';
        const host = document.getElementById("host");
        const shadow = host.attachShadow({ mode: "open" });
        shadow.innerHTML = '<div id="shadow-1"></div><div id="shadow-2"></div>';
        (0, vitest_1.expect)((0, xpathResolver_js_1.countXPathMatches)("//div")).toBe(4);
    });
    (0, vitest_1.it)("resolves nth over composed tree in document-order DFS", () => {
        document.body.innerHTML =
            '<div id="light-1"></div>' +
                '<shadow-host id="host"></shadow-host>' +
                '<div id="light-2"></div>';
        const host = document.getElementById("host");
        const shadow = host.attachShadow({ mode: "open" });
        shadow.innerHTML = '<div id="shadow-1"></div><div id="shadow-2"></div>';
        (0, vitest_1.expect)((0, xpathResolver_js_1.resolveXPathAtIndex)("//div", 0)?.id).toBe("light-1");
        (0, vitest_1.expect)((0, xpathResolver_js_1.resolveXPathAtIndex)("//div", 1)?.id).toBe("shadow-1");
        (0, vitest_1.expect)((0, xpathResolver_js_1.resolveXPathAtIndex)("//div", 2)?.id).toBe("shadow-2");
        (0, vitest_1.expect)((0, xpathResolver_js_1.resolveXPathAtIndex)("//div", 3)?.id).toBe("light-2");
    });
});
