// @browserbasehq/stagehand/dist/cjs/tests/unit/xpath-parser.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const xpathParser_js_1 = require("../../lib/v3/dom/locatorScripts/xpathParser.js");
(0, vitest_1.describe)("parseXPathSteps", () => {
    (0, vitest_1.describe)("basic tag parsing", () => {
        (0, vitest_1.it)("parses a simple absolute path", () => {
            (0, vitest_1.expect)((0, xpathParser_js_1.parseXPathSteps)("/html/body/div")).toEqual([
                { axis: "child", tag: "html", predicates: [] },
                { axis: "child", tag: "body", predicates: [] },
                { axis: "child", tag: "div", predicates: [] },
            ]);
        });
        (0, vitest_1.it)("lowercases tag names", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("/HTML/BODY");
            (0, vitest_1.expect)(steps[0].tag).toBe("html");
            (0, vitest_1.expect)(steps[1].tag).toBe("body");
        });
        (0, vitest_1.it)("treats wildcard correctly", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//*");
            (0, vitest_1.expect)(steps).toEqual([{ axis: "desc", tag: "*", predicates: [] }]);
        });
    });
    (0, vitest_1.describe)("axes", () => {
        (0, vitest_1.it)("distinguishes child (/) from descendant (//)", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("/html//div/span");
            (0, vitest_1.expect)(steps).toEqual([
                { axis: "child", tag: "html", predicates: [] },
                { axis: "desc", tag: "div", predicates: [] },
                { axis: "child", tag: "span", predicates: [] },
            ]);
        });
        (0, vitest_1.it)("handles leading //", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div");
            (0, vitest_1.expect)(steps[0].axis).toBe("desc");
        });
    });
    (0, vitest_1.describe)("positional indices", () => {
        (0, vitest_1.it)("parses positional index", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("/div[1]/span[3]");
            (0, vitest_1.expect)(steps[0]).toMatchObject({
                tag: "div",
                predicates: [{ type: "index", index: 1 }],
            });
            (0, vitest_1.expect)(steps[1]).toMatchObject({
                tag: "span",
                predicates: [{ type: "index", index: 3 }],
            });
        });
        (0, vitest_1.it)("clamps index to minimum 1", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("/div[0]");
            (0, vitest_1.expect)(steps[0].predicates[0]).toMatchObject({
                type: "index",
                index: 1,
            });
        });
        (0, vitest_1.it)("keeps multiple positional predicates in order", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[2][3]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "index", index: 2 },
                { type: "index", index: 3 },
            ]);
        });
    });
    (0, vitest_1.describe)("attribute predicates", () => {
        (0, vitest_1.it)("parses single attribute predicate with single quotes", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//img[@alt='Stagehand']");
            (0, vitest_1.expect)(steps).toEqual([
                {
                    axis: "desc",
                    tag: "img",
                    predicates: [{ type: "attrEquals", name: "alt", value: "Stagehand" }],
                },
            ]);
        });
        (0, vitest_1.it)("parses single attribute predicate with double quotes", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)('//img[@alt="Stagehand"]');
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "alt", value: "Stagehand" },
            ]);
        });
        (0, vitest_1.it)("parses multiple attribute predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@class='foo'][@id='bar']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "class", value: "foo" },
                { type: "attrEquals", name: "id", value: "bar" },
            ]);
        });
        (0, vitest_1.it)("parses attribute predicate combined with positional index", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@class='item'][2]");
            (0, vitest_1.expect)(steps[0]).toMatchObject({
                tag: "div",
                predicates: [
                    { type: "attrEquals", name: "class", value: "item" },
                    { type: "index", index: 2 },
                ],
            });
        });
        (0, vitest_1.it)("parses attribute with hyphenated name", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@data-testid='submit']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "data-testid", value: "submit" },
            ]);
        });
        (0, vitest_1.it)("parses attribute with empty value", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//input[@value='']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "value", value: "" },
            ]);
        });
        (0, vitest_1.it)("parses attribute value containing closing bracket", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@title='array[0]']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "title", value: "array[0]" },
            ]);
        });
        (0, vitest_1.it)("parses attribute value containing multiple brackets", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@data-json='[1,2,3]']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrEquals", name: "data-json", value: "[1,2,3]" },
            ]);
        });
        (0, vitest_1.it)("parses attribute value containing a closing bracket", () => {
            // The step splitter should ignore ] characters inside quotes.
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@title='a]b']/span");
            (0, vitest_1.expect)(steps).toEqual([
                {
                    axis: "desc",
                    tag: "div",
                    predicates: [{ type: "attrEquals", name: "title", value: "a]b" }],
                },
                { axis: "child", tag: "span", predicates: [] },
            ]);
        });
        (0, vitest_1.it)("parses attribute existence predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//iframe[@data-test]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrExists", name: "data-test" },
            ]);
        });
        (0, vitest_1.it)("parses attribute contains predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//iframe[contains(@src,'checkout')]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrContains", name: "src", value: "checkout" },
            ]);
        });
        (0, vitest_1.it)("parses attribute starts-with predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//button[starts-with(@id,'save-')]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "attrStartsWith", name: "id", value: "save-" },
            ]);
        });
    });
    (0, vitest_1.describe)("text predicates", () => {
        (0, vitest_1.it)("parses text equality", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//button[text()='Submit']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "textEquals", value: "Submit" },
            ]);
        });
        (0, vitest_1.it)("parses text contains", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[contains(text(),'Welcome')]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "textContains", value: "Welcome" },
            ]);
        });
        (0, vitest_1.it)("parses normalize-space on text", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[normalize-space(text())='Hello world']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "textEquals", value: "Hello world", normalize: true },
            ]);
        });
    });
    (0, vitest_1.describe)("boolean predicates", () => {
        (0, vitest_1.it)("parses and predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@a='x' and @b='y']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                {
                    type: "and",
                    predicates: [
                        { type: "attrEquals", name: "a", value: "x" },
                        { type: "attrEquals", name: "b", value: "y" },
                    ],
                },
            ]);
        });
        (0, vitest_1.it)("parses operators without surrounding whitespace", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[not(@x)and@y='z']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                {
                    type: "and",
                    predicates: [
                        { type: "not", predicate: { type: "attrExists", name: "x" } },
                        { type: "attrEquals", name: "y", value: "z" },
                    ],
                },
            ]);
        });
        (0, vitest_1.it)("parses or predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@a='x' or @b='y']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                {
                    type: "or",
                    predicates: [
                        { type: "attrEquals", name: "a", value: "x" },
                        { type: "attrEquals", name: "b", value: "y" },
                    ],
                },
            ]);
        });
        (0, vitest_1.it)("parses not predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//button[not(@disabled)]");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                { type: "not", predicate: { type: "attrExists", name: "disabled" } },
            ]);
        });
        (0, vitest_1.it)("does not treat @and as a boolean operator", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//div[@and='x' and @y='z']");
            (0, vitest_1.expect)(steps[0].predicates).toEqual([
                {
                    type: "and",
                    predicates: [
                        { type: "attrEquals", name: "and", value: "x" },
                        { type: "attrEquals", name: "y", value: "z" },
                    ],
                },
            ]);
        });
    });
    (0, vitest_1.describe)("multi-step with predicates", () => {
        (0, vitest_1.it)("parses complex path with mixed predicates", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("/html/body//div[@class='container']/ul/li[3]");
            (0, vitest_1.expect)(steps).toEqual([
                { axis: "child", tag: "html", predicates: [] },
                { axis: "child", tag: "body", predicates: [] },
                {
                    axis: "desc",
                    tag: "div",
                    predicates: [
                        { type: "attrEquals", name: "class", value: "container" },
                    ],
                },
                { axis: "child", tag: "ul", predicates: [] },
                { axis: "child", tag: "li", predicates: [{ type: "index", index: 3 }] },
            ]);
        });
    });
    (0, vitest_1.describe)("edge cases", () => {
        (0, vitest_1.it)("returns empty array for empty string", () => {
            (0, vitest_1.expect)((0, xpathParser_js_1.parseXPathSteps)("")).toEqual([]);
        });
        (0, vitest_1.it)("strips xpath= prefix", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("xpath=//div");
            (0, vitest_1.expect)(steps).toEqual([{ axis: "desc", tag: "div", predicates: [] }]);
        });
        (0, vitest_1.it)("strips XPATH= prefix (case-insensitive)", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("XPATH=//div");
            (0, vitest_1.expect)(steps).toEqual([{ axis: "desc", tag: "div", predicates: [] }]);
        });
        (0, vitest_1.it)("handles forward slashes inside attribute values", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//a[@href='/api/endpoint']");
            (0, vitest_1.expect)(steps).toEqual([
                {
                    axis: "desc",
                    tag: "a",
                    predicates: [
                        { type: "attrEquals", name: "href", value: "/api/endpoint" },
                    ],
                },
            ]);
        });
        (0, vitest_1.it)("handles URL attribute values with multiple slashes", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("//a[@data-url='http://example.com/path/to/page']");
            (0, vitest_1.expect)(steps).toEqual([
                {
                    axis: "desc",
                    tag: "a",
                    predicates: [
                        {
                            type: "attrEquals",
                            name: "data-url",
                            value: "http://example.com/path/to/page",
                        },
                    ],
                },
            ]);
        });
        (0, vitest_1.it)("handles whitespace", () => {
            const steps = (0, xpathParser_js_1.parseXPathSteps)("  //div  ");
            (0, vitest_1.expect)(steps.length).toBe(1);
            (0, vitest_1.expect)(steps[0].tag).toBe("div");
        });
    });
});
(0, vitest_1.describe)("applyPredicates", () => {
    const makeElement = (id) => {
        return {
            localName: "div",
            getAttribute: (name) => (name === "id" ? id : null),
        };
    };
    (0, vitest_1.it)("applies positional predicates sequentially", () => {
        const elements = ["a", "b", "c", "d"].map(makeElement);
        const predicates = [
            { type: "index", index: 2 },
            { type: "index", index: 3 },
        ];
        (0, vitest_1.expect)((0, xpathParser_js_1.applyPredicates)(elements, predicates)).toEqual([]);
    });
});
