// @browserbasehq/stagehand/dist/cjs/lib/v3/types/public/apiErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StagehandResponseParseError = exports.StagehandResponseBodyError = exports.StagehandServerError = exports.StagehandHttpError = exports.StagehandAPIUnauthorizedError = exports.StagehandAPIError = void 0;
class StagehandAPIError extends Error {
    constructor(message) {
        super(message);
        this.name = this.constructor.name;
    }
}
exports.StagehandAPIError = StagehandAPIError;
class StagehandAPIUnauthorizedError extends StagehandAPIError {
    constructor(message) {
        super(message || "Unauthorized request");
    }
}
exports.StagehandAPIUnauthorizedError = StagehandAPIUnauthorizedError;
class StagehandHttpError extends StagehandAPIError {
    constructor(message) {
        super(message);
    }
}
exports.StagehandHttpError = StagehandHttpError;
class StagehandServerError extends StagehandAPIError {
    constructor(message) {
        super(message);
    }
}
exports.StagehandServerError = StagehandServerError;
class StagehandResponseBodyError extends StagehandAPIError {
    constructor() {
        super("Response body is null");
    }
}
exports.StagehandResponseBodyError = StagehandResponseBodyError;
class StagehandResponseParseError extends StagehandAPIError {
    constructor(message) {
        super(message);
    }
}
exports.StagehandResponseParseError = StagehandResponseParseError;
