// @browserbasehq/stagehand/dist/cjs/lib/v3/types/public/sdkErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdpConnectionClosedError = exports.StagehandClosedError = exports.AgentAbortError = exports.StreamingCallbacksInNonStreamingModeError = exports.ConnectionTimeoutError = exports.PageNotFoundError = exports.ObserveTimeoutError = exports.ExtractTimeoutError = exports.ActTimeoutError = exports.TimeoutError = exports.ResponseParseError = exports.ResponseBodyError = exports.ElementNotVisibleError = exports.StagehandShadowSegmentNotFoundError = exports.StagehandShadowSegmentEmptyError = exports.StagehandShadowRootMissingError = exports.MCPConnectionError = exports.StagehandInitError = exports.ZodSchemaValidationError = exports.CuaModelRequiredError = exports.ExperimentalNotConfiguredError = exports.ExperimentalApiConflictError = exports.XPathResolutionError = exports.ContentFrameNotFoundError = exports.StagehandIframeError = exports.LLMResponseError = exports.StagehandClickError = exports.StagehandLocatorError = exports.StagehandDomProcessError = exports.StagehandEvalError = exports.CreateChatCompletionResponseError = exports.StagehandMissingArgumentError = exports.AgentScreenshotProviderError = exports.StagehandElementNotFoundError = exports.CookieSetError = exports.CookieValidationError = exports.StagehandInvalidArgumentError = exports.HandlerNotInitializedError = exports.MissingLLMConfigurationError = exports.CaptchaTimeoutError = exports.BrowserbaseSessionNotFoundError = exports.StagehandNotInitializedError = exports.InvalidAISDKModelFormatError = exports.UnsupportedAISDKModelProviderError = exports.UnsupportedModelProviderError = exports.UnsupportedModelError = exports.MissingEnvironmentVariableError = exports.StagehandEnvironmentError = exports.StagehandDefaultError = exports.StagehandError = void 0;
exports.UnderstudyCommandException = exports.StagehandSnapshotError = exports.StagehandSetExtraHTTPHeadersError = void 0;
// Avoid .js extension so bundlers resolve TS source
const version_js_1 = require("../../../version.js");
class StagehandError extends Error {
    cause;
    constructor(message, cause) {
        super(message);
        this.name = this.constructor.name;
        if (cause !== undefined) {
            this.cause = cause;
        }
    }
}
exports.StagehandError = StagehandError;
class StagehandDefaultError extends StagehandError {
    constructor(error) {
        if (error instanceof Error || error instanceof StagehandError) {
            super(`\nHey! We're sorry you ran into an error. \nStagehand version: ${version_js_1.STAGEHAND_VERSION} \nIf you need help, please open a Github issue or reach out to us on Discord: https://stagehand.dev/discord\n\nFull error:\n${error.message}`);
        }
    }
}
exports.StagehandDefaultError = StagehandDefaultError;
class StagehandEnvironmentError extends StagehandError {
    constructor(currentEnvironment, requiredEnvironment, feature) {
        super(`You seem to be setting the current environment to ${currentEnvironment}.` +
            `Ensure the environment is set to ${requiredEnvironment} if you want to use ${feature}.`);
    }
}
exports.StagehandEnvironmentError = StagehandEnvironmentError;
class MissingEnvironmentVariableError extends StagehandError {
    constructor(missingEnvironmentVariable, feature) {
        super(`${missingEnvironmentVariable} is required to use ${feature}.` +
            `Please set ${missingEnvironmentVariable} in your environment.`);
    }
}
exports.MissingEnvironmentVariableError = MissingEnvironmentVariableError;
class UnsupportedModelError extends StagehandError {
    constructor(supportedModels, feature) {
        const message = feature
            ? `${feature} requires a valid model.`
            : `Unsupported model.`;
        const guidance = `\n\nPlease use the provider/model format (e.g., "openai/gpt-4o", "anthropic/claude-sonnet-4-5", "google/gemini-3-flash-preview").` +
            `\n\nFor a complete list of supported models and providers, see: https://docs.stagehand.dev/v3/configuration/models#configuration-setup`;
        super(`${message}${guidance}`);
    }
}
exports.UnsupportedModelError = UnsupportedModelError;
class UnsupportedModelProviderError extends StagehandError {
    constructor(supportedProviders, feature) {
        super(feature
            ? `${feature} requires one of the following model providers: ${supportedProviders}`
            : `please use one of the supported model providers: ${supportedProviders}`);
    }
}
exports.UnsupportedModelProviderError = UnsupportedModelProviderError;
class UnsupportedAISDKModelProviderError extends StagehandError {
    constructor(provider, supportedProviders) {
        super(`${provider} is not currently supported for aiSDK. please use one of the supported model providers: ${supportedProviders}`);
    }
}
exports.UnsupportedAISDKModelProviderError = UnsupportedAISDKModelProviderError;
class InvalidAISDKModelFormatError extends StagehandError {
    constructor(modelName) {
        super(`${modelName} does not follow correct format for specifying aiSDK models. Please define your model as 'provider/model-name'. For example: \`model: 'openai/gpt-4o-mini'\``);
    }
}
exports.InvalidAISDKModelFormatError = InvalidAISDKModelFormatError;
class StagehandNotInitializedError extends StagehandError {
    constructor(prop) {
        super(`You seem to be calling \`${prop}\` on a page in an uninitialized \`Stagehand\` object. ` +
            `Ensure you are running \`await stagehand.init()\` on the Stagehand object before ` +
            `referencing the \`page\` object.`);
    }
}
exports.StagehandNotInitializedError = StagehandNotInitializedError;
class BrowserbaseSessionNotFoundError extends StagehandError {
    constructor() {
        super("No Browserbase session ID found");
    }
}
exports.BrowserbaseSessionNotFoundError = BrowserbaseSessionNotFoundError;
class CaptchaTimeoutError extends StagehandError {
    constructor() {
        super("Captcha timeout");
    }
}
exports.CaptchaTimeoutError = CaptchaTimeoutError;
class MissingLLMConfigurationError extends StagehandError {
    constructor() {
        super("No LLM API key or LLM Client configured. An LLM API key or a custom LLM Client " +
            "is required to use act, extract, or observe.");
    }
}
exports.MissingLLMConfigurationError = MissingLLMConfigurationError;
class HandlerNotInitializedError extends StagehandError {
    constructor(handlerType) {
        super(`${handlerType} handler not initialized`);
    }
}
exports.HandlerNotInitializedError = HandlerNotInitializedError;
class StagehandInvalidArgumentError extends StagehandError {
    constructor(message) {
        super(`InvalidArgumentError: ${message}`);
    }
}
exports.StagehandInvalidArgumentError = StagehandInvalidArgumentError;
class CookieValidationError extends StagehandError {
    constructor(message) {
        super(message);
    }
}
exports.CookieValidationError = CookieValidationError;
class CookieSetError extends StagehandError {
    constructor(message) {
        super(message);
    }
}
exports.CookieSetError = CookieSetError;
class StagehandElementNotFoundError extends StagehandError {
    constructor(xpaths) {
        super(`Could not find an element for the given xPath(s): ${xpaths}`);
    }
}
exports.StagehandElementNotFoundError = StagehandElementNotFoundError;
class AgentScreenshotProviderError extends StagehandError {
    constructor(message) {
        super(`ScreenshotProviderError: ${message}`);
    }
}
exports.AgentScreenshotProviderError = AgentScreenshotProviderError;
class StagehandMissingArgumentError extends StagehandError {
    constructor(message) {
        super(`MissingArgumentError: ${message}`);
    }
}
exports.StagehandMissingArgumentError = StagehandMissingArgumentError;
class CreateChatCompletionResponseError extends StagehandError {
    constructor(message) {
        super(`CreateChatCompletionResponseError: ${message}`);
    }
}
exports.CreateChatCompletionResponseError = CreateChatCompletionResponseError;
class StagehandEvalError extends StagehandError {
    constructor(message) {
        super(`StagehandEvalError: ${message}`);
    }
}
exports.StagehandEvalError = StagehandEvalError;
class StagehandDomProcessError extends StagehandError {
    constructor(message) {
        super(`Error Processing Dom: ${message}`);
    }
}
exports.StagehandDomProcessError = StagehandDomProcessError;
class StagehandLocatorError extends StagehandError {
    constructor(action, selector, message) {
        super(`Error ${action} Element with selector: ${selector} Reason: ${message}`);
    }
}
exports.StagehandLocatorError = StagehandLocatorError;
class StagehandClickError extends StagehandError {
    constructor(message, selector) {
        super(`Error Clicking Element with selector: ${selector} Reason: ${message}`);
    }
}
exports.StagehandClickError = StagehandClickError;
class LLMResponseError extends StagehandError {
    constructor(primitive, message) {
        super(`${primitive} LLM response error: ${message}`);
    }
}
exports.LLMResponseError = LLMResponseError;
class StagehandIframeError extends StagehandError {
    constructor(frameUrl, message) {
        super(`Unable to resolve frameId for iframe with URL: ${frameUrl} Full error: ${message}`);
    }
}
exports.StagehandIframeError = StagehandIframeError;
class ContentFrameNotFoundError extends StagehandError {
    constructor(selector) {
        super(`Unable to obtain a content frame for selector: ${selector}`);
    }
}
exports.ContentFrameNotFoundError = ContentFrameNotFoundError;
class XPathResolutionError extends StagehandError {
    constructor(xpath) {
        super(`XPath "${xpath}" does not resolve in the current page or frames`);
    }
}
exports.XPathResolutionError = XPathResolutionError;
class ExperimentalApiConflictError extends StagehandError {
    constructor() {
        super("`experimental` mode cannot be used together with the Stagehand API. " +
            "To use experimental features, set experimental: true and disableAPI: true in the stagehand constructor. " +
            "To use the Stagehand API, set experimental: false and disableAPI: false (or omit it) in the stagehand constructor.");
    }
}
exports.ExperimentalApiConflictError = ExperimentalApiConflictError;
class ExperimentalNotConfiguredError extends StagehandError {
    constructor(featureName) {
        super(`Feature "${featureName}" is an experimental feature, and cannot be configured when disableAPI: false.
    Please set experimental: true and disableAPI: true in the stagehand constructor to use this feature.
    If you wish to use the Stagehand API, please ensure ${featureName} is not defined in your function call,
    and set experimental: false, disableAPI: false (or omit it) in the Stagehand constructor.`);
    }
}
exports.ExperimentalNotConfiguredError = ExperimentalNotConfiguredError;
class CuaModelRequiredError extends StagehandError {
    constructor(availableModels) {
        super(`To use the computer use agent (CUA), please provide a CUA model in the agent constructor or stagehand config. ` +
            `Try one of our supported CUA models: ${availableModels.join(", ")}`);
    }
}
exports.CuaModelRequiredError = CuaModelRequiredError;
class ZodSchemaValidationError extends Error {
    received;
    issues;
    constructor(received, issues) {
        super(`Zod schema validation failed

— Received —
${JSON.stringify(received, null, 2)}

— Issues —
${JSON.stringify(issues, null, 2)}`);
        this.received = received;
        this.issues = issues;
        this.name = "ZodSchemaValidationError";
    }
}
exports.ZodSchemaValidationError = ZodSchemaValidationError;
class StagehandInitError extends StagehandError {
    constructor(message) {
        super(message);
    }
}
exports.StagehandInitError = StagehandInitError;
class MCPConnectionError extends StagehandError {
    serverUrl;
    originalError;
    constructor(serverUrl, originalError) {
        const errorMessage = originalError instanceof Error
            ? originalError.message
            : String(originalError);
        super(`Failed to connect to MCP server at "${serverUrl}". ${errorMessage}. ` +
            `Please verify the server URL is correct and the server is running.`);
        this.serverUrl = serverUrl;
        this.originalError = originalError;
    }
}
exports.MCPConnectionError = MCPConnectionError;
class StagehandShadowRootMissingError extends StagehandError {
    constructor(detail) {
        super(`No shadow root present on the resolved host` +
            (detail ? `: ${detail}` : ""));
    }
}
exports.StagehandShadowRootMissingError = StagehandShadowRootMissingError;
class StagehandShadowSegmentEmptyError extends StagehandError {
    constructor() {
        super(`Empty selector segment after shadow-DOM hop ("//")`);
    }
}
exports.StagehandShadowSegmentEmptyError = StagehandShadowSegmentEmptyError;
class StagehandShadowSegmentNotFoundError extends StagehandError {
    constructor(segment, hint) {
        super(`Shadow segment '${segment}' matched no element inside shadow root` +
            (hint ? ` ${hint}` : ""));
    }
}
exports.StagehandShadowSegmentNotFoundError = StagehandShadowSegmentNotFoundError;
class ElementNotVisibleError extends StagehandError {
    constructor(selector) {
        super(`Element not visible (no box model): ${selector}`);
    }
}
exports.ElementNotVisibleError = ElementNotVisibleError;
class ResponseBodyError extends StagehandError {
    constructor(message) {
        super(`Failed to retrieve response body: ${message}`);
    }
}
exports.ResponseBodyError = ResponseBodyError;
class ResponseParseError extends StagehandError {
    constructor(message) {
        super(`Failed to parse response: ${message}`);
    }
}
exports.ResponseParseError = ResponseParseError;
class TimeoutError extends StagehandError {
    constructor(operation, timeoutMs) {
        super(`${operation} timed out after ${timeoutMs}ms`);
    }
}
exports.TimeoutError = TimeoutError;
class ActTimeoutError extends TimeoutError {
    constructor(timeoutMs) {
        super("act()", timeoutMs);
        this.name = "ActTimeoutError";
    }
}
exports.ActTimeoutError = ActTimeoutError;
class ExtractTimeoutError extends TimeoutError {
    constructor(timeoutMs) {
        super("extract()", timeoutMs);
        this.name = "ExtractTimeoutError";
    }
}
exports.ExtractTimeoutError = ExtractTimeoutError;
class ObserveTimeoutError extends TimeoutError {
    constructor(timeoutMs) {
        super("observe()", timeoutMs);
        this.name = "ObserveTimeoutError";
    }
}
exports.ObserveTimeoutError = ObserveTimeoutError;
class PageNotFoundError extends StagehandError {
    constructor(identifier) {
        super(`No Page found for ${identifier}`);
    }
}
exports.PageNotFoundError = PageNotFoundError;
class ConnectionTimeoutError extends StagehandError {
    constructor(message) {
        super(`Connection timeout: ${message}`);
    }
}
exports.ConnectionTimeoutError = ConnectionTimeoutError;
class StreamingCallbacksInNonStreamingModeError extends StagehandError {
    invalidCallbacks;
    constructor(invalidCallbacks) {
        super(`Streaming-only callback(s) "${invalidCallbacks.join('", "')}" cannot be used in non-streaming mode. ` +
            `Set 'stream: true' in AgentConfig to use these callbacks.`);
        this.invalidCallbacks = invalidCallbacks;
    }
}
exports.StreamingCallbacksInNonStreamingModeError = StreamingCallbacksInNonStreamingModeError;
class AgentAbortError extends StagehandError {
    reason;
    constructor(reason) {
        const message = reason
            ? `Agent execution was aborted: ${reason}`
            : "Agent execution was aborted";
        super(message);
        this.reason = reason || "aborted";
    }
}
exports.AgentAbortError = AgentAbortError;
class StagehandClosedError extends StagehandError {
    constructor() {
        super("Stagehand session was closed");
    }
}
exports.StagehandClosedError = StagehandClosedError;
class CdpConnectionClosedError extends StagehandError {
    constructor(reason) {
        super(`CDP connection closed: ${reason}`);
    }
}
exports.CdpConnectionClosedError = CdpConnectionClosedError;
class StagehandSetExtraHTTPHeadersError extends StagehandError {
    failures;
    constructor(failures) {
        super(`setExtraHTTPHeaders failed for ${failures.length} session(s): ${failures.join(", ")}`);
        this.failures = failures;
    }
}
exports.StagehandSetExtraHTTPHeadersError = StagehandSetExtraHTTPHeadersError;
class StagehandSnapshotError extends StagehandError {
    constructor(cause) {
        const suffix = cause instanceof Error
            ? `: ${cause.message}`
            : cause
                ? `: ${String(cause)}`
                : "";
        super(`error taking snapshot${suffix}`, cause);
    }
}
exports.StagehandSnapshotError = StagehandSnapshotError;
class UnderstudyCommandException extends StagehandError {
    constructor(message, cause) {
        super(message, cause);
        this.name = "UnderstudyCommandException";
    }
}
exports.UnderstudyCommandException = UnderstudyCommandException;
