// @browserbasehq/stagehand/dist/esm/lib/v3/cache/CacheStorage.js
import fs from "fs";
import path from "path";
const jsonClone = (value) => {
    const serialized = JSON.stringify(value);
    if (serialized === undefined) {
        return value;
    }
    return JSON.parse(serialized);
};
export class CacheStorage {
    logger;
    dir;
    memoryStore;
    constructor(logger, dir, memoryStore) {
        this.logger = logger;
        this.dir = dir;
        this.memoryStore = memoryStore;
    }
    static create(cacheDir, logger, options) {
        if (!cacheDir) {
            return new CacheStorage(logger);
        }
        const resolved = path.resolve(cacheDir);
        try {
            fs.mkdirSync(resolved, { recursive: true });
            return new CacheStorage(logger, resolved);
        }
        catch (err) {
            const label = options?.label ?? "cache directory";
            logger({
                category: "cache",
                message: `unable to initialize ${label}: ${resolved}`,
                level: 1,
                auxiliary: {
                    error: { value: String(err), type: "string" },
                },
            });
            return new CacheStorage(logger);
        }
    }
    static createMemory(logger) {
        return new CacheStorage(logger, undefined, new Map());
    }
    get directory() {
        return this.dir;
    }
    get enabled() {
        return !!this.dir || !!this.memoryStore;
    }
    resolvePath(fileName) {
        if (!this.dir)
            return null;
        return path.join(this.dir, fileName);
    }
    async readJson(fileName) {
        if (this.memoryStore) {
            if (!this.memoryStore.has(fileName)) {
                return { value: null };
            }
            const existing = this.memoryStore.get(fileName);
            return { value: jsonClone(existing) };
        }
        const filePath = this.resolvePath(fileName);
        if (!filePath) {
            return { value: null };
        }
        try {
            const raw = await fs.promises.readFile(filePath, "utf8");
            return { value: JSON.parse(raw) };
        }
        catch (err) {
            const code = err?.code;
            if (code === "ENOENT") {
                return { value: null };
            }
            return { value: null, error: err, path: filePath };
        }
    }
    async writeJson(fileName, data) {
        if (this.memoryStore) {
            this.memoryStore.set(fileName, jsonClone(data));
            return {};
        }
        const filePath = this.resolvePath(fileName);
        if (!filePath) {
            return {};
        }
        try {
            await fs.promises.mkdir(path.dirname(filePath), { recursive: true });
            await fs.promises.writeFile(filePath, JSON.stringify(data, null, 2), "utf8");
            return {};
        }
        catch (err) {
            return { error: err, path: filePath };
        }
    }
}
