// @browserbasehq/stagehand/dist/esm/lib/v3/mcp/utils.js
import { jsonSchemaToZod } from "../../utils.js";
import { connectToMCPServer } from "./connection.js";
export const resolveTools = async (clients, userTools) => {
    const tools = { ...userTools };
    for (const client of clients) {
        let clientInstance;
        if (typeof client === "string") {
            clientInstance = await connectToMCPServer(client);
        }
        else {
            clientInstance = client;
        }
        let nextCursor = undefined;
        do {
            const clientTools = await clientInstance.listTools({
                cursor: nextCursor,
            });
            for (const tool of clientTools.tools) {
                tools[tool.name] = {
                    description: tool.description,
                    inputSchema: jsonSchemaToZod(tool.inputSchema),
                    execute: async (input) => {
                        const result = await clientInstance.callTool({
                            name: tool.name,
                            arguments: input,
                        });
                        return result;
                    },
                };
            }
            nextCursor = clientTools.nextCursor;
        } while (nextCursor);
    }
    return tools;
};
