// @browserbasehq/stagehand/dist/cjs/lib/v3/handlers/observeHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObserveHandler = void 0;
// lib/v3/handlers/observeHandler.ts
const inference_js_1 = require("../../inference.js");
const utils_js_1 = require("../../utils.js");
const logger_js_1 = require("../logger.js");
const methods_js_1 = require("../types/public/methods.js");
const index_js_1 = require("../understudy/a11y/snapshot/index.js");
const handlers_js_1 = require("../types/private/handlers.js");
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const timeoutGuard_js_1 = require("./handlerUtils/timeoutGuard.js");
class ObserveHandler {
    llmClient;
    defaultModelName;
    defaultClientOptions;
    resolveLlmClient;
    systemPrompt;
    logInferenceToFile;
    experimental;
    onMetrics;
    constructor(llmClient, defaultModelName, defaultClientOptions, resolveLlmClient, systemPrompt, logInferenceToFile, experimental, onMetrics) {
        this.llmClient = llmClient;
        this.defaultModelName = defaultModelName;
        this.defaultClientOptions = defaultClientOptions;
        this.resolveLlmClient = resolveLlmClient;
        this.systemPrompt = systemPrompt ?? "";
        this.logInferenceToFile = logInferenceToFile ?? false;
        this.experimental = experimental ?? false;
        this.onMetrics = onMetrics;
    }
    async observe(params) {
        const { instruction, page, timeout, selector, model, variables } = params;
        const llmClient = this.resolveLlmClient(model);
        const ensureTimeRemaining = (0, timeoutGuard_js_1.createTimeoutGuard)(timeout, (ms) => new sdkErrors_js_1.ObserveTimeoutError(ms));
        const effectiveInstruction = instruction ??
            "Find elements that can be used for any future actions in the page. These may be navigation links, related pages, section/subsection links, buttons, or other interactive elements. Be comprehensive: if there are multiple elements that may be relevant for future actions, return all of them.";
        (0, logger_js_1.v3Logger)({
            category: "observation",
            message: "starting observation",
            level: 1,
            auxiliary: {
                instruction: {
                    value: effectiveInstruction,
                    type: "string",
                },
            },
        });
        // Build the hybrid snapshot (a11y-centric text tree + lookup maps)
        const focusSelector = selector?.replace(/^xpath=/i, "") ?? "";
        ensureTimeRemaining();
        const snapshot = await (0, index_js_1.captureHybridSnapshot)(page, {
            experimental: this.experimental,
            focusSelector: focusSelector || undefined,
        });
        const combinedTree = snapshot.combinedTree;
        const combinedXpathMap = snapshot.combinedXpathMap ?? {};
        (0, logger_js_1.v3Logger)({
            category: "observation",
            message: "Got accessibility tree data",
            level: 1,
        });
        // Call the LLM to propose actionable elements
        ensureTimeRemaining();
        const observationResponse = await (0, inference_js_1.observe)({
            instruction: effectiveInstruction,
            domElements: combinedTree,
            llmClient,
            userProvidedInstructions: this.systemPrompt,
            logger: logger_js_1.v3Logger,
            logInferenceToFile: this.logInferenceToFile,
            supportedActions: Object.values(handlers_js_1.SupportedUnderstudyAction),
            variables,
        });
        const { prompt_tokens = 0, completion_tokens = 0, reasoning_tokens = 0, cached_input_tokens = 0, inference_time_ms = 0, } = observationResponse;
        // Update OBSERVE metrics from the LLM observation call
        this.onMetrics?.(methods_js_1.V3FunctionName.OBSERVE, prompt_tokens, completion_tokens, reasoning_tokens, cached_input_tokens, inference_time_ms);
        // Map elementIds -> selectors via combinedXpathMap
        const elementsWithSelectors = (await Promise.all(observationResponse.elements.map(async (element) => {
            const { elementId, ...rest } = element; // rest may or may not have method/arguments
            if (typeof elementId === "string" && elementId.includes("-")) {
                const lookUpIndex = elementId;
                const xpath = combinedXpathMap[lookUpIndex];
                const trimmedXpath = (0, utils_js_1.trimTrailingTextNode)(xpath);
                if (!trimmedXpath)
                    return undefined;
                // For dragAndDrop, convert element ID in arguments to xpath (target element)
                let resolvedArgs = rest.arguments;
                if (rest.method === "dragAndDrop" &&
                    Array.isArray(rest.arguments) &&
                    rest.arguments.length > 0) {
                    const targetArg = rest.arguments[0];
                    // Check if argument looks like an element ID (e.g., "1-67")
                    if (typeof targetArg === "string" &&
                        /^\d+-\d+$/.test(targetArg)) {
                        const argXpath = combinedXpathMap[targetArg];
                        const trimmedArgXpath = (0, utils_js_1.trimTrailingTextNode)(argXpath);
                        if (trimmedArgXpath) {
                            resolvedArgs = [
                                `xpath=${trimmedArgXpath}`,
                                ...rest.arguments.slice(1),
                            ];
                        }
                        else {
                            // Target element lookup failed, filter out this action
                            (0, logger_js_1.v3Logger)({
                                category: "observation",
                                message: "dragAndDrop target element lookup failed",
                                level: 0,
                                auxiliary: {
                                    targetElementId: { value: targetArg, type: "string" },
                                    sourceElementId: { value: elementId, type: "string" },
                                },
                            });
                            return undefined;
                        }
                    }
                    else {
                        (0, logger_js_1.v3Logger)({
                            category: "observation",
                            message: "dragAndDrop target element invalid ID format",
                            level: 0,
                            auxiliary: {
                                targetElementId: { value: targetArg, type: "string" },
                                sourceElementId: { value: elementId, type: "string" },
                            },
                        });
                        return undefined;
                    }
                }
                return {
                    ...rest,
                    arguments: resolvedArgs,
                    selector: `xpath=${trimmedXpath}`,
                };
            }
            // shadow-root fallback:
            return {
                description: "an element inside a shadow DOM",
                method: "not-supported",
                arguments: [],
                selector: "not-supported",
            };
        }))).filter((e) => e !== undefined);
        (0, logger_js_1.v3Logger)({
            category: "observation",
            message: "found elements",
            level: 1,
            auxiliary: {
                elements: {
                    value: JSON.stringify(elementsWithSelectors),
                    type: "object",
                },
            },
        });
        return elementsWithSelectors;
    }
}
exports.ObserveHandler = ObserveHandler;
