// @browserbasehq/stagehand/dist/cjs/lib/v3/external_clients/aisdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AISdkClient = void 0;
const ai_1 = require("ai");
const LLMClient_js_1 = require("../llm/LLMClient.js");
class AISdkClient extends LLMClient_js_1.LLMClient {
    type = "aisdk";
    model;
    constructor({ model }) {
        super(model.modelId);
        this.model = model;
    }
    async createChatCompletion({ options, }) {
        const formattedMessages = options.messages.map((message) => {
            if (Array.isArray(message.content)) {
                if (message.role === "system") {
                    const systemMessage = {
                        role: "system",
                        content: message.content
                            .map((c) => ("text" in c ? c.text : ""))
                            .join("\n"),
                    };
                    return systemMessage;
                }
                const contentParts = message.content.map((content) => {
                    if ("image_url" in content) {
                        const imageContent = {
                            type: "image",
                            image: content.image_url.url,
                        };
                        return imageContent;
                    }
                    else {
                        const textContent = {
                            type: "text",
                            text: content.text,
                        };
                        return textContent;
                    }
                });
                if (message.role === "user") {
                    const userMessage = {
                        role: "user",
                        content: contentParts,
                    };
                    return userMessage;
                }
                else {
                    const textOnlyParts = contentParts.map((part) => ({
                        type: "text",
                        text: part.type === "image" ? "[Image]" : part.text,
                    }));
                    const assistantMessage = {
                        role: "assistant",
                        content: textOnlyParts,
                    };
                    return assistantMessage;
                }
            }
            return {
                role: message.role,
                content: message.content,
            };
        });
        if (options.response_model) {
            const response = await (0, ai_1.generateObject)({
                model: this.model,
                messages: formattedMessages,
                schema: options.response_model.schema,
            });
            return {
                data: response.object,
                usage: {
                    prompt_tokens: response.usage.inputTokens ?? 0,
                    completion_tokens: response.usage.outputTokens ?? 0,
                    reasoning_tokens: response.usage.reasoningTokens ?? 0,
                    cached_input_tokens: response.usage.cachedInputTokens ?? 0,
                    total_tokens: response.usage.totalTokens ?? 0,
                },
            };
        }
        const tools = {};
        for (const rawTool of options.tools) {
            tools[rawTool.name] = {
                description: rawTool.description,
                inputSchema: rawTool.parameters,
            };
        }
        const response = await (0, ai_1.generateText)({
            model: this.model,
            messages: formattedMessages,
            tools,
        });
        return {
            data: response.text,
            usage: {
                prompt_tokens: response.usage.inputTokens ?? 0,
                completion_tokens: response.usage.outputTokens ?? 0,
                reasoning_tokens: response.usage.reasoningTokens ?? 0,
                cached_input_tokens: response.usage.cachedInputTokens ?? 0,
                total_tokens: response.usage.totalTokens ?? 0,
            },
        };
    }
}
exports.AISdkClient = AISdkClient;
