// @browserbasehq/stagehand/dist/cjs/tests/integration/agent-experimental-validation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const zod_1 = require("zod");
const ai_1 = require("ai");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
const sdkErrors_js_1 = require("../../lib/v3/types/public/sdkErrors.js");
// Define a mock custom tool for testing
const mockCustomTool = (0, ai_1.tool)({
    description: "A mock tool for testing",
    inputSchema: zod_1.z.object({
        input: zod_1.z.string().describe("The input string"),
    }),
    execute: async ({ input }) => {
        return `Processed: ${input}`;
    },
});
test_1.test.describe("Stagehand agent experimental feature validation", () => {
    test_1.test.describe("Invalid argument errors", () => {
        let v3;
        test_1.test.beforeEach(async () => {
            v3 = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: false,
            });
            await v3.init();
        });
        test_1.test.afterEach(async () => {
            await v3?.close?.().catch(() => { });
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError when CUA and streaming are both enabled", async () => {
            try {
                v3.agent({
                    cua: true,
                    stream: true,
                    model: "anthropic/claude-sonnet-4-20250514",
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                (0, test_1.expect)(error.message).toContain("streaming");
                (0, test_1.expect)(error.message).toContain("not supported with CUA");
            }
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError for CUA + streaming even with experimental: true", async () => {
            // Close the non-experimental instance
            await v3.close();
            // Create an experimental instance
            const v3Experimental = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: true,
            });
            await v3Experimental.init();
            try {
                v3Experimental.agent({
                    cua: true,
                    stream: true,
                    model: "anthropic/claude-sonnet-4-20250514",
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                (0, test_1.expect)(error.message).toContain("streaming");
                (0, test_1.expect)(error.message).toContain("not supported with CUA");
            }
            finally {
                await v3Experimental.close();
            }
        });
    });
    test_1.test.describe("Experimental feature errors without experimental: true", () => {
        let v3;
        test_1.test.beforeEach(async () => {
            v3 = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: false,
            });
            await v3.init();
        });
        test_1.test.afterEach(async () => {
            await v3?.close?.().catch(() => { });
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for MCP integrations", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
                integrations: ["https://mcp.example.com"],
            });
            try {
                await agent.execute("test");
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("MCP integrations and custom tools");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for custom tools", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
                tools: {
                    mockCustomTool,
                },
            });
            try {
                await agent.execute("test");
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("MCP integrations and custom tools");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for streaming mode", async () => {
            try {
                const agent = v3.agent({
                    stream: true,
                    model: "anthropic/claude-sonnet-4-20250514",
                });
                await agent.execute("test instruction");
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("streaming");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for callbacks", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
            });
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onStepFinish: async () => { },
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("callbacks");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for abort signal", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
            });
            const controller = new AbortController();
            try {
                await agent.execute({
                    instruction: "test",
                    signal: controller.signal,
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("abort signal");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for message continuation", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
            });
            try {
                await agent.execute({
                    instruction: "test",
                    messages: [{ role: "user", content: "previous message" }],
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("message continuation");
            }
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError listing multiple features", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-sonnet-4-20250514",
            });
            const controller = new AbortController();
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: { onStepFinish: async () => { } },
                    signal: controller.signal,
                    messages: [{ role: "user", content: "previous" }],
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                const message = error.message;
                (0, test_1.expect)(message).toContain("callbacks");
                (0, test_1.expect)(message).toContain("abort signal");
                (0, test_1.expect)(message).toContain("message continuation");
            }
        });
    });
    test_1.test.describe("CUA agent unsupported features", () => {
        let v3;
        test_1.test.beforeEach(async () => {
            v3 = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: false,
            });
            await v3.init();
        });
        test_1.test.afterEach(async () => {
            await v3?.close?.().catch(() => { });
        });
        (0, test_1.test)("throws ExperimentalNotConfiguredError for CUA with integrations", async () => {
            // MCP integrations are still an experimental feature check (not unsupported)
            try {
                v3.agent({
                    cua: true,
                    model: "anthropic/claude-sonnet-4-20250514",
                    integrations: ["https://mcp.example.com"],
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.ExperimentalNotConfiguredError);
                (0, test_1.expect)(error.message).toContain("MCP integrations and custom tools");
            }
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError for CUA with abort signal (not supported)", async () => {
            const agent = v3.agent({
                cua: true,
                model: "anthropic/claude-sonnet-4-20250514",
            });
            const controller = new AbortController();
            try {
                await agent.execute({
                    instruction: "test",
                    signal: controller.signal,
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                (0, test_1.expect)(error.message).toContain("abort signal");
                (0, test_1.expect)(error.message).toContain("not supported with CUA");
            }
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError for CUA with message continuation (not supported)", async () => {
            const agent = v3.agent({
                cua: true,
                model: "anthropic/claude-sonnet-4-20250514",
            });
            try {
                await agent.execute({
                    instruction: "test",
                    messages: [{ role: "user", content: "previous message" }],
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                (0, test_1.expect)(error.message).toContain("message continuation");
                (0, test_1.expect)(error.message).toContain("not supported with CUA");
            }
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError for CUA with multiple unsupported features", async () => {
            const agent = v3.agent({
                cua: true,
                model: "anthropic/claude-sonnet-4-20250514",
            });
            const controller = new AbortController();
            try {
                await agent.execute({
                    instruction: "test",
                    signal: controller.signal,
                    messages: [{ role: "user", content: "previous message" }],
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                const message = error.message;
                (0, test_1.expect)(message).toContain("abort signal");
                (0, test_1.expect)(message).toContain("message continuation");
                (0, test_1.expect)(message).toContain("are not supported with CUA");
            }
        });
        (0, test_1.test)("throws StagehandInvalidArgumentError for CUA unsupported features even with experimental: true", async () => {
            // Close the non-experimental instance
            await v3.close();
            // Create an experimental instance
            const v3Experimental = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: true,
            });
            await v3Experimental.init();
            const agent = v3Experimental.agent({
                cua: true,
                model: "anthropic/claude-sonnet-4-20250514",
            });
            const controller = new AbortController();
            try {
                await agent.execute({
                    instruction: "test",
                    signal: controller.signal,
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StagehandInvalidArgumentError);
                (0, test_1.expect)(error.message).toContain("not supported with CUA");
            }
            finally {
                await v3Experimental.close();
            }
        });
    });
    test_1.test.describe("Valid configurations with experimental: true", () => {
        let v3;
        test_1.test.beforeEach(async () => {
            v3 = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: true,
            });
            await v3.init();
        });
        test_1.test.afterEach(async () => {
            await v3?.close?.().catch(() => { });
        });
        (0, test_1.test)("allows CUA without streaming", () => {
            (0, test_1.expect)(() => v3.agent({
                cua: true,
                model: "anthropic/claude-sonnet-4-20250514",
            })).not.toThrow();
        });
        (0, test_1.test)("allows streaming mode", () => {
            (0, test_1.expect)(() => v3.agent({
                stream: true,
                model: "anthropic/claude-sonnet-4-20250514",
            })).not.toThrow();
        });
        (0, test_1.test)("allows basic agent without experimental features", async () => {
            const v3NonExperimental = new v3_js_1.V3({
                ...v3_config_js_1.v3TestConfig,
                experimental: false,
            });
            await v3NonExperimental.init();
            try {
                // This should work - just creating a basic agent with no experimental features
                (0, test_1.expect)(() => v3NonExperimental.agent({
                    model: "anthropic/claude-sonnet-4-20250514",
                })).not.toThrow();
            }
            finally {
                await v3NonExperimental.close();
            }
        });
    });
});
