// @browserbasehq/stagehand/dist/cjs/tests/integration/agent-captcha-autosolve.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
const isBrowserbase = (process.env.STAGEHAND_BROWSER_TARGET ?? "local").toLowerCase() ===
    "browserbase";
test_1.test.describe("Agent captcha auto-solve on Browserbase", () => {
    test_1.test.skip(!isBrowserbase, "Requires Browserbase environment");
    let v3;
    let logs;
    test_1.test.beforeEach(async () => {
        logs = [];
        v3 = new v3_js_1.V3((0, v3_config_js_1.getV3TestConfig)({
            env: "BROWSERBASE",
            verbose: 2,
            logger: (line) => {
                logs.push(line);
                console.log(`[${line.category}] ${line.message}`);
            },
            browserbaseSessionCreateParams: {
                browserSettings: {
                    solveCaptchas: true,
                },
            },
        }));
        await v3.init();
        console.log("BB session URL:", v3.browserbaseSessionURL);
    });
    test_1.test.afterEach(async () => {
        await v3?.close?.().catch(() => { });
    });
    (0, test_1.test)("reCAPTCHA v2 auto-solve (Google demo)", async () => {
        test_1.test.setTimeout(180_000);
        const page = v3.context.pages()[0];
        // Google's official reCAPTCHA v2 demo — same URL the stealth team tests.
        // Use domcontentloaded since BB's route interception can delay full load.
        await page.goto("https://www.google.com/recaptcha/api2/demo", {
            waitUntil: "domcontentloaded",
        });
        // Give BB time to intercept the anchor request and solve the captcha
        await new Promise((r) => setTimeout(r, 30_000));
        const agent = v3.agent({
            mode: "dom",
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const result = await agent.execute({
            instruction: 'Click the "Submit" button and report the exact text shown on the result page.',
            maxSteps: 15,
        });
        console.log("reCAPTCHA v2 result:", result.message);
        (0, test_1.expect)(result.completed).toBe(true);
        (0, test_1.expect)(result.message.toLowerCase()).toContain("success");
    });
});
