// @browserbasehq/stagehand/dist/cjs/tests/integration/agent-abort-signal.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
const sdkErrors_js_1 = require("../../lib/v3/types/public/sdkErrors.js");
test_1.test.describe("Stagehand agent abort signal", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3({
            ...v3_config_js_1.v3TestConfig,
            experimental: true,
        });
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await v3?.close?.().catch(() => { });
    });
    (0, test_1.test)("non-streaming: abort signal stops execution and throws AgentAbortError", async () => {
        test_1.test.setTimeout(60000);
        const agent = v3.agent({
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        const controller = new AbortController();
        // Abort after 500ms - should be enough for the LLM to start but not finish
        setTimeout(() => controller.abort(), 500);
        await (0, test_1.expect)(agent.execute({
            instruction: "Describe every visual element on this page in extreme detail. Describe at least 100 different elements.",
            maxSteps: 50,
            signal: controller.signal,
        })).rejects.toThrow(sdkErrors_js_1.AgentAbortError);
    });
    (0, test_1.test)("streaming: abort signal stops stream and rejects result with AgentAbortError", async () => {
        test_1.test.setTimeout(60000);
        const agent = v3.agent({
            stream: true,
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        const controller = new AbortController();
        // Abort after 500ms
        setTimeout(() => controller.abort(), 500);
        const streamResult = await agent.execute({
            instruction: "Describe every visual element on this page in extreme detail. Describe at least 100 different elements.",
            maxSteps: 50,
            signal: controller.signal,
        });
        // Handle both stream consumption and result promise together
        // The result promise will reject with AgentAbortError when aborted
        const consumeStream = async () => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume chunks until stream ends
            }
        };
        // Both should complete - stream ends and result rejects
        const [, resultError] = await Promise.allSettled([
            consumeStream(),
            streamResult.result,
        ]);
        // The result should have rejected with AgentAbortError
        (0, test_1.expect)(resultError.status).toBe("rejected");
        (0, test_1.expect)(resultError.reason).toBeInstanceOf(sdkErrors_js_1.AgentAbortError);
    });
    (0, test_1.test)("non-streaming: already aborted signal throws AgentAbortError immediately", async () => {
        test_1.test.setTimeout(20000);
        const agent = v3.agent({
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        // Create an already aborted controller
        const controller = new AbortController();
        controller.abort();
        await (0, test_1.expect)(agent.execute({
            instruction: "This should not run.",
            maxSteps: 3,
            signal: controller.signal,
        })).rejects.toThrow(sdkErrors_js_1.AgentAbortError);
    });
    (0, test_1.test)("non-streaming: execution completes normally without abort signal", async () => {
        test_1.test.setTimeout(60000);
        const agent = v3.agent({
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        // No signal provided - should complete normally
        const result = await agent.execute({
            instruction: "Describe this page briefly.",
            maxSteps: 3,
        });
        (0, test_1.expect)(result.success).toBe(true);
        (0, test_1.expect)(result.completed).toBe(true);
    });
    (0, test_1.test)("streaming: execution completes normally without abort signal", async () => {
        test_1.test.setTimeout(60000);
        const agent = v3.agent({
            stream: true,
            model: "anthropic/claude-haiku-4-5-20251001",
        });
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        // No signal provided - should complete normally
        const streamResult = await agent.execute({
            instruction: "Describe this page briefly.",
            maxSteps: 3,
        });
        // Consume the stream first
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        for await (const _ of streamResult.textStream) {
            // Just consume
        }
        // Now get the final result
        const result = await streamResult.result;
        (0, test_1.expect)(result.success).toBe(true);
        (0, test_1.expect)(result.completed).toBe(true);
    });
});
