// @browserbasehq/stagehand/dist/cjs/tests/integration/agent-streaming.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
test_1.test.describe("Stagehand agent streaming behavior", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3({
            ...v3_config_js_1.v3TestConfig,
            experimental: true, // Required for streaming
        });
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await v3?.close?.().catch(() => { });
    });
    test_1.test.describe("agent({ stream: true })", () => {
        (0, test_1.test)("AgentStreamResult has textStream as async iterable", async () => {
            test_1.test.setTimeout(60000);
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            // Navigate to a simple page first
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "What is the title of this page? Describe it briefly.",
                maxSteps: 3,
            });
            // Verify it's an AgentStreamResult with streaming capabilities
            (0, test_1.expect)(streamResult).toHaveProperty("textStream");
            (0, test_1.expect)(streamResult).toHaveProperty("result");
            // textStream should be async iterable
            (0, test_1.expect)(typeof streamResult.textStream[Symbol.asyncIterator]).toBe("function");
            // result should be a promise
            (0, test_1.expect)(streamResult.result).toBeInstanceOf(Promise);
        });
        (0, test_1.test)("textStream yields chunks incrementally", async () => {
            test_1.test.setTimeout(60000);
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "Say hello briefly.",
                maxSteps: 3,
            });
            // Collect chunks from the stream
            const chunks = [];
            for await (const chunk of streamResult.textStream) {
                chunks.push(chunk);
            }
            // Should have received at least some chunks (streaming behavior)
            // The exact content depends on the LLM response
            (0, test_1.expect)(Array.isArray(chunks)).toBe(true);
            (0, test_1.expect)(chunks.length).toBeGreaterThan(0);
        });
        (0, test_1.test)("result promise resolves to AgentResult after stream completes", async () => {
            test_1.test.setTimeout(60000);
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "What is this page about? Describe it briefly.",
                maxSteps: 5,
            });
            // Consume the stream first
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            // Now get the final result
            const finalResult = await streamResult.result;
            // Verify it's a proper AgentResult
            (0, test_1.expect)(finalResult).toHaveProperty("success");
            (0, test_1.expect)(finalResult).toHaveProperty("message");
            (0, test_1.expect)(finalResult).toHaveProperty("actions");
            (0, test_1.expect)(finalResult).toHaveProperty("completed");
            (0, test_1.expect)(typeof finalResult.success).toBe("boolean");
            (0, test_1.expect)(typeof finalResult.message).toBe("string");
            (0, test_1.expect)(Array.isArray(finalResult.actions)).toBe(true);
        });
    });
    test_1.test.describe("agent({ stream: false }) or agent()", () => {
        (0, test_1.test)("execute returns AgentResult without streaming properties", async () => {
            test_1.test.setTimeout(60000);
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const result = await agent.execute({
                instruction: "What is this page? Describe it briefly.",
                maxSteps: 3,
            });
            // Should be AgentResult, not AgentStreamResult
            (0, test_1.expect)(result).toHaveProperty("success");
            (0, test_1.expect)(result).toHaveProperty("message");
            (0, test_1.expect)(result).toHaveProperty("actions");
            (0, test_1.expect)(result).toHaveProperty("completed");
            // Should NOT have streaming properties
            (0, test_1.expect)(result).not.toHaveProperty("textStream");
        });
    });
    test_1.test.describe("CUA disables streaming", () => {
        (0, test_1.test)("throws StagehandInvalidArgumentError when cua: true and stream: true", () => {
            (0, test_1.expect)(() => {
                v3.agent({
                    cua: true,
                    stream: true,
                    model: "anthropic/claude-haiku-4-5-20251001",
                });
            }).toThrow("streaming is not supported with CUA");
        });
        (0, test_1.test)("allows cua: true without stream", () => {
            // Should not throw
            const agent = v3.agent({
                cua: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            (0, test_1.expect)(agent).toHaveProperty("execute");
        });
        (0, test_1.test)("allows stream: true without cua", () => {
            // Should not throw
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            (0, test_1.expect)(agent).toHaveProperty("execute");
        });
    });
});
