// @browserbasehq/stagehand/dist/cjs/tests/integration/agent-callbacks.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
const sdkErrors_js_1 = require("../../lib/v3/types/public/sdkErrors.js");
test_1.test.describe("Stagehand agent callbacks behavior", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3({
            ...v3_config_js_1.v3TestConfig,
            experimental: true, // Required for callbacks and streaming
        });
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await v3?.close?.().catch(() => { });
    });
    test_1.test.describe("Non-streaming callbacks (stream: false)", () => {
        (0, test_1.test)("onStepFinish callback is called for each step", async () => {
            test_1.test.setTimeout(60000);
            const stepFinishEvents = [];
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            await agent.execute({
                instruction: "What is the title of this page? Mark the task as complete after answering.",
                maxSteps: 5,
                callbacks: {
                    onStepFinish: async (event) => {
                        stepFinishEvents.push(event);
                    },
                },
            });
            // Should have at least one step finish event
            (0, test_1.expect)(stepFinishEvents.length).toBeGreaterThan(0);
            // Each event should have expected properties
            for (const event of stepFinishEvents) {
                (0, test_1.expect)(event).toHaveProperty("finishReason");
                (0, test_1.expect)(event).toHaveProperty("text");
            }
        });
        (0, test_1.test)("prepareStep callback is called before each step", async () => {
            test_1.test.setTimeout(60000);
            let prepareStepCallCount = 0;
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            await agent.execute({
                instruction: "Simply describe the page briefly.",
                maxSteps: 3,
                callbacks: {
                    prepareStep: async (stepContext) => {
                        prepareStepCallCount++;
                        return stepContext;
                    },
                },
            });
            // prepareStep should have been called at least once
            (0, test_1.expect)(prepareStepCallCount).toBeGreaterThan(0);
        });
        (0, test_1.test)("callbacks receive tool call information", async () => {
            test_1.test.setTimeout(60000);
            const toolCalls = [];
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            await agent.execute({
                instruction: "Take a screenshot and describe what you see briefly. Then mark the task as complete.",
                maxSteps: 3,
                callbacks: {
                    onStepFinish: async (event) => {
                        if (event.toolCalls) {
                            for (const tc of event.toolCalls) {
                                toolCalls.push({
                                    toolName: tc.toolName,
                                    input: tc.input,
                                });
                            }
                        }
                    },
                },
            });
            // Should have captured at least one tool call (e.g. screenshot)
            (0, test_1.expect)(toolCalls.length).toBeGreaterThan(0);
            (0, test_1.expect)(toolCalls.some((tc) => tc.toolName === "screenshot" || tc.toolName === "ariaTree")).toBe(true);
        });
    });
    test_1.test.describe("Streaming callbacks (stream: true)", () => {
        (0, test_1.test)("onStepFinish callback is called for each step in stream mode", async () => {
            test_1.test.setTimeout(60000);
            const stepFinishEvents = [];
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "What is this page? Describe it briefly.",
                maxSteps: 5,
                callbacks: {
                    onStepFinish: async (event) => {
                        stepFinishEvents.push(event);
                    },
                },
            });
            // Consume the stream
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            // Wait for result to complete
            await streamResult.result;
            // Should have at least one step finish event
            (0, test_1.expect)(stepFinishEvents.length).toBeGreaterThan(0);
        });
        (0, test_1.test)("onChunk callback is called for each chunk", async () => {
            test_1.test.setTimeout(60000);
            let chunkCount = 0;
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "Say hello briefly and describe the page.",
                maxSteps: 3,
                callbacks: {
                    onChunk: async () => {
                        chunkCount++;
                    },
                },
            });
            // Consume the stream
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            await streamResult.result;
            // Should have received chunks
            (0, test_1.expect)(chunkCount).toBeGreaterThan(0);
        });
        (0, test_1.test)("onFinish callback is called when stream completes", async () => {
            test_1.test.setTimeout(60000);
            let finishCalled = false;
            let finishEvent = null;
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "Simply describe the page briefly.",
                maxSteps: 3,
                callbacks: {
                    onFinish: (event) => {
                        finishCalled = true;
                        finishEvent = event;
                    },
                },
            });
            // Consume the stream
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            await streamResult.result;
            // onFinish should have been called
            (0, test_1.expect)(finishCalled).toBe(true);
            (0, test_1.expect)(finishEvent).not.toBeNull();
        });
        (0, test_1.test)("prepareStep callback works in stream mode", async () => {
            test_1.test.setTimeout(60000);
            let prepareStepCallCount = 0;
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "Simply describe the page briefly.",
                maxSteps: 3,
                callbacks: {
                    prepareStep: async (stepContext) => {
                        prepareStepCallCount++;
                        return stepContext;
                    },
                },
            });
            // Consume the stream
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            await streamResult.result;
            // prepareStep should have been called at least once
            (0, test_1.expect)(prepareStepCallCount).toBeGreaterThan(0);
        });
    });
    test_1.test.describe("Streaming-only callbacks runtime validation", () => {
        (0, test_1.test)("throws StreamingCallbacksInNonStreamingModeError when onChunk is used", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onChunk: (() => { }),
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StreamingCallbacksInNonStreamingModeError);
                (0, test_1.expect)(error.invalidCallbacks).toEqual(["onChunk"]);
            }
        });
        (0, test_1.test)("throws StreamingCallbacksInNonStreamingModeError when onFinish is used", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onFinish: (() => { }),
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StreamingCallbacksInNonStreamingModeError);
                (0, test_1.expect)(error.invalidCallbacks).toEqual(["onFinish"]);
            }
        });
        (0, test_1.test)("throws StreamingCallbacksInNonStreamingModeError when onError is used", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onError: (() => { }),
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StreamingCallbacksInNonStreamingModeError);
                (0, test_1.expect)(error.invalidCallbacks).toEqual(["onError"]);
            }
        });
        (0, test_1.test)("throws StreamingCallbacksInNonStreamingModeError when onAbort is used", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onAbort: (() => { }),
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StreamingCallbacksInNonStreamingModeError);
                (0, test_1.expect)(error.invalidCallbacks).toEqual(["onAbort"]);
            }
        });
        (0, test_1.test)("error includes all invalid callbacks when multiple are used", async () => {
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            try {
                await agent.execute({
                    instruction: "test",
                    callbacks: {
                        onChunk: (() => { }),
                        onFinish: (() => { }),
                    },
                });
                throw new Error("Expected error to be thrown");
            }
            catch (error) {
                (0, test_1.expect)(error).toBeInstanceOf(sdkErrors_js_1.StreamingCallbacksInNonStreamingModeError);
                (0, test_1.expect)(error.invalidCallbacks).toEqual(["onChunk", "onFinish"]);
            }
        });
    });
    test_1.test.describe("Combined callbacks", () => {
        (0, test_1.test)("multiple callbacks can be used together", async () => {
            test_1.test.setTimeout(60000);
            let prepareStepCount = 0;
            let stepFinishCount = 0;
            const agent = v3.agent({
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            await agent.execute({
                instruction: "Simply describe the page briefly.",
                maxSteps: 3,
                callbacks: {
                    prepareStep: async (stepContext) => {
                        prepareStepCount++;
                        return stepContext;
                    },
                    onStepFinish: async () => {
                        stepFinishCount++;
                    },
                },
            });
            // Both callbacks should have been called
            (0, test_1.expect)(prepareStepCount).toBeGreaterThan(0);
            (0, test_1.expect)(stepFinishCount).toBeGreaterThan(0);
        });
        (0, test_1.test)("streaming with multiple callbacks", async () => {
            test_1.test.setTimeout(60000);
            let prepareStepCount = 0;
            let stepFinishCount = 0;
            let chunkCount = 0;
            let finishCalled = false;
            const agent = v3.agent({
                stream: true,
                model: "anthropic/claude-haiku-4-5-20251001",
            });
            const page = v3.context.pages()[0];
            await page.goto("https://example.com");
            const streamResult = await agent.execute({
                instruction: "Say hello briefly and describe the page.",
                maxSteps: 3,
                callbacks: {
                    prepareStep: async (stepContext) => {
                        prepareStepCount++;
                        return stepContext;
                    },
                    onStepFinish: async () => {
                        stepFinishCount++;
                    },
                    onChunk: async () => {
                        chunkCount++;
                    },
                    onFinish: () => {
                        finishCalled = true;
                    },
                },
            });
            // Consume the stream
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for await (const _ of streamResult.textStream) {
                // Just consume
            }
            await streamResult.result;
            // All callbacks should have been called
            (0, test_1.expect)(prepareStepCount).toBeGreaterThan(0);
            (0, test_1.expect)(stepFinishCount).toBeGreaterThan(0);
            (0, test_1.expect)(chunkCount).toBeGreaterThan(0);
            (0, test_1.expect)(finishCalled).toBe(true);
        });
    });
});
