// @browserbasehq/stagehand/dist/cjs/tests/integration/cookies.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_dynamic_config_js_1 = require("./v3.dynamic.config.js");
const testUtils_js_1 = require("./testUtils.js");
const BASE_URL = "https://browserbase.github.io/stagehand-eval-sites/sites/example/";
test_1.test.describe("cookies", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3(v3_dynamic_config_js_1.v3DynamicTestConfig);
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await (0, testUtils_js_1.closeV3)(v3);
    });
    (0, test_1.test)("addCookies sets a cookie visible to the page", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        (0, test_1.expect)(page).toBeDefined();
        await page.goto(BASE_URL);
        const name = `stagehand_cookie_${Date.now()}`;
        await ctx.addCookies([
            {
                name,
                value: "1",
                url: BASE_URL,
                httpOnly: false,
            },
        ]);
        await page.reload();
        const cookieString = await page.evaluate(() => document.cookie);
        (0, test_1.expect)(cookieString).toContain(`${name}=1`);
        const cookies = await ctx.cookies(BASE_URL);
        (0, test_1.expect)(cookies.some((c) => c.name === name && c.value === "1")).toBe(true);
    });
    (0, test_1.test)("cookies() with no URL returns all cookies", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        const name = `stagehand_all_${Date.now()}`;
        await ctx.addCookies([
            { name, value: "all", url: BASE_URL, httpOnly: false },
        ]);
        const all = await ctx.cookies();
        (0, test_1.expect)(all.some((c) => c.name === name)).toBe(true);
    });
    (0, test_1.test)("clearCookies() removes all cookies", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        await ctx.addCookies([
            { name: "to_clear_a", value: "1", url: BASE_URL, httpOnly: false },
            { name: "to_clear_b", value: "2", url: BASE_URL, httpOnly: false },
        ]);
        // Verify cookies were set
        let cookies = await ctx.cookies(BASE_URL);
        (0, test_1.expect)(cookies.some((c) => c.name === "to_clear_a")).toBe(true);
        (0, test_1.expect)(cookies.some((c) => c.name === "to_clear_b")).toBe(true);
        await ctx.clearCookies();
        cookies = await ctx.cookies(BASE_URL);
        (0, test_1.expect)(cookies.some((c) => c.name === "to_clear_a")).toBe(false);
        (0, test_1.expect)(cookies.some((c) => c.name === "to_clear_b")).toBe(false);
    });
    (0, test_1.test)("clearCookies() with name filter removes only matching cookies", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        await ctx.addCookies([
            { name: "keep_me", value: "1", url: BASE_URL, httpOnly: false },
            { name: "remove_me", value: "2", url: BASE_URL, httpOnly: false },
        ]);
        await ctx.clearCookies({ name: "remove_me" });
        const cookies = await ctx.cookies(BASE_URL);
        (0, test_1.expect)(cookies.some((c) => c.name === "keep_me")).toBe(true);
        (0, test_1.expect)(cookies.some((c) => c.name === "remove_me")).toBe(false);
    });
    (0, test_1.test)("clearCookies() with regex filter removes matching cookies", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        await ctx.addCookies([
            { name: "_ga_ABC", value: "1", url: BASE_URL, httpOnly: false },
            { name: "_ga_DEF", value: "2", url: BASE_URL, httpOnly: false },
            { name: "session", value: "3", url: BASE_URL, httpOnly: false },
        ]);
        await ctx.clearCookies({ name: /^_ga/ });
        const cookies = await ctx.cookies(BASE_URL);
        (0, test_1.expect)(cookies.some((c) => c.name === "session")).toBe(true);
        (0, test_1.expect)(cookies.some((c) => c.name === "_ga_ABC")).toBe(false);
        (0, test_1.expect)(cookies.some((c) => c.name === "_ga_DEF")).toBe(false);
    });
    (0, test_1.test)("cookies are visible from a second page on the same domain", async () => {
        const ctx = v3.context;
        const page1 = ctx.pages()[0];
        await page1.goto(BASE_URL);
        const name = `stagehand_multi_${Date.now()}`;
        await ctx.addCookies([
            { name, value: "shared", url: BASE_URL, httpOnly: false },
        ]);
        const page2 = await ctx.newPage();
        await page2.goto(BASE_URL);
        const cookieString = await page2.evaluate(() => document.cookie);
        (0, test_1.expect)(cookieString).toContain(`${name}=shared`);
    });
    (0, test_1.test)("cookies persist across navigation to a different path", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        const name = `stagehand_nav_${Date.now()}`;
        await ctx.addCookies([
            {
                name,
                value: "persisted",
                domain: "browserbase.github.io",
                path: "/",
                httpOnly: false,
            },
        ]);
        // Navigate to a different path on the same domain
        await page.goto("https://browserbase.github.io/stagehand-eval-sites/");
        const cookieString = await page.evaluate(() => document.cookie);
        (0, test_1.expect)(cookieString).toContain(`${name}=persisted`);
    });
    (0, test_1.test)("httpOnly cookie is hidden from document.cookie but returned by cookies()", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        const name = `stagehand_http_${Date.now()}`;
        await ctx.addCookies([
            { name, value: "secret", url: BASE_URL, httpOnly: true },
        ]);
        await page.reload();
        // document.cookie must NOT include httpOnly cookies
        const cookieString = await page.evaluate(() => document.cookie);
        (0, test_1.expect)(cookieString).not.toContain(name);
        // But the context API should still return it
        const cookies = await ctx.cookies(BASE_URL);
        const match = cookies.find((c) => c.name === name);
        (0, test_1.expect)(match).toBeDefined();
        (0, test_1.expect)(match.value).toBe("secret");
        (0, test_1.expect)(match.httpOnly).toBe(true);
    });
    (0, test_1.test)("cookies() returns correct shape for a fully-specified cookie", async () => {
        const ctx = v3.context;
        const page = ctx.pages()[0];
        await page.goto(BASE_URL);
        const name = `stagehand_shape_${Date.now()}`;
        const expires = Math.floor(Date.now() / 1000) + 3600; // 1 hour from now
        await ctx.addCookies([
            {
                name,
                value: "full",
                domain: "browserbase.github.io",
                path: "/",
                expires,
                httpOnly: true,
                secure: true,
                sameSite: "Lax",
            },
        ]);
        const cookies = await ctx.cookies(BASE_URL);
        const match = cookies.find((c) => c.name === name);
        (0, test_1.expect)(match).toBeDefined();
        // Validate every field on the returned Cookie object
        (0, test_1.expect)(match.value).toBe("full");
        (0, test_1.expect)(match.domain).toMatch(/browserbase\.github\.io/);
        (0, test_1.expect)(match.path).toBe("/");
        (0, test_1.expect)(match.expires).toBeGreaterThan(0);
        (0, test_1.expect)(match.httpOnly).toBe(true);
        (0, test_1.expect)(match.secure).toBe(true);
        (0, test_1.expect)(match.sameSite).toBe("Lax");
        // Ensure no extra fields leak through from CDP
        const keys = Object.keys(match);
        (0, test_1.expect)(keys.sort()).toEqual([
            "name",
            "value",
            "domain",
            "path",
            "expires",
            "httpOnly",
            "secure",
            "sameSite",
        ].sort());
    });
});
