// @browserbasehq/stagehand/dist/cjs/tests/integration/page-send-cdp.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_config_js_1 = require("./v3.config.js");
test_1.test.describe("Page sendCDP method", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3(v3_config_js_1.v3TestConfig);
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await v3?.close?.().catch(() => { });
    });
    (0, test_1.test)("sends CDP commands and requires domain to be enabled first", async () => {
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        // Try to add a virtual authenticator without enabling WebAuthn first
        // This should fail because the domain needs to be enabled
        await (0, test_1.expect)(page.sendCDP("WebAuthn.addVirtualAuthenticator", {
            options: {
                protocol: "ctap2",
                transport: "usb",
                hasResidentKey: false,
                hasUserVerification: false,
                isUserVerified: false,
            },
        })).rejects.toThrow();
        // Enable the WebAuthn domain
        await page.sendCDP("WebAuthn.enable");
        // Now adding a virtual authenticator should succeed
        const result = await page.sendCDP("WebAuthn.addVirtualAuthenticator", {
            options: {
                protocol: "ctap2",
                transport: "usb",
                hasResidentKey: false,
                hasUserVerification: false,
                isUserVerified: false,
            },
        });
        // Verify we got an authenticator ID back
        (0, test_1.expect)(result).toHaveProperty("authenticatorId");
        (0, test_1.expect)(typeof result.authenticatorId).toBe("string");
        (0, test_1.expect)(result.authenticatorId.length).toBeGreaterThan(0);
    });
});
