// @browserbasehq/stagehand/dist/cjs/tests/integration/shadow-iframe-oopif.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const puppeteer_core_1 = __importDefault(require("puppeteer-core"));
const playwright_core_1 = require("playwright-core");
const patchright_core_1 = require("patchright-core");
const v3_dynamic_config_js_1 = require("./v3.dynamic.config.js");
const testUtils_js_1 = require("./testUtils.js");
async function runCase(v3, c, framework) {
    let cleanup = null;
    // Acquire the correct page for the requested framework
    let page;
    switch (framework) {
        case "v3": {
            const v3Page = v3.context.pages()[0];
            await v3Page.goto(c.url, { waitUntil: "networkidle" });
            page = v3Page;
            break;
        }
        case "puppeteer": {
            const browser = await puppeteer_core_1.default.connect({
                browserWSEndpoint: v3.connectURL(),
                defaultViewport: null,
            });
            const pages = await browser.pages();
            const puppeteerPage = pages[0];
            await puppeteerPage.goto(c.url, { waitUntil: "networkidle0" });
            page = puppeteerPage;
            cleanup = async () => {
                try {
                    await browser.close();
                }
                catch {
                    //
                }
            };
            break;
        }
        case "playwright": {
            const pwBrowser = await playwright_core_1.chromium.connectOverCDP(v3.connectURL());
            const pwContext = pwBrowser.contexts()[0];
            const pwPage = pwContext.pages()[0];
            await pwPage.goto(c.url, { waitUntil: "networkidle" });
            page = pwPage;
            cleanup = async () => {
                try {
                    await pwBrowser.close();
                }
                catch {
                    // ignore
                }
            };
            break;
        }
        case "patchright": {
            const prBrowser = await patchright_core_1.chromium.connectOverCDP(v3.connectURL());
            const prContext = prBrowser.contexts()[0];
            const prPage = prContext.pages()[0];
            await prPage.goto(c.url, { waitUntil: "networkidle" });
            page = prPage;
            cleanup = async () => {
                try {
                    await prBrowser.close();
                }
                catch {
                    // ignore
                }
            };
            break;
        }
    }
    try {
        if (!page)
            throw new Error("Missing page for selected framework");
        await v3.act(c.action, { page });
        // Post-action extraction; verify expected text appears
        const extraction = await v3.extract({ page });
        const text = extraction.pageText ?? "";
        for (const s of c.expectedSubstrings) {
            (0, test_1.expect)(text.includes(s), `expected pageText to include substring: ${s}`).toBeTruthy();
        }
    }
    finally {
        await cleanup?.();
    }
}
const cases = [
    {
        title: "Closed shadow root inside OOPIF",
        url: "https://browserbase.github.io/stagehand-eval-sites/sites/closed-shadow-root-in-oopif/",
        action: {
            selector: "xpath=/html/body/main/section/iframe/html/body/shadow-demo//div/button",
            method: "click",
            arguments: [""],
            description: "click button inside closed shadow root in OOPIF",
        },
        expectedSubstrings: ["button successfully clicked"],
    },
    {
        title: "Open shadow root inside OOPIF",
        url: "https://browserbase.github.io/stagehand-eval-sites/sites/open-shadow-root-in-oopif/",
        action: {
            selector: "xpath=/html/body/main/section/iframe/html/body/shadow-demo//div/button",
            method: "click",
            arguments: [""],
            description: "",
        },
        expectedSubstrings: ["button successfully clicked"],
    },
    {
        title: "OOPIF inside open shadow root",
        url: "https://browserbase.github.io/stagehand-eval-sites/sites/oopif-in-open-shadow-dom/",
        action: {
            selector: "xpath=/html/body/shadow-host//section/iframe/html/body/main/section[1]/form/div/div[1]/input",
            method: "fill",
            arguments: ["nunya"],
            description: "",
        },
        expectedSubstrings: ["nunya"],
    },
    {
        title: "OOPIF inside closed shadow root",
        url: "https://browserbase.github.io/stagehand-eval-sites/sites/oopif-in-closed-shadow-dom/",
        action: {
            selector: "xpath=/html/body/shadow-host//section/iframe/html/body/main/section[1]/form/div/div[1]/input",
            method: "fill",
            arguments: ["nunya"],
            description: "fill input inside OOPIF",
        },
        expectedSubstrings: ["nunya"],
    },
];
test_1.test.describe
    .parallel("Stagehand v3: shadow <-> iframe OOPIF scenarios", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3(v3_dynamic_config_js_1.v3DynamicTestConfig);
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await (0, testUtils_js_1.closeV3)(v3);
    });
    const frameworks = [
        "v3",
        "playwright",
        "puppeteer",
        "patchright",
    ];
    for (const fw of frameworks) {
        for (const c of cases) {
            (0, test_1.test)(`[${fw}] ${c.title}`, async () => {
                await runCase(v3, c, fw);
            });
        }
    }
});
