// @browserbasehq/stagehand/dist/cjs/tests/integration/timeouts.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_dynamic_config_js_1 = require("./v3.dynamic.config.js");
const zod_1 = require("zod");
const testUtils_js_1 = require("./testUtils.js");
const ai_1 = require("ai");
function createToolTimeoutTestLlmClient(toolName, toolInput) {
    const usage = {
        prompt_tokens: 0,
        completion_tokens: 0,
        reasoning_tokens: 0,
        cached_input_tokens: 0,
        total_tokens: 0,
    };
    let generateCallCount = 0;
    const model = {
        provider: "mock",
        modelId: "mock/tool-timeout-test",
        specificationVersion: "v2",
        supportedUrls: {},
        doGenerate: async () => {
            generateCallCount += 1;
            if (generateCallCount === 1) {
                return {
                    content: [
                        {
                            type: "tool-call",
                            toolCallId: "tool-1",
                            toolName,
                            input: JSON.stringify(toolInput),
                        },
                    ],
                    finishReason: "tool-calls",
                    usage: { inputTokens: 0, outputTokens: 0, totalTokens: 0 },
                    warnings: [],
                };
            }
            return {
                content: [
                    {
                        type: "tool-call",
                        toolCallId: "done-1",
                        toolName: "done",
                        input: JSON.stringify({ reasoning: "done", taskComplete: true }),
                    },
                ],
                finishReason: "tool-calls",
                usage: { inputTokens: 0, outputTokens: 0, totalTokens: 0 },
                warnings: [],
            };
        },
        doStream: async () => {
            throw new Error("doStream not implemented in timeout test model");
        },
    };
    const llm = {
        type: "openai",
        modelName: "openai/gpt-4.1-mini",
        hasVision: false,
        clientOptions: {},
        model,
        getLanguageModel: () => model,
        generateText: ai_1.generateText,
        createChatCompletion: async (options) => {
            const responseModelName = options?.options?.response_model?.name;
            if (responseModelName === "act") {
                return {
                    data: {
                        elementId: "1-0",
                        description: "click body",
                        method: "click",
                        arguments: [],
                        twoStep: false,
                    },
                    usage,
                };
            }
            if (responseModelName === "Observation") {
                return { data: { elements: [] }, usage };
            }
            if (responseModelName === "Extraction") {
                return { data: {}, usage };
            }
            if (responseModelName === "Metadata") {
                return { data: { completed: true, progress: "" }, usage };
            }
            return { data: {}, usage };
        },
    };
    return llm;
}
function findToolOutput(stepEvents, toolName) {
    for (const event of stepEvents) {
        if (!event.toolCalls || !event.toolResults)
            continue;
        const toolIndex = event.toolCalls.findIndex((tc) => tc.toolName === toolName);
        if (toolIndex !== -1) {
            return event.toolResults[toolIndex]?.output;
        }
    }
    return undefined;
}
async function runAgentToolTimeoutScenario(toolName, toolInput, options) {
    const llmClient = createToolTimeoutTestLlmClient(toolName, toolInput);
    const stepEvents = [];
    const v3 = new v3_js_1.V3({
        ...v3_dynamic_config_js_1.v3DynamicTestConfig,
        experimental: true,
        llmClient,
    });
    await v3.init();
    try {
        const page = v3.context.pages()[0];
        await page.goto("https://example.com");
        const agent = v3.agent({
            ...(options?.mode ? { mode: options.mode } : {}),
        });
        await agent.execute({
            instruction: `Use ${toolName} and then finish`,
            maxSteps: 2,
            toolTimeout: 1,
            callbacks: {
                onStepFinish: (event) => {
                    stepEvents.push({
                        toolCalls: event.toolCalls?.map((tc) => ({
                            toolName: tc.toolName,
                        })),
                        toolResults: event.toolResults?.map((tr) => ({
                            output: tr.output,
                        })),
                    });
                },
            },
        });
        const toolOutput = findToolOutput(stepEvents, toolName);
        if (!toolOutput) {
            throw new Error(`No tool output captured for ${toolName}`);
        }
        return { toolOutput };
    }
    finally {
        await (0, testUtils_js_1.closeV3)(v3);
    }
}
test_1.test.describe("V3 hard timeouts", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3(v3_dynamic_config_js_1.v3DynamicTestConfig);
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await (0, testUtils_js_1.closeV3)(v3);
    });
    (0, test_1.test)("observe() enforces timeoutMs", async () => {
        // Tiny timeout to force the race to hit the timeout branch
        await (0, test_1.expect)(v3.observe("find something", { timeout: 5 })).rejects.toThrow(/timed out/i);
    });
    (0, test_1.test)("extract() enforces timeoutMs", async () => {
        const schema = zod_1.z.object({ title: zod_1.z.string().optional() });
        await (0, test_1.expect)(v3.extract("Extract title", schema, { timeout: 5 })).rejects.toThrow(/timed out/i);
    });
    (0, test_1.test)("act() enforces timeoutMs", async () => {
        await (0, test_1.expect)(v3.act("do nothing", { timeout: 5 })).rejects.toThrow(/timed out/i);
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for act tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("act", {
            action: "click somewhere",
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for extract tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("extract", {
            instruction: "extract the page title",
            schema: { type: "object", properties: { title: { type: "string" } } },
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for fillForm tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("fillForm", {
            fields: [{ action: "type hello into name" }],
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for ariaTree", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("ariaTree", {});
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for goto tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("goto", {
            url: "https://example.com/slow",
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for navback tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("navback", {
            reasoningText: "going back",
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for screenshot tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("screenshot", {});
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for scroll tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("scroll", {
            direction: "down",
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for keys tool", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("keys", {
            method: "press",
            value: "Enter",
        });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for click tool (hybrid)", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("click", { describe: "click element", coordinates: [100, 100] }, { mode: "hybrid" });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for type tool (hybrid)", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("type", {
            describe: "type into field",
            text: "hello",
            coordinates: [100, 100],
        }, { mode: "hybrid" });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for dragAndDrop tool (hybrid)", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("dragAndDrop", {
            describe: "drag element",
            startCoordinates: [100, 100],
            endCoordinates: [200, 200],
        }, { mode: "hybrid" });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for clickAndHold tool (hybrid)", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("clickAndHold", {
            describe: "hold element",
            coordinates: [100, 100],
            duration: 1000,
        }, { mode: "hybrid" });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
    (0, test_1.test)("agent toolTimeout enforces timeout for fillFormVision tool (hybrid)", async () => {
        const { toolOutput } = await runAgentToolTimeoutScenario("fillFormVision", {
            fields: [
                {
                    action: "type hello into name",
                    value: "hello",
                    coordinates: { x: 100, y: 100 },
                },
                {
                    action: "type world into email",
                    value: "world",
                    coordinates: { x: 100, y: 200 },
                },
            ],
        }, { mode: "hybrid" });
        const output = toolOutput;
        (0, test_1.expect)(output.success).toBe(false);
        (0, test_1.expect)(output.error).toContain("TimeoutError");
        (0, test_1.expect)(output.error).toContain("1ms");
    });
});
