// @browserbasehq/stagehand/dist/cjs/tests/integration/v3.dynamic.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.v3DynamicTestConfig = void 0;
exports.getV3DynamicTestConfig = getV3DynamicTestConfig;
const browserTarget = (process.env.STAGEHAND_BROWSER_TARGET ?? "local").toLowerCase();
const isBrowserbase = browserTarget === "browserbase";
const browserbaseRegionRaw = process.env.BROWSERBASE_REGION;
const browserbaseRegion = [
    "us-west-2",
    "us-east-1",
    "eu-central-1",
    "ap-southeast-1",
].includes(browserbaseRegionRaw)
    ? browserbaseRegionRaw
    : undefined;
const baseConfig = {
    verbose: 0,
    disablePino: true,
    logger: (line) => console.log(line),
    disableAPI: true,
};
exports.v3DynamicTestConfig = isBrowserbase
    ? {
        ...baseConfig,
        env: "BROWSERBASE",
        apiKey: process.env.BROWSERBASE_API_KEY,
        projectId: process.env.BROWSERBASE_PROJECT_ID,
        disableAPI: true,
        selfHeal: false,
        ...(browserbaseRegion
            ? { browserbaseSessionCreateParams: { region: browserbaseRegion } }
            : {}),
    }
    : {
        ...baseConfig,
        env: "LOCAL",
        localBrowserLaunchOptions: {
            executablePath: process.env.CHROME_PATH,
            args: process.env.CI ? ["--no-sandbox"] : undefined,
            headless: true,
            viewport: { width: 1288, height: 711 },
        },
    };
function getV3DynamicTestConfig(overrides = {}) {
    return { ...exports.v3DynamicTestConfig, ...overrides };
}
exports.default = getV3DynamicTestConfig;
