// @browserbasehq/stagehand/dist/cjs/tests/integration/xpath-for-location-deep.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const v3_js_1 = require("../../lib/v3/v3.js");
const v3_dynamic_config_js_1 = require("./v3.dynamic.config.js");
const index_js_1 = require("../../lib/v3/understudy/a11y/snapshot/index.js");
const executionContextRegistry_js_1 = require("../../lib/v3/understudy/executionContextRegistry.js");
const testUtils_js_1 = require("./testUtils.js");
test_1.test.describe("resolveNodeForLocationDeep", () => {
    let v3;
    test_1.test.beforeEach(async () => {
        v3 = new v3_js_1.V3(v3_dynamic_config_js_1.v3DynamicTestConfig);
        await v3.init();
    });
    test_1.test.afterEach(async () => {
        await (0, testUtils_js_1.closeV3)(v3);
    });
    (0, test_1.test)("click resolves inside same-process iframe and returns absolute XPath", async () => {
        const page = await v3.context.awaitActivePage();
        // Set consistent viewport size to ensure stable rendering across environments
        await page.setViewportSize(1280, 720);
        await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/iframe-hn/", { waitUntil: "networkidle" });
        await page.waitForSelector("section iframe", {
            state: "attached",
            timeout: 10000,
        });
        const frame = await page.frameLocator("section iframe").resolveFrame();
        await executionContextRegistry_js_1.executionContexts.waitForMainWorld(frame.session, frame.frameId, 5000);
        // scroll to the bottom of the page
        await page.evaluate(() => {
            window.scrollTo(0, document.body.scrollHeight);
        });
        // scroll to the bottom of the iframe
        await frame.evaluate(() => {
            window.scrollTo(0, document.body.scrollHeight);
        });
        // Wait a bit for the iframe content to settle after scrolling
        await new Promise((resolve) => setTimeout(resolve, 500));
        // Get the iframe's position in the main page
        const iframeOffset = await page.evaluate(() => {
            const iframe = document.querySelector("section iframe");
            if (!iframe)
                return null;
            const rect = iframe.getBoundingClientRect();
            return {
                left: rect.left,
                top: rect.top,
            };
        });
        // Get the link's position within the iframe
        const linkOffsetInFrame = await frame.evaluate(() => {
            // Find the 88th row, 3rd column link (the one we're testing)
            const table = document.querySelector("center > table > tbody > tr:nth-child(3) > td > table");
            if (!table)
                return null;
            const row88 = table.querySelector("tbody > tr:nth-child(88)");
            if (!row88)
                return null;
            const cell3 = row88.querySelector("td:nth-child(3)");
            if (!cell3)
                return null;
            const link = cell3.querySelector("span > a");
            if (!link)
                return null;
            const rect = link.getBoundingClientRect();
            // Return center coordinates of the link relative to iframe
            return {
                x: rect.left + rect.width / 2,
                y: rect.top + rect.height / 2,
            };
        });
        // Combine iframe offset and link offset to get page-level coordinates
        // Fallback to hardcoded coordinates if element not found (shouldn't happen)
        const x = iframeOffset && linkOffsetInFrame
            ? iframeOffset.left + linkOffsetInFrame.x
            : 356;
        const y = iframeOffset && linkOffsetInFrame
            ? iframeOffset.top + linkOffsetInFrame.y
            : 503;
        const result = await (0, index_js_1.resolveXpathForLocation)(page, x, y);
        console.log("=== Coordinates used:", { x, y });
        console.log("=== Result:", result);
        const xpath = result.absoluteXPath;
        (0, test_1.expect)(xpath).toBe("/html[1]/body[1]/main[1]/section[3]/iframe[1]/html[1]/body[1]/center[1]/table[1]/tbody[1]/tr[3]/td[1]/table[1]/tbody[1]/tr[88]/td[3]/span[1]/a[1]");
    });
});
