// @browserbasehq/stagehand/dist/cjs/lib/v3/mcp/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveTools = void 0;
const utils_js_1 = require("../../utils.js");
const connection_js_1 = require("./connection.js");
const resolveTools = async (clients, userTools) => {
    const tools = { ...userTools };
    for (const client of clients) {
        let clientInstance;
        if (typeof client === "string") {
            clientInstance = await (0, connection_js_1.connectToMCPServer)(client);
        }
        else {
            clientInstance = client;
        }
        let nextCursor = undefined;
        do {
            const clientTools = await clientInstance.listTools({
                cursor: nextCursor,
            });
            for (const tool of clientTools.tools) {
                tools[tool.name] = {
                    description: tool.description,
                    inputSchema: (0, utils_js_1.jsonSchemaToZod)(tool.inputSchema),
                    execute: async (input) => {
                        const result = await clientInstance.callTool({
                            name: tool.name,
                            arguments: input,
                        });
                        return result;
                    },
                };
            }
            nextCursor = clientTools.nextCursor;
        } while (nextCursor);
    }
    return tools;
};
exports.resolveTools = resolveTools;
