// @browserbasehq/stagehand/dist/cjs/lib/v3/shutdown/supervisorClient.js
"use strict";
/**
 * Parent-side helper for spawning the shutdown supervisor process.
 *
 * The supervisor runs out-of-process and watches a lifeline pipe. If the parent
 * dies, the supervisor performs best-effort cleanup (Chrome kill or Browserbase
 * session release) when keepAlive is false.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startShutdownSupervisor = startShutdownSupervisor;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const node_child_process_1 = require("node:child_process");
const node_module_1 = require("node:module");
const shutdownErrors_js_1 = require("../types/private/shutdownErrors.js");
const runtimePaths_js_1 = require("../runtimePaths.js");
const moduleFilename = (0, runtimePaths_js_1.getCurrentFilePath)();
const moduleDir = node_path_1.default.dirname(moduleFilename);
const nodeRequire = (0, node_module_1.createRequire)(moduleFilename);
const isSeaRuntime = () => {
    try {
        const sea = nodeRequire("node:sea");
        return Boolean(sea.isSea?.());
    }
    catch {
        return false;
    }
};
// SEA: re-exec current binary with supervisor args.
// Non-SEA: execute Stagehand CLI entrypoint with supervisor args.
const resolveCliPath = () => `${moduleDir}/../cli.js`;
const resolveSupervisorCommand = (config) => {
    const baseArgs = ["--supervisor", serializeConfigArg(config)];
    if (isSeaRuntime()) {
        return { command: process.execPath, args: baseArgs };
    }
    const cliPath = resolveCliPath();
    if (!node_fs_1.default.existsSync(cliPath))
        return null;
    const needsTsxLoader = node_fs_1.default.existsSync(`${moduleDir}/supervisor.ts`) &&
        !node_fs_1.default.existsSync(`${moduleDir}/supervisor.js`);
    return {
        command: process.execPath,
        args: needsTsxLoader
            ? ["--import", "tsx", cliPath, ...baseArgs]
            : [cliPath, ...baseArgs],
    };
};
// Single JSON arg keeps supervisor bootstrap parsing tiny and versionable.
const serializeConfigArg = (config) => `--supervisor-config=${JSON.stringify({
    ...config,
    parentPid: process.pid,
})}`;
/**
 * Start a supervisor process for crash cleanup. Returns a handle that can
 * stop the supervisor during a normal shutdown.
 */
function startShutdownSupervisor(config, opts) {
    const resolved = resolveSupervisorCommand(config);
    if (!resolved) {
        opts?.onError?.(new shutdownErrors_js_1.ShutdownSupervisorResolveError("Shutdown supervisor entry missing (expected Stagehand CLI entrypoint)."), "resolve");
        return null;
    }
    const child = (0, node_child_process_1.spawn)(resolved.command, resolved.args, {
        // stdin is the parent lifeline.
        // Preserve supervisor stderr so crash-cleanup debug lines are visible.
        stdio: ["pipe", "ignore", "inherit"],
        detached: true,
    });
    child.on("error", (error) => {
        opts?.onError?.(new shutdownErrors_js_1.ShutdownSupervisorSpawnError(`Shutdown supervisor failed to start: ${error.message}`), "spawn");
    });
    try {
        child.unref();
        const stdin = child.stdin;
        stdin?.unref?.();
    }
    catch {
        // best-effort: avoid keeping the event loop alive
    }
    const stop = () => {
        // Normal close path: terminate supervisor directly.
        try {
            child.kill("SIGTERM");
        }
        catch {
            // ignore
        }
    };
    return { stop };
}
