// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/deepLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeepLocatorDelegate = void 0;
exports.deepLocatorThroughIframes = deepLocatorThroughIframes;
exports.resolveLocatorTarget = resolveLocatorTarget;
exports.resolveLocatorWithHops = resolveLocatorWithHops;
exports.deepLocatorFromPage = deepLocatorFromPage;
const locator_js_1 = require("./locator.js");
const frameLocator_js_1 = require("./frameLocator.js");
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const focusSelectors_js_1 = require("./a11y/snapshot/focusSelectors.js");
/** Parse XPath into steps preserving '/' vs '//' and the raw token (with [n]) */
function parseXPath(path) {
    const s = path.trim();
    let i = 0;
    const steps = [];
    while (i < s.length) {
        let axis = "child";
        if (s.startsWith("//", i)) {
            axis = "desc";
            i += 2;
        }
        else if (s[i] === "/") {
            axis = "child";
            i += 1;
        }
        const start = i;
        while (i < s.length && s[i] !== "/")
            i++;
        const raw = s.slice(start, i).trim();
        if (!raw)
            continue;
        const name = raw.replace(/\[\d+\]\s*$/u, "").toLowerCase();
        steps.push({ axis, raw, name });
    }
    return steps;
}
function buildXPathFromSteps(steps) {
    let out = "";
    for (const st of steps) {
        out += st.axis === "desc" ? "//" : "/";
        out += st.raw; // keep predicates intact
    }
    return out || "/";
}
/** Build a Locator scoped to the correct frame for a deep XPath crossing iframes. */
async function deepLocatorThroughIframes(page, root, xpathOrSelector) {
    const target = await resolveDeepXPathTarget(page, root, xpathOrSelector);
    return new locator_js_1.Locator(target.frame, target.selector);
}
/**
 * Unified resolver that supports '>>' hop notation, deep XPath across iframes,
 * and plain single-frame selectors. Keeps hop logic in one shared place.
 */
async function resolveLocatorTarget(page, root, selectorRaw) {
    const sel = selectorRaw.trim();
    const parts = sel
        .split(">>")
        .map((s) => s.trim())
        .filter(Boolean);
    if (parts.length > 1) {
        // Build a FrameLocator chain for all but the last segment
        let fl = (0, frameLocator_js_1.frameLocatorFromFrame)(page, root, parts[0]);
        for (let i = 1; i < parts.length - 1; i++) {
            fl = fl.frameLocator(parts[i]);
        }
        const targetFrame = await fl.resolveFrame();
        return { frame: targetFrame, selector: parts[parts.length - 1] };
    }
    // No hops — delegate to XPath-aware deep resolver when needed
    const isXPath = sel.startsWith("xpath=") || sel.startsWith("/");
    if (isXPath) {
        return resolveDeepXPathTarget(page, root, sel);
    }
    return { frame: root, selector: sel };
}
async function resolveLocatorWithHops(page, root, selectorRaw) {
    const target = await resolveLocatorTarget(page, root, selectorRaw);
    return new locator_js_1.Locator(target.frame, target.selector);
}
/**
 * DeepLocatorDelegate: a lightweight wrapper that looks like a Locator and
 * resolves to the correct frame/element on each call using hop/deep-XPath logic.
 *
 * Returned by `page.deepLocator()` for ergonomic, await-free chaining:
 *   page.deepLocator('iframe#ifrA >> #btn').click()
 */
class DeepLocatorDelegate {
    page;
    root;
    selector;
    nthIndex;
    constructor(page, root, selector, nthIndex = 0) {
        this.page = page;
        this.root = root;
        this.selector = selector;
        this.nthIndex = nthIndex;
    }
    async real() {
        const base = await resolveLocatorWithHops(this.page, this.root, this.selector);
        return base.nth(this.nthIndex);
    }
    // Locator API delegates
    async click(options) {
        return (await this.real()).click(options);
    }
    async count() {
        return (await this.real()).count();
    }
    async hover() {
        return (await this.real()).hover();
    }
    async fill(value) {
        return (await this.real()).fill(value);
    }
    async type(text, options) {
        return (await this.real()).type(text, options);
    }
    async selectOption(values) {
        return (await this.real()).selectOption(values);
    }
    async scrollTo(percent) {
        return (await this.real()).scrollTo(percent);
    }
    async isVisible() {
        return (await this.real()).isVisible();
    }
    async isChecked() {
        return (await this.real()).isChecked();
    }
    async inputValue() {
        return (await this.real()).inputValue();
    }
    async textContent() {
        return (await this.real()).textContent();
    }
    async innerHtml() {
        return (await this.real()).innerHtml();
    }
    async innerText() {
        return (await this.real()).innerText();
    }
    async centroid() {
        return (await this.real()).centroid();
    }
    async backendNodeId() {
        return (await this.real()).backendNodeId();
    }
    async highlight(options) {
        return (await this.real()).highlight(options);
    }
    async sendClickEvent(options) {
        return (await this.real()).sendClickEvent(options);
    }
    async setInputFiles(files) {
        return (await this.real()).setInputFiles(files);
    }
    first() {
        return this.nth(0);
    }
    nth(index) {
        const value = Number(index);
        if (!Number.isFinite(value) || value < 0) {
            throw new sdkErrors_js_1.StagehandInvalidArgumentError("deepLocator().nth() expects a non-negative index");
        }
        const nextIndex = Math.floor(value);
        if (nextIndex === this.nthIndex)
            return this;
        return new DeepLocatorDelegate(this.page, this.root, this.selector, nextIndex);
    }
}
exports.DeepLocatorDelegate = DeepLocatorDelegate;
/** Factory to create a deep locator delegate from a Page + root frame. */
function deepLocatorFromPage(page, root, selector) {
    return new DeepLocatorDelegate(page, root, selector);
}
async function resolveDeepXPathTarget(page, root, xpathOrSelector) {
    let path = xpathOrSelector.trim();
    if (path.startsWith("xpath="))
        path = path.slice("xpath=".length).trim();
    if (!path.startsWith("/"))
        path = "/" + path;
    const steps = parseXPath(path);
    let fl;
    let buf = [];
    const flushIntoFrameLocator = () => {
        if (!buf.length)
            return;
        const selectorForIframe = "xpath=" + buildXPathFromSteps(buf);
        fl = fl
            ? fl.frameLocator(selectorForIframe)
            : (0, frameLocator_js_1.frameLocatorFromFrame)(page, root, selectorForIframe);
        buf = [];
    };
    for (const st of steps) {
        buf.push(st);
        if (focusSelectors_js_1.IFRAME_STEP_RE.test(st.name))
            flushIntoFrameLocator();
    }
    const finalSelector = "xpath=" + buildXPathFromSteps(buf);
    const targetFrame = fl ? await fl.resolveFrame() : root;
    return { frame: targetFrame, selector: finalSelector };
}
