// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/executionContextRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executionContexts = exports.ExecutionContextRegistry = void 0;
class ExecutionContextRegistry {
    byFrame = new WeakMap();
    byExec = new WeakMap();
    /** Wire listeners for this session. Call BEFORE Runtime.enable. */
    attachSession(session) {
        const onCreated = (evt) => {
            const aux = (evt.context.auxData ?? {});
            if (aux.isDefault === true && typeof aux.frameId === "string") {
                this.register(session, aux.frameId, evt.context.id);
            }
        };
        const onDestroyed = (evt) => {
            const rev = this.byExec.get(session);
            const fwd = this.byFrame.get(session);
            if (!rev || !fwd)
                return;
            const frameId = rev.get(evt.executionContextId);
            if (!frameId)
                return;
            rev.delete(evt.executionContextId);
            if (fwd.get(frameId) === evt.executionContextId)
                fwd.delete(frameId);
        };
        const onCleared = () => {
            this.byFrame.delete(session);
            this.byExec.delete(session);
        };
        session.on("Runtime.executionContextCreated", onCreated);
        session.on("Runtime.executionContextDestroyed", onDestroyed);
        session.on("Runtime.executionContextsCleared", onCleared);
    }
    getMainWorld(session, frameId) {
        return this.byFrame.get(session)?.get(frameId) ?? null;
    }
    async waitForMainWorld(session, frameId, timeoutMs = 800) {
        const cached = this.getMainWorld(session, frameId);
        if (cached)
            return cached;
        await session.send("Runtime.enable").catch(() => { });
        const after = this.getMainWorld(session, frameId);
        if (after)
            return after;
        return await new Promise((resolve, reject) => {
            let done = false;
            const onCreated = (evt) => {
                const aux = (evt.context.auxData ?? {});
                if (aux.isDefault === true && aux.frameId === frameId) {
                    this.register(session, frameId, evt.context.id);
                    if (!done) {
                        done = true;
                        clearTimeout(timer);
                        session.off("Runtime.executionContextCreated", onCreated);
                        resolve(evt.context.id);
                    }
                }
            };
            const timer = setTimeout(() => {
                if (!done) {
                    done = true;
                    session.off("Runtime.executionContextCreated", onCreated);
                    reject(new Error(`main world not ready for frame ${frameId}`));
                }
            }, timeoutMs);
            session.on("Runtime.executionContextCreated", onCreated);
        });
    }
    register(session, frameId, ctxId) {
        let fwd = this.byFrame.get(session);
        if (!fwd) {
            fwd = new Map();
            this.byFrame.set(session, fwd);
        }
        let rev = this.byExec.get(session);
        if (!rev) {
            rev = new Map();
            this.byExec.set(session, rev);
        }
        fwd.set(frameId, ctxId);
        rev.set(ctxId, frameId);
    }
}
exports.ExecutionContextRegistry = ExecutionContextRegistry;
exports.executionContexts = new ExecutionContextRegistry();
