// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/fileUploadUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeInputFiles = normalizeInputFiles;
exports.toBuffer = toBuffer;
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const buffer_1 = require("buffer");
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const DEFAULT_MIME_TYPE = "application/octet-stream";
/**
 * Normalize user-provided setInputFiles arguments into in-memory payloads.
 * - Resolves string paths relative to the provided base directory.
 * - Validates that each path exists and is a regular file.
 * - Converts all buffers into Node Buffers for downstream processing.
 */
async function normalizeInputFiles(files, opts = {}) {
    if (files === null || files === undefined)
        return [];
    const flattened = Array.isArray(files)
        ? files
        : [files];
    if (!flattened.length)
        return [];
    const baseDir = opts.baseDir ?? process.cwd();
    const normalized = [];
    for (const entry of flattened) {
        if (typeof entry === "string") {
            const absolutePath = path_1.default.isAbsolute(entry)
                ? entry
                : path_1.default.resolve(baseDir, entry);
            const stat = await statFile(absolutePath);
            if (!stat.isFile()) {
                throw new sdkErrors_js_1.StagehandInvalidArgumentError(`setInputFiles(): expected a file but received directory or special entry at ${absolutePath}`);
            }
            const buffer = await fs_1.promises.readFile(absolutePath);
            normalized.push({
                name: path_1.default.basename(absolutePath) || "upload.bin",
                mimeType: DEFAULT_MIME_TYPE,
                buffer,
                lastModified: stat.mtimeMs || Date.now(),
                absolutePath,
            });
            continue;
        }
        if (entry && typeof entry === "object" && "buffer" in entry) {
            const payload = entry;
            const buffer = toBuffer(payload.buffer);
            normalized.push({
                name: payload.name || "upload.bin",
                mimeType: payload.mimeType || DEFAULT_MIME_TYPE,
                buffer,
                lastModified: typeof payload.lastModified === "number"
                    ? payload.lastModified
                    : Date.now(),
            });
            continue;
        }
        throw new sdkErrors_js_1.StagehandInvalidArgumentError("setInputFiles(): expected file path(s) or payload object(s)");
    }
    return normalized;
}
async function statFile(absolutePath) {
    try {
        return await fs_1.promises.stat(absolutePath);
    }
    catch (error) {
        const code = error?.code;
        if (code === "ENOENT") {
            throw new sdkErrors_js_1.StagehandInvalidArgumentError(`setInputFiles(): file not found at ${absolutePath}`);
        }
        throw error;
    }
}
function toBuffer(data) {
    if (buffer_1.Buffer.isBuffer(data))
        return data;
    if (data instanceof Uint8Array)
        return buffer_1.Buffer.from(data);
    if (typeof data === "string")
        return buffer_1.Buffer.from(data);
    if (data instanceof ArrayBuffer)
        return buffer_1.Buffer.from(new Uint8Array(data));
    throw new sdkErrors_js_1.StagehandInvalidArgumentError("Unsupported file payload buffer type");
}
