// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/piercer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installV3PiercerIntoSession = installV3PiercerIntoSession;
exports.tapPiercerConsole = tapPiercerConsole;
const logger_js_1 = require("../logger.js");
const scriptV3Content_js_1 = require("../dom/build/scriptV3Content.js");
const reRenderScriptContent_js_1 = require("../dom/build/reRenderScriptContent.js");
async function installV3PiercerIntoSession(session) {
    const pageEnabled = await session
        .send("Page.enable")
        .then(() => true)
        .catch(() => false);
    if (!pageEnabled)
        return false;
    await session.send("Runtime.enable").catch(() => { });
    try {
        await session.send("Page.addScriptToEvaluateOnNewDocument", { source: scriptV3Content_js_1.v3ScriptContent, runImmediately: true });
    }
    catch (e) {
        const msg = String(e?.message ?? e ?? "");
        // If the session vanished during attach (common with short-lived OOPIFs),
        // swallow and report failure so callers can early-return.
        if (msg.includes("Session with given id not found"))
            return false;
        // For other errors, keep going but don't throw — the next evaluate is idempotent.
    }
    await session
        .send("Runtime.evaluate", {
        expression: scriptV3Content_js_1.v3ScriptContent,
        returnByValue: true,
        awaitPromise: true,
    })
        .catch(() => { });
    // After the piercer is in place, re-render any custom elements whose
    // shadow roots were created before we patched attachShadow so their
    // closed roots are recreated under the hook.
    await session
        .send("Runtime.evaluate", {
        expression: reRenderScriptContent_js_1.reRenderScriptContent,
        returnByValue: true,
        awaitPromise: false,
    })
        .catch(() => { });
    return true;
}
/** (Optional) stream patch logs in your node console during bring-up */
function tapPiercerConsole(session, label) {
    session.on("Runtime.consoleAPICalled", (evt) => {
        const head = evt.args?.[0]?.value;
        if (head?.startsWith?.("[v3-piercer]")) {
            (0, logger_js_1.v3Logger)({
                category: "piercer",
                message: `[${label}] ${head}`,
                level: 2,
                auxiliary: {
                    value: {
                        value: String(evt.args?.[1]?.value ?? ""),
                        type: "string",
                    },
                },
            });
        }
    });
}
