// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/response.js
"use strict";
/**
 * Response
 * -----------------
 *
 * This module implements a Playwright-inspired response wrapper that exposes
 * navigation metadata and helpers for retrieving HTTP response bodies. The
 * abstraction is consumed by navigation routines (e.g. `Page.goto`) so callers
 * can synchronously inspect status codes, lazily fetch body text, or await the
 * network layer finishing the request. The implementation is built directly on
 * Chrome DevTools Protocol primitives – it holds the originating `requestId`
 * so it can request payloads via `Network.getResponseBody`, and it listens for
 * `responseReceivedExtraInfo`, `loadingFinished`, and `loadingFailed` events to
 * hydrate the richer header view and resolve callers waiting on completion.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = void 0;
exports.isSerializableResponse = isSerializableResponse;
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
function isSerializableResponse(value) {
    if (!value || typeof value !== "object")
        return false;
    const candidate = value;
    if (typeof candidate.requestId !== "string")
        return false;
    if (!candidate.response || typeof candidate.response !== "object") {
        return false;
    }
    return true;
}
function createDeferred() {
    let resolve;
    let reject;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return { promise, resolve, reject };
}
/** Normalise header names to lowercase for case-insensitive lookups. */
function normaliseHeaderName(name) {
    return name.toLowerCase();
}
/** Split multi-value header strings into discrete values while trimming. */
function splitHeaderValues(value) {
    return value
        .split(/\r?\n/)
        .map((part) => part.trim())
        .filter(Boolean);
}
/**
 * Parse an HTTP header text block (as emitted by CDP) into an ordered array of
 * name/value pairs while preserving the wire casing.
 */
function parseHeadersText(headersText) {
    if (!headersText)
        return [];
    const lines = headersText.split(/\r?\n/);
    const entries = [];
    for (const line of lines) {
        if (!line || line.startsWith("HTTP/"))
            continue;
        const index = line.indexOf(":");
        if (index === -1)
            continue;
        const name = line.slice(0, index).trim();
        const value = line.slice(index + 1).trim();
        entries.push({ name, value });
    }
    return entries;
}
/**
 * Thin wrapper around CDP response metadata that mirrors the ergonomics of
 * Playwright's `Response` class. The class intentionally keeps the same method
 * names so upstream integrations can transition with minimal code changes.
 */
class Response {
    page;
    session;
    requestId;
    frameId;
    loaderId;
    response;
    fromServiceWorkerFlag;
    serverAddress;
    headersObject;
    headersArrayCache = null;
    allHeadersCache = null;
    headerValuesMap = new Map();
    finishedDeferred = createDeferred();
    finishedSettled = false;
    extraInfoHeaders = null;
    extraInfoHeadersText;
    /**
     * Build a response wrapper from the CDP notification associated with a
     * navigation. The constructor captures the owning page/session so follow-up
     * methods (body/text/json) can query CDP on-demand. The `response` payload is
     * the raw `Protocol.Network.Response` object emitted by Chrome.
     */
    constructor(params) {
        this.page = params.page;
        this.session = params.session;
        this.requestId = params.requestId;
        this.frameId = params.frameId;
        this.loaderId = params.loaderId;
        this.response = params.response;
        this.fromServiceWorkerFlag = params.fromServiceWorker;
        if (params.response.remoteIPAddress &&
            params.response.remotePort !== undefined) {
            this.serverAddress = {
                ipAddress: params.response.remoteIPAddress,
                port: params.response.remotePort,
            };
        }
        else {
            this.serverAddress = null;
        }
        this.headersObject = {};
        for (const [name, value] of Object.entries(this.response.headers ?? {})) {
            const lower = normaliseHeaderName(name);
            if (value === undefined)
                continue;
            const values = splitHeaderValues(String(value));
            this.headerValuesMap.set(lower, values);
            this.headersObject[lower] = values.join(", ");
        }
    }
    /** URL associated with the navigation request. */
    url() {
        return this.response.url;
    }
    /** HTTP status code reported by Chrome. */
    status() {
        return this.response.status;
    }
    /** Human-readable status text that accompanied the response. */
    statusText() {
        return this.response.statusText;
    }
    /** Convenience predicate that checks for 2xx statuses. */
    ok() {
        const status = this.status();
        return status >= 200 && status <= 299;
    }
    /** Returns the Stagehand frame object that initiated the navigation. */
    frame() {
        if (!this.frameId)
            return null;
        try {
            return this.page.frameForId(this.frameId);
        }
        catch {
            return null;
        }
    }
    /** Indicates whether the response was serviced by a Service Worker. */
    fromServiceWorker() {
        return this.fromServiceWorkerFlag;
    }
    /**
     * Returns TLS security metadata when provided by the browser. In practice
     * this includes certificate issuer, protocol, and validity interval.
     */
    async securityDetails() {
        return this.response.securityDetails ?? null;
    }
    /** Returns the resolved server address for the navigation when available. */
    async serverAddr() {
        return this.serverAddress ?? null;
    }
    /**
     * Returns the response headers normalised to lowercase keys. Matches the
     * behaviour of Playwright's `headers()` by eliding duplicate header entries.
     */
    headers() {
        return { ...this.headersObject };
    }
    /**
     * Returns all headers including those only surfaced through
     * `responseReceivedExtraInfo` such as `set-cookie`. Values are reported as the
     * browser sends them (no further splitting or concatenation).
     */
    async allHeaders() {
        if (this.allHeadersCache)
            return { ...this.allHeadersCache };
        const source = this.extraInfoHeaders ?? this.response.headers ?? {};
        const map = {};
        for (const [name, value] of Object.entries(source)) {
            map[name] = String(value);
        }
        this.allHeadersCache = map;
        return { ...map };
    }
    /** Returns a concatenated header string for the supplied header name. */
    async headerValue(name) {
        const values = await this.headerValues(name);
        if (!values.length)
            return null;
        return values.join(", ");
    }
    /** Returns all values for a header (case-insensitive lookup). */
    async headerValues(name) {
        const lower = normaliseHeaderName(name);
        if (this.extraInfoHeaders) {
            const raw = this.extraInfoHeaders[name] ?? this.extraInfoHeaders[lower];
            if (raw !== undefined) {
                return splitHeaderValues(String(raw));
            }
        }
        const values = this.headerValuesMap.get(lower);
        return values ? [...values] : [];
    }
    /**
     * Returns header entries preserving their original wire casing and ordering.
     * Falls back to the CDP object when the raw header text is unavailable.
     */
    async headersArray() {
        if (this.headersArrayCache)
            return [...this.headersArrayCache];
        const entriesFromText = parseHeadersText(this.extraInfoHeadersText);
        if (entriesFromText.length > 0) {
            this.headersArrayCache = entriesFromText;
            return [...entriesFromText];
        }
        const entries = [];
        const source = this.extraInfoHeaders ?? this.response.headers ?? {};
        for (const [name, value] of Object.entries(source)) {
            const values = splitHeaderValues(String(value));
            for (const val of values) {
                entries.push({ name, value: val });
            }
        }
        this.headersArrayCache = entries;
        return [...entries];
    }
    /**
     * Requests the raw response body from Chrome DevTools Protocol. The method is
     * intentionally lazy because not every caller needs the payload, and CDP only
     * allows retrieving it once the response completes.
     */
    async body() {
        const result = await this.session
            .send("Network.getResponseBody", { requestId: this.requestId })
            .catch((error) => {
            throw new sdkErrors_js_1.ResponseBodyError(String(error));
        });
        if (result.base64Encoded) {
            return Buffer.from(result.body, "base64");
        }
        return Buffer.from(result.body, "utf-8");
    }
    /** Decodes the response body as UTF-8 text. */
    async text() {
        const buffer = await this.body();
        return buffer.toString("utf-8");
    }
    /** Parses the response body as JSON and throws if parsing fails. */
    async json() {
        const text = await this.text();
        try {
            return JSON.parse(text);
        }
        catch (error) {
            throw new sdkErrors_js_1.ResponseParseError(String(error));
        }
    }
    /**
     * Resolves once the underlying network request completes or fails. Mirrors
     * Playwright's behaviour by resolving to `null` on success and to an `Error`
     * instance when Chrome reports `Network.loadingFailed`.
     */
    async finished() {
        return this.finishedDeferred.promise;
    }
    /**
     * Internal helper invoked by the navigation tracker when CDP reports extra
     * header information. This keeps the cached header views in sync with the
     * richer metadata.
     */
    applyExtraInfo(event) {
        this.extraInfoHeaders = event.headers;
        this.extraInfoHeadersText = event.headersText;
        this.allHeadersCache = null;
        this.headersArrayCache = null;
        this.headersObject = {};
        this.headerValuesMap.clear();
        const source = event.headers ?? {};
        for (const [name, value] of Object.entries(source)) {
            const lower = normaliseHeaderName(name);
            const segments = splitHeaderValues(String(value));
            this.headerValuesMap.set(lower, segments);
            this.headersObject[lower] = segments.join(", ");
        }
    }
    /**
     * Internal helper for creating a Response object from a Serializable
     * goto response from the Stagehand API
     */
    static fromSerializable(serialized, context) {
        const reconstructed = new Response({
            page: context.page,
            session: context.session,
            requestId: serialized.requestId,
            frameId: serialized.frameId,
            loaderId: serialized.loaderId,
            response: serialized.response,
            fromServiceWorker: serialized.fromServiceWorkerFlag ?? false,
        });
        if (serialized.extraInfoHeaders) {
            reconstructed.applyExtraInfo({
                requestId: serialized.requestId,
                headers: serialized.extraInfoHeaders,
                headersText: serialized.extraInfoHeadersText,
            });
        }
        if (serialized.finishedSettled) {
            reconstructed.markFinished(null);
        }
        return reconstructed;
    }
    /** Marks the response as finished and resolves the `finished()` promise. */
    markFinished(error) {
        if (this.finishedSettled)
            return;
        this.finishedSettled = true;
        if (error) {
            this.finishedDeferred.resolve(error);
        }
        else {
            this.finishedDeferred.resolve(null);
        }
    }
}
exports.Response = Response;
