// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/selectorResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSelectorResolver = void 0;
const locatorScripts_generated_js_1 = require("../dom/build/locatorScripts.generated.js");
const logger_js_1 = require("../logger.js");
const executionContextRegistry_js_1 = require("./executionContextRegistry.js");
class FrameSelectorResolver {
    frame;
    constructor(frame) {
        this.frame = frame;
    }
    static parseSelector(raw) {
        const trimmed = raw.trim();
        const isText = /^text=/i.test(trimmed);
        const looksLikeXPath = /^xpath=/i.test(trimmed) ||
            trimmed.startsWith("/") ||
            trimmed.startsWith("(");
        const isCssPrefixed = /^css=/i.test(trimmed);
        if (isText) {
            let value = trimmed.replace(/^text=/i, "").trim();
            if ((value.startsWith('"') && value.endsWith('"')) ||
                (value.startsWith("'") && value.endsWith("'"))) {
                value = value.slice(1, -1);
            }
            return { kind: "text", value };
        }
        if (looksLikeXPath) {
            const value = trimmed.replace(/^xpath=/i, "");
            return { kind: "xpath", value };
        }
        let selector = isCssPrefixed ? trimmed.replace(/^css=/i, "") : trimmed;
        if (selector.includes(">>")) {
            selector = selector
                .split(">>")
                .map((piece) => piece.trim())
                .filter(Boolean)
                .join(" ");
        }
        return { kind: "css", value: selector };
    }
    async resolveFirst(query) {
        return this.resolveAtIndex(query, 0);
    }
    async resolveAll(query, { limit = Infinity } = {}) {
        if (limit <= 0)
            return [];
        switch (query.kind) {
            case "css":
                return this.resolveCss(query.value, limit);
            case "text":
                return this.resolveText(query.value, limit);
            case "xpath":
                return this.resolveXPath(query.value, limit);
            default:
                return [];
        }
    }
    async count(query) {
        switch (query.kind) {
            case "css":
                return this.countCss(query.value);
            case "text":
                return this.countText(query.value);
            case "xpath":
                return this.countXPath(query.value);
            default:
                return 0;
        }
    }
    async resolveAtIndex(query, index) {
        if (index < 0 || !Number.isFinite(index))
            return null;
        const results = await this.resolveAll(query, { limit: index + 1 });
        return results[index] ?? null;
    }
    buildLocatorInvocation(name, args) {
        const call = `${locatorScripts_generated_js_1.locatorScriptGlobalRefs[name]}(${args.join(", ")})`;
        return `(() => { ${locatorScripts_generated_js_1.locatorScriptBootstrap}; return ${call}; })()`;
    }
    async resolveCss(selector, limit) {
        if (limit <= 0)
            return [];
        const session = this.frame.session;
        const { executionContextId } = await session.send("Page.createIsolatedWorld", {
            frameId: this.frame.frameId,
            worldName: "v3-world",
        });
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const results = [];
        let loggedFallback = false;
        for (let index = 0; index < limit; index += 1) {
            const primaryExpr = this.buildLocatorInvocation("resolveCssSelector", [
                JSON.stringify(selector),
                String(index),
            ]);
            const primary = await this.evaluateElement(primaryExpr, executionContextId);
            if (primary) {
                results.push(primary);
                continue;
            }
            if (!loggedFallback) {
                (0, logger_js_1.v3Logger)({
                    category: "locator",
                    message: "css pierce-fallback",
                    level: 2,
                    auxiliary: {
                        frameId: { value: String(this.frame.frameId), type: "string" },
                        selector: { value: selector, type: "string" },
                    },
                });
                loggedFallback = true;
            }
            const fallbackExpr = this.buildLocatorInvocation("resolveCssSelectorPierce", [JSON.stringify(selector), String(index)]);
            const fallback = await this.evaluateElement(fallbackExpr, ctxId);
            if (fallback) {
                results.push(fallback);
                continue;
            }
            break;
        }
        return results;
    }
    async resolveText(value, limit) {
        if (limit <= 0)
            return [];
        const session = this.frame.session;
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const results = [];
        for (let index = 0; index < limit; index += 1) {
            const expr = this.buildLocatorInvocation("resolveTextSelector", [
                JSON.stringify(value),
                String(index),
            ]);
            const resolved = await this.evaluateElement(expr, ctxId);
            if (!resolved)
                break;
            results.push(resolved);
        }
        return results;
    }
    async resolveXPath(value, limit) {
        if (limit <= 0)
            return [];
        const session = this.frame.session;
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const results = [];
        for (let index = 0; index < limit; index += 1) {
            const expr = this.buildLocatorInvocation("resolveXPathMainWorld", [
                JSON.stringify(value),
                String(index),
            ]);
            const resolved = await this.evaluateElement(expr, ctxId);
            if (!resolved)
                break;
            results.push(resolved);
        }
        return results;
    }
    async countCss(selector) {
        const session = this.frame.session;
        const { executionContextId } = await session.send("Page.createIsolatedWorld", {
            frameId: this.frame.frameId,
            worldName: "v3-world",
        });
        const primaryExpr = this.buildLocatorInvocation("countCssMatchesPrimary", [
            JSON.stringify(selector),
        ]);
        const primary = await this.evaluateCount(primaryExpr, executionContextId);
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const fallbackExpr = this.buildLocatorInvocation("countCssMatchesPierce", [
            JSON.stringify(selector),
        ]);
        const fallback = await this.evaluateCount(fallbackExpr, ctxId);
        return Math.max(primary, fallback);
    }
    async countText(value) {
        const session = this.frame.session;
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const expr = this.buildLocatorInvocation("countTextMatches", [
            JSON.stringify(value),
        ]);
        try {
            const evalRes = await session.send("Runtime.evaluate", {
                expression: expr,
                contextId: ctxId,
                returnByValue: true,
                awaitPromise: true,
            });
            if (evalRes.exceptionDetails) {
                const details = evalRes.exceptionDetails;
                (0, logger_js_1.v3Logger)({
                    category: "locator",
                    message: "count text evaluate exception",
                    level: 0,
                    auxiliary: {
                        frameId: { value: String(this.frame.frameId), type: "string" },
                        selector: { value: value, type: "string" },
                        exception: {
                            value: details.text ??
                                String(details.exception?.description ??
                                    details.exception?.value ??
                                    ""),
                            type: "string",
                        },
                    },
                });
                return 0;
            }
            const data = (evalRes.result.value ?? {});
            const num = typeof data.count === "number" ? data.count : Number(data.count);
            if (!Number.isFinite(num))
                return 0;
            return Math.max(0, Math.floor(num));
        }
        catch {
            return 0;
        }
    }
    async countXPath(value) {
        const session = this.frame.session;
        const ctxId = await executionContextRegistry_js_1.executionContexts.waitForMainWorld(session, this.frame.frameId, 1000);
        const expr = this.buildLocatorInvocation("countXPathMatchesMainWorld", [
            JSON.stringify(value),
        ]);
        try {
            const evalRes = await session.send("Runtime.evaluate", {
                expression: expr,
                contextId: ctxId,
                returnByValue: true,
                awaitPromise: true,
            });
            if (evalRes.exceptionDetails) {
                return 0;
            }
            const num = typeof evalRes.result.value === "number"
                ? evalRes.result.value
                : Number(evalRes.result.value);
            if (!Number.isFinite(num))
                return 0;
            return Math.max(0, Math.floor(num));
        }
        catch {
            return 0;
        }
    }
    async resolveFromObjectId(objectId) {
        const session = this.frame.session;
        let nodeId;
        try {
            const rn = await session.send("DOM.requestNode", { objectId });
            nodeId = rn.nodeId ?? null;
        }
        catch {
            nodeId = null;
        }
        return { objectId, nodeId };
    }
    async evaluateCount(expression, contextId) {
        const session = this.frame.session;
        try {
            const evalRes = await session.send("Runtime.evaluate", {
                expression,
                contextId,
                returnByValue: true,
                awaitPromise: true,
            });
            if (evalRes.exceptionDetails) {
                return 0;
            }
            const value = evalRes.result.value;
            const num = typeof value === "number" ? value : Number(value);
            if (!Number.isFinite(num))
                return 0;
            return Math.max(0, Math.floor(num));
        }
        catch {
            return 0;
        }
    }
    async evaluateElement(expression, contextId) {
        const session = this.frame.session;
        try {
            const evalRes = await session.send("Runtime.evaluate", {
                expression,
                contextId,
                returnByValue: false,
                awaitPromise: true,
            });
            if (evalRes.exceptionDetails || !evalRes.result.objectId) {
                return null;
            }
            return this.resolveFromObjectId(evalRes.result.objectId);
        }
        catch {
            return null;
        }
    }
}
exports.FrameSelectorResolver = FrameSelectorResolver;
