// @browserbasehq/stagehand/dist/cjs/tests/unit/public-api/export-surface.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const stagehand_1 = __importStar(require("@browserbasehq/stagehand")), Stagehand = stagehand_1;
const public_error_types_test_js_1 = require("./public-error-types.test.js");
// Type matcher guidelines:
//
// toEqualTypeOf – Default. Assert full, deep type equality; any type change should fail.
//   e.g. expectTypeOf<ReturnType<typeof foo>>().toEqualTypeOf<FooResult>()
//
// toMatchObjectType – Assert (part of) an object's shape while allowing extra fields.
//   e.g. expectTypeOf(user).toMatchObjectType<{ id: string; email: string }>()
//
// toExtend – Assert that a type is compatible with a broader contract (assignable/extends).
//   e.g. expectTypeOf<User>().toExtend<BaseUser>()
const publicApiShape = {
    __internalMaybeRunShutdownSupervisorFromArgv: Stagehand.__internalMaybeRunShutdownSupervisorFromArgv,
    __internalCreateInMemoryAgentCacheHandle: Stagehand.__internalCreateInMemoryAgentCacheHandle,
    AISdkClient: Stagehand.AISdkClient,
    Api: Stagehand.Api,
    AVAILABLE_CUA_MODELS: Stagehand.AVAILABLE_CUA_MODELS,
    AgentProvider: Stagehand.AgentProvider,
    AnnotatedScreenshotText: Stagehand.AnnotatedScreenshotText,
    ConsoleMessage: Stagehand.ConsoleMessage,
    CustomOpenAIClient: Stagehand.CustomOpenAIClient,
    LLMClient: Stagehand.LLMClient,
    LOG_LEVEL_NAMES: Stagehand.LOG_LEVEL_NAMES,
    Response: Stagehand.Response,
    Stagehand: Stagehand.Stagehand,
    V3: Stagehand.V3,
    V3Evaluator: Stagehand.V3Evaluator,
    V3FunctionName: Stagehand.V3FunctionName,
    connectToMCPServer: Stagehand.connectToMCPServer,
    default: stagehand_1.default,
    defaultExtractSchema: Stagehand.defaultExtractSchema,
    getAISDKLanguageModel: Stagehand.getAISDKLanguageModel,
    getZodType: Stagehand.getZodType,
    injectUrls: Stagehand.injectUrls,
    isRunningInBun: Stagehand.isRunningInBun,
    isZod3Schema: Stagehand.isZod3Schema,
    isZod4Schema: Stagehand.isZod4Schema,
    jsonSchemaToZod: Stagehand.jsonSchemaToZod,
    loadApiKeyFromEnv: Stagehand.loadApiKeyFromEnv,
    localBrowserLaunchOptionsSchema: Stagehand.localBrowserLaunchOptionsSchema,
    modelToAgentProviderMap: Stagehand.modelToAgentProviderMap,
    pageTextSchema: Stagehand.pageTextSchema,
    providerEnvVarMap: Stagehand.providerEnvVarMap,
    toGeminiSchema: Stagehand.toGeminiSchema,
    toJsonSchema: Stagehand.toJsonSchema,
    tool: Stagehand.tool,
    transformSchema: Stagehand.transformSchema,
    trimTrailingTextNode: Stagehand.trimTrailingTextNode,
    validateZodSchema: Stagehand.validateZodSchema,
    ...public_error_types_test_js_1.publicErrorTypes,
};
(0, vitest_1.describe)("Stagehand public API export surface", () => {
    (0, vitest_1.it)("public API shape matches module exports", () => {
        const _check = publicApiShape;
        void _check;
    });
    (0, vitest_1.it)("does not expose unexpected top-level exports", () => {
        const expected = Object.keys(publicApiShape).sort();
        const actual = Object.keys(Stagehand).sort();
        (0, vitest_1.expect)(actual).toStrictEqual(expected);
    });
    (0, vitest_1.it)("default export mirrors the named export surface", () => {
        const expected = Object.keys(Stagehand)
            .filter((key) => key !== "default")
            .sort();
        const actual = Object.keys(stagehand_1.default).sort();
        (0, vitest_1.expect)(actual).toStrictEqual(expected);
    });
});
