// @browserbasehq/stagehand/dist/cjs/tests/unit/public-api/public-error-types.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicErrorTypes = void 0;
const vitest_1 = require("vitest");
const Stagehand = __importStar(require("@browserbasehq/stagehand"));
exports.publicErrorTypes = {
    AgentAbortError: Stagehand.AgentAbortError,
    CdpConnectionClosedError: Stagehand.CdpConnectionClosedError,
    AgentScreenshotProviderError: Stagehand.AgentScreenshotProviderError,
    BrowserbaseSessionNotFoundError: Stagehand.BrowserbaseSessionNotFoundError,
    CaptchaTimeoutError: Stagehand.CaptchaTimeoutError,
    ConnectionTimeoutError: Stagehand.ConnectionTimeoutError,
    ContentFrameNotFoundError: Stagehand.ContentFrameNotFoundError,
    CookieSetError: Stagehand.CookieSetError,
    CookieValidationError: Stagehand.CookieValidationError,
    CreateChatCompletionResponseError: Stagehand.CreateChatCompletionResponseError,
    CuaModelRequiredError: Stagehand.CuaModelRequiredError,
    ElementNotVisibleError: Stagehand.ElementNotVisibleError,
    ExperimentalApiConflictError: Stagehand.ExperimentalApiConflictError,
    ExperimentalNotConfiguredError: Stagehand.ExperimentalNotConfiguredError,
    HandlerNotInitializedError: Stagehand.HandlerNotInitializedError,
    InvalidAISDKModelFormatError: Stagehand.InvalidAISDKModelFormatError,
    LLMResponseError: Stagehand.LLMResponseError,
    MCPConnectionError: Stagehand.MCPConnectionError,
    MissingEnvironmentVariableError: Stagehand.MissingEnvironmentVariableError,
    MissingLLMConfigurationError: Stagehand.MissingLLMConfigurationError,
    PageNotFoundError: Stagehand.PageNotFoundError,
    ResponseBodyError: Stagehand.ResponseBodyError,
    ResponseParseError: Stagehand.ResponseParseError,
    StagehandAPIError: Stagehand.StagehandAPIError,
    StagehandAPIUnauthorizedError: Stagehand.StagehandAPIUnauthorizedError,
    StagehandClickError: Stagehand.StagehandClickError,
    StagehandClosedError: Stagehand.StagehandClosedError,
    StagehandDefaultError: Stagehand.StagehandDefaultError,
    StagehandDomProcessError: Stagehand.StagehandDomProcessError,
    StagehandElementNotFoundError: Stagehand.StagehandElementNotFoundError,
    StagehandEnvironmentError: Stagehand.StagehandEnvironmentError,
    StagehandError: Stagehand.StagehandError,
    StagehandEvalError: Stagehand.StagehandEvalError,
    StagehandHttpError: Stagehand.StagehandHttpError,
    StagehandIframeError: Stagehand.StagehandIframeError,
    StagehandInitError: Stagehand.StagehandInitError,
    StagehandInvalidArgumentError: Stagehand.StagehandInvalidArgumentError,
    StagehandLocatorError: Stagehand.StagehandLocatorError,
    StagehandMissingArgumentError: Stagehand.StagehandMissingArgumentError,
    StagehandNotInitializedError: Stagehand.StagehandNotInitializedError,
    StagehandResponseBodyError: Stagehand.StagehandResponseBodyError,
    StagehandResponseParseError: Stagehand.StagehandResponseParseError,
    StagehandServerError: Stagehand.StagehandServerError,
    StagehandShadowRootMissingError: Stagehand.StagehandShadowRootMissingError,
    StagehandShadowSegmentEmptyError: Stagehand.StagehandShadowSegmentEmptyError,
    StagehandShadowSegmentNotFoundError: Stagehand.StagehandShadowSegmentNotFoundError,
    StreamingCallbacksInNonStreamingModeError: Stagehand.StreamingCallbacksInNonStreamingModeError,
    StagehandSnapshotError: Stagehand.StagehandSnapshotError,
    TimeoutError: Stagehand.TimeoutError,
    UnsupportedAISDKModelProviderError: Stagehand.UnsupportedAISDKModelProviderError,
    UnsupportedModelError: Stagehand.UnsupportedModelError,
    UnsupportedModelProviderError: Stagehand.UnsupportedModelProviderError,
    XPathResolutionError: Stagehand.XPathResolutionError,
    ZodSchemaValidationError: Stagehand.ZodSchemaValidationError,
    ActTimeoutError: Stagehand.ActTimeoutError,
    ObserveTimeoutError: Stagehand.ObserveTimeoutError,
    ExtractTimeoutError: Stagehand.ExtractTimeoutError,
    UnderstudyCommandException: Stagehand.UnderstudyCommandException,
    StagehandSetExtraHTTPHeadersError: Stagehand.StagehandSetExtraHTTPHeadersError,
};
const errorTypes = Object.keys(exports.publicErrorTypes);
(0, vitest_1.describe)("Stagehand public error types", () => {
    (0, vitest_1.describe)("errors", () => {
        vitest_1.it.each(errorTypes)("%s extends Error", (errorTypeName) => {
            const ErrorClass = Stagehand[errorTypeName];
            (0, vitest_1.expectTypeOf)().toExtend();
            void ErrorClass; // Mark as used to satisfy ESLint
        });
    });
});
