// @browserbasehq/stagehand/dist/cjs/lib/v3/dom/locatorScripts/counts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.countCssMatchesPrimary = countCssMatchesPrimary;
exports.countCssMatchesPierce = countCssMatchesPierce;
exports.countTextMatches = countTextMatches;
exports.countXPathMatchesMainWorld = countXPathMatchesMainWorld;
const xpathResolver_js_1 = require("./xpathResolver.js");
function countCssMatchesPrimary(selectorRaw) {
    const selector = String(selectorRaw ?? "").trim();
    if (!selector)
        return 0;
    const seen = new WeakSet();
    const visit = (root) => {
        if (!root || seen.has(root))
            return 0;
        seen.add(root);
        let total = 0;
        try {
            const queryable = root;
            if (typeof queryable.querySelectorAll === "function") {
                total += queryable.querySelectorAll(selector).length;
            }
        }
        catch {
            // ignore query errors
        }
        try {
            const doc = root instanceof Document
                ? root
                : (root?.ownerDocument ?? document);
            const walker = doc.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
            let node;
            while ((node = walker.nextNode())) {
                if (node instanceof Element && node.shadowRoot) {
                    total += visit(node.shadowRoot);
                }
            }
        }
        catch {
            // ignore traversal errors
        }
        return total;
    };
    try {
        return visit(document);
    }
    catch {
        try {
            return document.querySelectorAll(selector).length;
        }
        catch {
            return 0;
        }
    }
}
function countCssMatchesPierce(selectorRaw) {
    const selector = String(selectorRaw ?? "").trim();
    if (!selector)
        return 0;
    const backdoor = window.__stagehandV3__;
    if (!backdoor || typeof backdoor.getClosedRoot !== "function") {
        try {
            return document.querySelectorAll(selector).length;
        }
        catch {
            return 0;
        }
    }
    const seen = new WeakSet();
    const queue = [];
    const enqueue = (node) => {
        if (!node || seen.has(node))
            return;
        seen.add(node);
        queue.push(node);
    };
    enqueue(document);
    let total = 0;
    const visitElement = (element) => {
        const open = element.shadowRoot;
        if (open)
            enqueue(open);
        try {
            const closed = backdoor.getClosedRoot(element);
            if (closed)
                enqueue(closed);
        }
        catch {
            // ignore
        }
    };
    while (queue.length) {
        const root = queue.shift();
        if (!root)
            continue;
        try {
            const queryable = root;
            if (typeof queryable.querySelectorAll === "function") {
                total += queryable.querySelectorAll(selector).length;
            }
        }
        catch {
            // ignore query errors
        }
        try {
            const doc = root instanceof Document
                ? root
                : root instanceof ShadowRoot
                    ? (root.host?.ownerDocument ?? document)
                    : (root.ownerDocument ?? document);
            const walker = doc.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
            let node;
            while ((node = walker.nextNode())) {
                if (node instanceof Element) {
                    visitElement(node);
                }
            }
        }
        catch {
            // ignore traversal errors
        }
    }
    return total;
}
function countTextMatches(rawNeedle) {
    const needle = String(rawNeedle ?? "");
    if (!needle) {
        return { count: 0, sample: [], error: null };
    }
    const needleLc = needle.toLowerCase();
    const skipTags = new Set([
        "SCRIPT",
        "STYLE",
        "TEMPLATE",
        "NOSCRIPT",
        "HEAD",
        "TITLE",
        "LINK",
        "META",
        "HTML",
        "BODY",
    ]);
    const shouldSkip = (node) => {
        if (!node)
            return false;
        const tag = node.tagName?.toUpperCase() ?? "";
        return skipTags.has(tag);
    };
    const extractText = (element) => {
        try {
            if (shouldSkip(element))
                return "";
            const inner = element.innerText;
            if (typeof inner === "string" && inner.trim())
                return inner.trim();
        }
        catch {
            // ignore
        }
        try {
            const text = element.textContent;
            if (typeof text === "string")
                return text.trim();
        }
        catch {
            // ignore
        }
        return "";
    };
    const matches = (element) => {
        const text = extractText(element);
        return !!text && text.toLowerCase().includes(needleLc);
    };
    const backdoor = window.__stagehandV3__;
    const getClosedRoot = backdoor && typeof backdoor.getClosedRoot === "function"
        ? (host) => {
            try {
                return backdoor.getClosedRoot(host) ?? null;
            }
            catch {
                return null;
            }
        }
        : (host) => {
            void host;
            return null;
        };
    const seen = new WeakSet();
    const queue = [];
    const enqueue = (node) => {
        if (!node || seen.has(node))
            return;
        seen.add(node);
        queue.push(node);
    };
    const walkerFor = (root) => {
        try {
            const doc = root instanceof Document
                ? root
                : (root?.ownerDocument ?? document);
            return doc.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
        }
        catch {
            return null;
        }
    };
    const matchesList = [];
    enqueue(document);
    while (queue.length) {
        const root = queue.shift();
        if (!root)
            continue;
        if (root instanceof Element && matches(root)) {
            matchesList.push({
                element: root,
                tag: root.tagName ?? "",
                id: root.id ?? "",
                className: root.className ?? "",
                text: extractText(root),
            });
        }
        const walker = walkerFor(root);
        if (!walker)
            continue;
        let node;
        while ((node = walker.nextNode())) {
            if (!(node instanceof Element))
                continue;
            if (matches(node)) {
                matchesList.push({
                    element: node,
                    tag: node.tagName ?? "",
                    id: node.id ?? "",
                    className: node.className ?? "",
                    text: extractText(node),
                });
            }
            const open = node.shadowRoot;
            if (open)
                enqueue(open);
            const closed = getClosedRoot(node);
            if (closed)
                enqueue(closed);
        }
    }
    const innermost = [];
    for (const item of matchesList) {
        const el = item.element;
        let skip = false;
        for (const other of matchesList) {
            if (item === other)
                continue;
            try {
                if (el.contains(other.element)) {
                    skip = true;
                    break;
                }
            }
            catch {
                // ignore containment errors
            }
        }
        if (!skip)
            innermost.push(item);
    }
    const count = innermost.length;
    const sample = innermost.slice(0, 5).map((item) => ({
        tag: item.tag,
        id: item.id,
        class: item.className,
        text: item.text,
    }));
    return { count, sample, error: null };
}
function countXPathMatchesMainWorld(rawXp) {
    return (0, xpathResolver_js_1.countXPathMatches)(rawXp, { pierceShadow: true });
}
