// @browserbasehq/stagehand/dist/cjs/lib/v3/llm/GroqClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroqClient = void 0;
const openai_1 = __importDefault(require("openai"));
const LLMClient_js_1 = require("./LLMClient.js");
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const zodCompat_js_1 = require("../zodCompat.js");
class GroqClient extends LLMClient_js_1.LLMClient {
    type = "groq";
    client;
    hasVision = false;
    constructor({ modelName, clientOptions, userProvidedInstructions, }) {
        super(modelName, userProvidedInstructions);
        // Create OpenAI client with the base URL set to Groq API
        this.client = new openai_1.default({
            baseURL: "https://api.groq.com/openai/v1",
            apiKey: clientOptions?.apiKey || process.env.GROQ_API_KEY,
            ...clientOptions,
        });
        this.modelName = modelName;
        this.clientOptions = clientOptions;
    }
    async createChatCompletion({ options, retries, logger, }) {
        const optionsWithoutImage = { ...options };
        delete optionsWithoutImage.image;
        logger({
            category: "groq",
            message: "creating chat completion",
            level: 2,
            auxiliary: {
                options: {
                    value: JSON.stringify(optionsWithoutImage),
                    type: "object",
                },
            },
        });
        // Format messages for Groq API (using OpenAI format)
        const formattedMessages = options.messages.map((msg) => {
            const baseMessage = {
                content: typeof msg.content === "string"
                    ? msg.content
                    : Array.isArray(msg.content) &&
                        msg.content.length > 0 &&
                        "text" in msg.content[0]
                        ? msg.content[0].text
                        : "",
            };
            // Groq supports system, user, and assistant roles
            if (msg.role === "system") {
                return { ...baseMessage, role: "system" };
            }
            else if (msg.role === "assistant") {
                return { ...baseMessage, role: "assistant" };
            }
            else {
                // Default to user for any other role
                return { ...baseMessage, role: "user" };
            }
        });
        // Format tools if provided
        let tools = options.tools?.map((tool) => ({
            type: "function",
            function: {
                name: tool.name,
                description: tool.description,
                parameters: {
                    type: "object",
                    properties: tool.parameters.properties,
                    required: tool.parameters.required,
                },
            },
        }));
        // Add response model as a tool if provided
        if (options.response_model) {
            const jsonSchema = (0, zodCompat_js_1.toJsonSchema)(options.response_model.schema);
            const schemaProperties = jsonSchema.properties || {};
            const schemaRequired = jsonSchema.required || [];
            const responseTool = {
                type: "function",
                function: {
                    name: "print_extracted_data",
                    description: "Prints the extracted data based on the provided schema.",
                    parameters: {
                        type: "object",
                        properties: schemaProperties,
                        required: schemaRequired,
                    },
                },
            };
            tools = tools ? [...tools, responseTool] : [responseTool];
        }
        try {
            // Use OpenAI client with Groq API
            const apiResponse = await this.client.chat.completions.create({
                model: this.modelName.split("groq-")[1],
                messages: [
                    ...formattedMessages,
                    // Add explicit instruction to return JSON if we have a response model
                    ...(options.response_model
                        ? [
                            {
                                role: "system",
                                content: `IMPORTANT: Your response must be valid JSON that matches this schema: ${JSON.stringify(options.response_model.schema)}`,
                            },
                        ]
                        : []),
                ],
                temperature: options.temperature || 0.7,
                max_tokens: options.maxOutputTokens,
                tools: tools,
                tool_choice: options.tool_choice || "auto",
            });
            // Format the response to match the expected LLMResponse format
            const response = {
                id: apiResponse.id,
                object: "chat.completion",
                created: Date.now(),
                model: this.modelName.split("groq-")[1],
                choices: [
                    {
                        index: 0,
                        message: {
                            role: "assistant",
                            content: apiResponse.choices[0]?.message?.content || null,
                            tool_calls: apiResponse.choices[0]?.message?.tool_calls || [],
                        },
                        finish_reason: apiResponse.choices[0]?.finish_reason || "stop",
                    },
                ],
                usage: {
                    prompt_tokens: apiResponse.usage?.prompt_tokens || 0,
                    completion_tokens: apiResponse.usage?.completion_tokens || 0,
                    total_tokens: apiResponse.usage?.total_tokens || 0,
                },
            };
            logger({
                category: "groq",
                message: "response",
                level: 2,
                auxiliary: {
                    response: {
                        value: JSON.stringify(response),
                        type: "object",
                    },
                    requestId: {
                        value: options.requestId,
                        type: "string",
                    },
                },
            });
            // If there's no response model, return the entire response object
            if (!options.response_model) {
                return response;
            }
            // Otherwise, try parsing the JSON from the tool call or content
            const toolCall = response.choices[0]?.message?.tool_calls?.[0];
            if (toolCall?.function?.arguments) {
                try {
                    const result = JSON.parse(toolCall.function.arguments);
                    const finalResponse = {
                        data: result,
                        usage: response.usage,
                    };
                    return finalResponse;
                }
                catch (e) {
                    logger({
                        category: "groq",
                        message: "failed to parse tool call arguments as JSON, retrying",
                        level: 0,
                        auxiliary: {
                            error: {
                                value: e.message,
                                type: "string",
                            },
                        },
                    });
                }
            }
            // If we have content but no tool calls, try to parse the content as JSON
            const content = response.choices[0]?.message?.content;
            if (content) {
                try {
                    // Try to extract JSON from the content
                    const jsonMatch = content.match(/\{[\s\S]*\}/);
                    if (jsonMatch) {
                        const result = JSON.parse(jsonMatch[0]);
                        const finalResponse = {
                            data: result,
                            usage: response.usage,
                        };
                        return finalResponse;
                    }
                }
                catch (e) {
                    logger({
                        category: "groq",
                        message: "failed to parse content as JSON",
                        level: 0,
                        auxiliary: {
                            error: {
                                value: e.message,
                                type: "string",
                            },
                        },
                    });
                }
            }
            // If we still haven't found valid JSON and have retries left, try again
            if (!retries || retries < 5) {
                return this.createChatCompletion({
                    options,
                    logger,
                    retries: (retries ?? 0) + 1,
                });
            }
            throw new sdkErrors_js_1.CreateChatCompletionResponseError("Invalid response schema");
        }
        catch (error) {
            logger({
                category: "groq",
                message: "error creating chat completion",
                level: 0,
                auxiliary: {
                    error: {
                        value: error.message,
                        type: "string",
                    },
                    requestId: {
                        value: options.requestId,
                        type: "string",
                    },
                },
            });
            throw error;
        }
    }
}
exports.GroqClient = GroqClient;
