// @browserbasehq/stagehand/dist/cjs/lib/v3/llm/OpenAIClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAIClient = void 0;
const openai_1 = __importDefault(require("openai"));
const utils_js_1 = require("../../utils.js");
const LLMClient_js_1 = require("./LLMClient.js");
const sdkErrors_js_1 = require("../types/public/sdkErrors.js");
const zodCompat_js_1 = require("../zodCompat.js");
class OpenAIClient extends LLMClient_js_1.LLMClient {
    type = "openai";
    client;
    constructor({ modelName, clientOptions, }) {
        super(modelName);
        this.clientOptions = clientOptions;
        this.client = new openai_1.default(clientOptions);
        this.modelName = modelName;
    }
    async createChatCompletion({ options: optionsInitial, logger, retries = 3, }) {
        let options = optionsInitial;
        // O1 models do not support most of the options. So we override them.
        // For schema and tools, we add them as user messages.
        let isToolsOverridedForO1 = false;
        if (this.modelName.startsWith("o1") || this.modelName.startsWith("o3")) {
            /* eslint-disable */
            // Remove unsupported options
            let { tool_choice, top_p, frequency_penalty, presence_penalty, temperature, } = options;
            ({
                tool_choice,
                top_p,
                frequency_penalty,
                presence_penalty,
                temperature,
                ...options
            } = options);
            /* eslint-enable */
            // Remove unsupported options
            options.messages = options.messages.map((message) => ({
                ...message,
                role: "user",
            }));
            if (options.tools && options.response_model) {
                throw new sdkErrors_js_1.StagehandError("Cannot use both tool and response_model for o1 models");
            }
            if (options.tools) {
                // Remove unsupported options
                const { tools, ...rest } = options;
                options = rest;
                isToolsOverridedForO1 = true;
                options.messages.push({
                    role: "user",
                    content: `You have the following tools available to you:\n${JSON.stringify(tools)}

          Respond with the following zod schema format to use a method: {
            "name": "<tool_name>",
            "arguments": <tool_args>
          }
          
          Do not include any other text or formattings like \`\`\` in your response. Just the JSON object.`,
                });
            }
        }
        if (options.temperature &&
            (this.modelName.startsWith("o1") || this.modelName.startsWith("o3"))) {
            throw new sdkErrors_js_1.StagehandError("Temperature is not supported for o1 models");
        }
        const { requestId, ...optionsWithoutImageAndRequestId } = options;
        logger({
            category: "openai",
            message: "creating chat completion",
            level: 2,
            auxiliary: {
                options: {
                    value: JSON.stringify({
                        ...optionsWithoutImageAndRequestId,
                        requestId,
                    }),
                    type: "object",
                },
                modelName: {
                    value: this.modelName,
                    type: "string",
                },
            },
        });
        if (options.image) {
            const screenshotMessage = {
                role: "user",
                content: [
                    {
                        type: "image_url",
                        image_url: {
                            url: `data:image/jpeg;base64,${options.image.buffer.toString("base64")}`,
                        },
                    },
                    ...(options.image.description
                        ? [{ type: "text", text: options.image.description }]
                        : []),
                ],
            };
            options.messages.push(screenshotMessage);
        }
        let responseFormat;
        if (options.response_model) {
            // For O1 models, we need to add the schema as a user message.
            if (this.modelName.startsWith("o1") || this.modelName.startsWith("o3")) {
                try {
                    const parsedSchema = JSON.stringify((0, zodCompat_js_1.toJsonSchema)(options.response_model.schema));
                    options.messages.push({
                        role: "user",
                        content: `Respond in this zod schema format:\n${parsedSchema}\n

          Do not include any other text, formatting or markdown in your output. Do not include \`\`\` or \`\`\`json in your response. Only the JSON object itself.`,
                    });
                }
                catch (error) {
                    logger({
                        category: "openai",
                        message: "Failed to parse response model schema",
                        level: 0,
                    });
                    if (retries > 0) {
                        // as-casting to account for o1 models not supporting all options
                        return this.createChatCompletion({
                            options: options,
                            logger,
                            retries: retries - 1,
                        });
                    }
                    throw error;
                }
            }
            else {
                responseFormat = {
                    type: "json_schema",
                    json_schema: {
                        name: options.response_model.name,
                        schema: (0, zodCompat_js_1.toJsonSchema)(options.response_model.schema),
                    },
                };
            }
        }
        /* eslint-disable */
        // Remove unsupported options
        const { response_model, ...openAiOptions } = {
            ...optionsWithoutImageAndRequestId,
            model: this.modelName,
        };
        /* eslint-enable */
        logger({
            category: "openai",
            message: "creating chat completion",
            level: 2,
            auxiliary: {
                openAiOptions: {
                    value: JSON.stringify(openAiOptions),
                    type: "object",
                },
            },
        });
        const formattedMessages = options.messages.map((message) => {
            if (Array.isArray(message.content)) {
                const contentParts = message.content.map((content) => {
                    if ("image_url" in content) {
                        const imageContent = {
                            image_url: {
                                url: content.image_url.url,
                            },
                            type: "image_url",
                        };
                        return imageContent;
                    }
                    else {
                        const textContent = {
                            text: content.text,
                            type: "text",
                        };
                        return textContent;
                    }
                });
                if (message.role === "system") {
                    const formattedMessage = {
                        ...message,
                        role: "system",
                        content: contentParts.filter((content) => content.type === "text"),
                    };
                    return formattedMessage;
                }
                else if (message.role === "user") {
                    const formattedMessage = {
                        ...message,
                        role: "user",
                        content: contentParts,
                    };
                    return formattedMessage;
                }
                else {
                    const formattedMessage = {
                        ...message,
                        role: "assistant",
                        content: contentParts.filter((content) => content.type === "text"),
                    };
                    return formattedMessage;
                }
            }
            const formattedMessage = {
                role: "user",
                content: message.content,
            };
            return formattedMessage;
        });
        const body = {
            ...openAiOptions,
            model: this.modelName,
            messages: formattedMessages,
            response_format: responseFormat,
            stream: false,
            tools: options.tools?.map((tool) => ({
                function: {
                    name: tool.name,
                    description: tool.description,
                    parameters: tool.parameters,
                },
                type: "function",
            })),
        };
        const response = await this.client.chat.completions.create(body);
        // For O1 models, we need to parse the tool call response manually and add it to the response.
        if (isToolsOverridedForO1) {
            try {
                const parsedContent = JSON.parse(response.choices[0].message.content);
                response.choices[0].message.tool_calls = [
                    {
                        function: {
                            name: parsedContent["name"],
                            arguments: JSON.stringify(parsedContent["arguments"]),
                        },
                        type: "function",
                        id: "-1",
                    },
                ];
                response.choices[0].message.content = null;
            }
            catch (error) {
                logger({
                    category: "openai",
                    message: "Failed to parse tool call response",
                    level: 0,
                    auxiliary: {
                        error: {
                            value: error.message,
                            type: "string",
                        },
                        content: {
                            value: response.choices[0].message.content,
                            type: "string",
                        },
                    },
                });
                if (retries > 0) {
                    // as-casting to account for o1 models not supporting all options
                    return this.createChatCompletion({
                        options: options,
                        logger,
                        retries: retries - 1,
                    });
                }
                throw error;
            }
        }
        logger({
            category: "openai",
            message: "response",
            level: 2,
            auxiliary: {
                response: {
                    value: JSON.stringify(response),
                    type: "object",
                },
                requestId: {
                    value: requestId,
                    type: "string",
                },
            },
        });
        if (options.response_model) {
            const extractedData = response.choices[0].message.content;
            const parsedData = JSON.parse(extractedData);
            try {
                (0, utils_js_1.validateZodSchema)(options.response_model.schema, parsedData);
            }
            catch (e) {
                logger({
                    category: "openai",
                    message: "Response failed Zod schema validation",
                    level: 0,
                });
                if (retries > 0) {
                    // as-casting to account for o1 models not supporting all options
                    return this.createChatCompletion({
                        options: options,
                        logger,
                        retries: retries - 1,
                    });
                }
                if (e instanceof sdkErrors_js_1.ZodSchemaValidationError) {
                    logger({
                        category: "openai",
                        message: `Error during OpenAI chat completion: ${e.message}`,
                        level: 0,
                        auxiliary: {
                            errorDetails: {
                                value: `Message: ${e.message}${e.stack ? "\nStack: " + e.stack : ""}`,
                                type: "string",
                            },
                            requestId: { value: requestId, type: "string" },
                        },
                    });
                    throw new sdkErrors_js_1.CreateChatCompletionResponseError(e.message);
                }
                throw e;
            }
            return {
                data: parsedData,
                usage: response.usage,
            };
        }
        // if the function was called with a response model, it would have returned earlier
        // so we can safely cast here to T, which defaults to ChatCompletion
        return response;
    }
}
exports.OpenAIClient = OpenAIClient;
