// @browserbasehq/stagehand/dist/cjs/lib/inferenceLogUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendSummary = appendSummary;
exports.writeTimestampedTxtFile = writeTimestampedTxtFile;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
/**
 * Create (or ensure) a parent directory named "inference_summary".
 */
function ensureInferenceSummaryDir() {
    const inferenceDir = path_1.default.join(process.cwd(), "inference_summary");
    if (!fs_1.default.existsSync(inferenceDir)) {
        fs_1.default.mkdirSync(inferenceDir, { recursive: true });
    }
    return inferenceDir;
}
/**
 * Appends a new entry to the act_summary.json file, then writes the file back out.
 */
function appendSummary(inferenceType, entry) {
    const summaryPath = getSummaryJsonPath(inferenceType);
    const arrayKey = `${inferenceType}_summary`;
    const existingData = readSummaryFile(inferenceType);
    existingData[arrayKey].push(entry);
    fs_1.default.writeFileSync(summaryPath, JSON.stringify(existingData, null, 2));
}
/** A simple timestamp utility for filenames. */
function getTimestamp() {
    return new Date()
        .toISOString()
        .replace(/[^0-9T]/g, "")
        .replace("T", "_");
}
/**
 * Writes `data` as JSON into a file in `directory`, using a prefix plus timestamp.
 * Returns both the file name and the timestamp used, so you can log them.
 */
function writeTimestampedTxtFile(directory, prefix, data) {
    const baseDir = ensureInferenceSummaryDir();
    const subDir = path_1.default.join(baseDir, directory);
    if (!fs_1.default.existsSync(subDir)) {
        fs_1.default.mkdirSync(subDir, { recursive: true });
    }
    const timestamp = getTimestamp();
    const fileName = `${timestamp}_${prefix}.txt`;
    const filePath = path_1.default.join(subDir, fileName);
    fs_1.default.writeFileSync(filePath, JSON.stringify(data, null, 2).replace(/\\n/g, "\n"));
    return { fileName, timestamp };
}
/**
 * Returns the path to the `<inferenceType>_summary.json` file.
 *
 * For example, if `inferenceType = "act"`, this will be:
 *   `./inference_summary/act_summary/act_summary.json`
 */
function getSummaryJsonPath(inferenceType) {
    const baseDir = ensureInferenceSummaryDir();
    const subDir = path_1.default.join(baseDir, `${inferenceType}_summary`);
    if (!fs_1.default.existsSync(subDir)) {
        fs_1.default.mkdirSync(subDir, { recursive: true });
    }
    return path_1.default.join(subDir, `${inferenceType}_summary.json`);
}
/**
 * Reads the `<inferenceType>_summary.json` file, returning an object
 * with the top-level array named `<inferenceType>_summary`, if it exists.
 *
 * E.g. if inferenceType is "act", we expect a shape like:
 * {
 *   "act_summary": [ ... ]
 * }
 *
 * If the file or array is missing, returns { "<inferenceType>_summary": [] }.
 */
function readSummaryFile(inferenceType) {
    const summaryPath = getSummaryJsonPath(inferenceType);
    // The top-level array key, e.g. "act_summary", "observe_summary", "extract_summary"
    const arrayKey = `${inferenceType}_summary`;
    if (!fs_1.default.existsSync(summaryPath)) {
        return { [arrayKey]: [] };
    }
    try {
        const raw = fs_1.default.readFileSync(summaryPath, "utf8");
        const parsed = JSON.parse(raw);
        if (parsed &&
            typeof parsed === "object" &&
            Array.isArray(parsed[arrayKey])) {
            return parsed;
        }
    }
    catch {
        // If we fail to parse for any reason, fall back to empty array
    }
    return { [arrayKey]: [] };
}
