// @ai-sdk/google-vertex/dist/maas/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/maas/index.ts
var index_exports = {};
__export(index_exports, {
  createVertexMaas: () => createVertexMaas2,
  vertexMaas: () => vertexMaas
});
module.exports = __toCommonJS(index_exports);

// src/maas/google-vertex-maas-provider-node.ts
var import_provider_utils2 = require("@ai-sdk/provider-utils");

// src/google-vertex-auth-google-auth-library.ts
var import_google_auth_library = require("google-auth-library");
var authInstance = null;
var authOptions = null;
function getAuth(options) {
  if (!authInstance || options !== authOptions) {
    authInstance = new import_google_auth_library.GoogleAuth({
      scopes: ["https://www.googleapis.com/auth/cloud-platform"],
      ...options
    });
    authOptions = options;
  }
  return authInstance;
}
async function generateAuthToken(options) {
  const auth = getAuth(options || {});
  const client = await auth.getClient();
  const token = await client.getAccessToken();
  return (token == null ? void 0 : token.token) || null;
}

// src/maas/google-vertex-maas-provider.ts
var import_openai_compatible = require("@ai-sdk/openai-compatible");
var import_provider_utils = require("@ai-sdk/provider-utils");
function createVertexMaas(options = {}) {
  const loadLocation = () => (0, import_provider_utils.loadOptionalSetting)({
    settingValue: options.location,
    environmentVariableName: "GOOGLE_VERTEX_LOCATION"
  });
  const loadProject = () => (0, import_provider_utils.loadSetting)({
    settingValue: options.project,
    settingName: "project",
    environmentVariableName: "GOOGLE_VERTEX_PROJECT",
    description: "Google Vertex project"
  });
  const constructBaseURL = () => {
    var _a;
    const projectId = loadProject();
    const location = (_a = loadLocation()) != null ? _a : "global";
    return `https://aiplatform.googleapis.com/v1/projects/${projectId}/locations/${location}/endpoints/openapi`;
  };
  const loadBaseURL = () => {
    var _a;
    return (0, import_provider_utils.withoutTrailingSlash)((_a = options.baseURL) != null ? _a : "") || constructBaseURL();
  };
  let cachedProvider;
  const getProvider = () => cachedProvider != null ? cachedProvider : cachedProvider = (0, import_openai_compatible.createOpenAICompatible)({
    name: "vertex.maas",
    baseURL: loadBaseURL(),
    fetch: options.fetch
  });
  const provider = (modelId) => getProvider()(modelId);
  provider.specificationVersion = "v2";
  provider.languageModel = (modelId) => getProvider().languageModel(modelId);
  provider.chatModel = (modelId) => getProvider().chatModel(modelId);
  provider.completionModel = (modelId) => getProvider().completionModel(modelId);
  provider.textEmbeddingModel = (modelId) => getProvider().textEmbeddingModel(modelId);
  provider.imageModel = (modelId) => getProvider().imageModel(modelId);
  return provider;
}

// src/maas/google-vertex-maas-provider-node.ts
function createVertexMaas2(options = {}) {
  const customFetch = async (url, init) => {
    var _a;
    const token = await generateAuthToken(options.googleAuthOptions);
    const resolvedHeaders = await (0, import_provider_utils2.resolve)(options.headers);
    const authHeaders = {
      ...resolvedHeaders,
      Authorization: `Bearer ${token}`
    };
    const fetchInit = {
      ...init,
      headers: {
        ...init == null ? void 0 : init.headers,
        ...authHeaders
      }
    };
    return ((_a = options.fetch) != null ? _a : fetch)(url, fetchInit);
  };
  return createVertexMaas({
    ...options,
    fetch: customFetch,
    headers: void 0
    // Don't pass headers, we handle them in fetch
  });
}
var vertexMaas = createVertexMaas2();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createVertexMaas,
  vertexMaas
});
