// @ai-sdk/google-vertex/dist/maas/index.mjs
// src/maas/google-vertex-maas-provider-node.ts
import { resolve } from "@ai-sdk/provider-utils";

// src/google-vertex-auth-google-auth-library.ts
import { GoogleAuth } from "google-auth-library";
var authInstance = null;
var authOptions = null;
function getAuth(options) {
  if (!authInstance || options !== authOptions) {
    authInstance = new GoogleAuth({
      scopes: ["https://www.googleapis.com/auth/cloud-platform"],
      ...options
    });
    authOptions = options;
  }
  return authInstance;
}
async function generateAuthToken(options) {
  const auth = getAuth(options || {});
  const client = await auth.getClient();
  const token = await client.getAccessToken();
  return (token == null ? void 0 : token.token) || null;
}

// src/maas/google-vertex-maas-provider.ts
import { createOpenAICompatible } from "@ai-sdk/openai-compatible";
import {
  loadOptionalSetting,
  loadSetting,
  withoutTrailingSlash
} from "@ai-sdk/provider-utils";
function createVertexMaas(options = {}) {
  const loadLocation = () => loadOptionalSetting({
    settingValue: options.location,
    environmentVariableName: "GOOGLE_VERTEX_LOCATION"
  });
  const loadProject = () => loadSetting({
    settingValue: options.project,
    settingName: "project",
    environmentVariableName: "GOOGLE_VERTEX_PROJECT",
    description: "Google Vertex project"
  });
  const constructBaseURL = () => {
    var _a;
    const projectId = loadProject();
    const location = (_a = loadLocation()) != null ? _a : "global";
    return `https://aiplatform.googleapis.com/v1/projects/${projectId}/locations/${location}/endpoints/openapi`;
  };
  const loadBaseURL = () => {
    var _a;
    return withoutTrailingSlash((_a = options.baseURL) != null ? _a : "") || constructBaseURL();
  };
  let cachedProvider;
  const getProvider = () => cachedProvider != null ? cachedProvider : cachedProvider = createOpenAICompatible({
    name: "vertex.maas",
    baseURL: loadBaseURL(),
    fetch: options.fetch
  });
  const provider = (modelId) => getProvider()(modelId);
  provider.specificationVersion = "v2";
  provider.languageModel = (modelId) => getProvider().languageModel(modelId);
  provider.chatModel = (modelId) => getProvider().chatModel(modelId);
  provider.completionModel = (modelId) => getProvider().completionModel(modelId);
  provider.textEmbeddingModel = (modelId) => getProvider().textEmbeddingModel(modelId);
  provider.imageModel = (modelId) => getProvider().imageModel(modelId);
  return provider;
}

// src/maas/google-vertex-maas-provider-node.ts
function createVertexMaas2(options = {}) {
  const customFetch = async (url, init) => {
    var _a;
    const token = await generateAuthToken(options.googleAuthOptions);
    const resolvedHeaders = await resolve(options.headers);
    const authHeaders = {
      ...resolvedHeaders,
      Authorization: `Bearer ${token}`
    };
    const fetchInit = {
      ...init,
      headers: {
        ...init == null ? void 0 : init.headers,
        ...authHeaders
      }
    };
    return ((_a = options.fetch) != null ? _a : fetch)(url, fetchInit);
  };
  return createVertexMaas({
    ...options,
    fetch: customFetch,
    headers: void 0
    // Don't pass headers, we handle them in fetch
  });
}
var vertexMaas = createVertexMaas2();
export {
  createVertexMaas2 as createVertexMaas,
  vertexMaas
};
