// @browserbasehq/stagehand/dist/cjs/lib/v3/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.__internalMaybeRunShutdownSupervisorFromArgv = exports.__internalCreateInMemoryAgentCacheHandle = exports.getAISDKLanguageModel = exports.tool = exports.V3Evaluator = exports.connectToMCPServer = exports.toJsonSchema = exports.isZod3Schema = exports.isZod4Schema = exports.jsonSchemaToZod = exports.trimTrailingTextNode = exports.loadApiKeyFromEnv = exports.providerEnvVarMap = exports.injectUrls = exports.transformSchema = exports.getZodType = exports.toGeminiSchema = exports.isRunningInBun = exports.validateZodSchema = exports.modelToAgentProviderMap = exports.AgentProvider = exports.LLMClient = exports.AnnotatedScreenshotText = exports.Stagehand = exports.V3 = void 0;
const PublicApi = __importStar(require("./types/public/index.js"));
const v3_js_1 = require("./v3.js");
const LLMClient_js_1 = require("./llm/LLMClient.js");
const AgentProvider_js_1 = require("./agent/AgentProvider.js");
const utils_js_1 = require("../utils.js");
const zodCompat_js_1 = require("./zodCompat.js");
const connection_js_1 = require("./mcp/connection.js");
const v3Evaluator_js_1 = require("../v3Evaluator.js");
const ai_1 = require("ai");
const LLMProvider_js_1 = require("./llm/LLMProvider.js");
const serverAgentCache_js_1 = require("./cache/serverAgentCache.js");
const supervisor_js_1 = require("./shutdown/supervisor.js");
var v3_js_2 = require("./v3.js");
Object.defineProperty(exports, "V3", { enumerable: true, get: function () { return v3_js_2.V3; } });
var v3_js_3 = require("./v3.js");
Object.defineProperty(exports, "Stagehand", { enumerable: true, get: function () { return v3_js_3.V3; } });
__exportStar(require("./types/public/index.js"), exports);
var LLMClient_js_2 = require("./llm/LLMClient.js");
Object.defineProperty(exports, "AnnotatedScreenshotText", { enumerable: true, get: function () { return LLMClient_js_2.AnnotatedScreenshotText; } });
Object.defineProperty(exports, "LLMClient", { enumerable: true, get: function () { return LLMClient_js_2.LLMClient; } });
var AgentProvider_js_2 = require("./agent/AgentProvider.js");
Object.defineProperty(exports, "AgentProvider", { enumerable: true, get: function () { return AgentProvider_js_2.AgentProvider; } });
Object.defineProperty(exports, "modelToAgentProviderMap", { enumerable: true, get: function () { return AgentProvider_js_2.modelToAgentProviderMap; } });
var utils_js_2 = require("../utils.js");
Object.defineProperty(exports, "validateZodSchema", { enumerable: true, get: function () { return utils_js_2.validateZodSchema; } });
Object.defineProperty(exports, "isRunningInBun", { enumerable: true, get: function () { return utils_js_2.isRunningInBun; } });
Object.defineProperty(exports, "toGeminiSchema", { enumerable: true, get: function () { return utils_js_2.toGeminiSchema; } });
Object.defineProperty(exports, "getZodType", { enumerable: true, get: function () { return utils_js_2.getZodType; } });
Object.defineProperty(exports, "transformSchema", { enumerable: true, get: function () { return utils_js_2.transformSchema; } });
Object.defineProperty(exports, "injectUrls", { enumerable: true, get: function () { return utils_js_2.injectUrls; } });
Object.defineProperty(exports, "providerEnvVarMap", { enumerable: true, get: function () { return utils_js_2.providerEnvVarMap; } });
Object.defineProperty(exports, "loadApiKeyFromEnv", { enumerable: true, get: function () { return utils_js_2.loadApiKeyFromEnv; } });
Object.defineProperty(exports, "trimTrailingTextNode", { enumerable: true, get: function () { return utils_js_2.trimTrailingTextNode; } });
Object.defineProperty(exports, "jsonSchemaToZod", { enumerable: true, get: function () { return utils_js_2.jsonSchemaToZod; } });
var zodCompat_js_2 = require("./zodCompat.js");
Object.defineProperty(exports, "isZod4Schema", { enumerable: true, get: function () { return zodCompat_js_2.isZod4Schema; } });
Object.defineProperty(exports, "isZod3Schema", { enumerable: true, get: function () { return zodCompat_js_2.isZod3Schema; } });
Object.defineProperty(exports, "toJsonSchema", { enumerable: true, get: function () { return zodCompat_js_2.toJsonSchema; } });
var connection_js_2 = require("./mcp/connection.js");
Object.defineProperty(exports, "connectToMCPServer", { enumerable: true, get: function () { return connection_js_2.connectToMCPServer; } });
var v3Evaluator_js_2 = require("../v3Evaluator.js");
Object.defineProperty(exports, "V3Evaluator", { enumerable: true, get: function () { return v3Evaluator_js_2.V3Evaluator; } });
var ai_2 = require("ai");
Object.defineProperty(exports, "tool", { enumerable: true, get: function () { return ai_2.tool; } });
var LLMProvider_js_2 = require("./llm/LLMProvider.js");
Object.defineProperty(exports, "getAISDKLanguageModel", { enumerable: true, get: function () { return LLMProvider_js_2.getAISDKLanguageModel; } });
var serverAgentCache_js_2 = require("./cache/serverAgentCache.js");
Object.defineProperty(exports, "__internalCreateInMemoryAgentCacheHandle", { enumerable: true, get: function () { return serverAgentCache_js_2.__internalCreateInMemoryAgentCacheHandle; } });
var supervisor_js_2 = require("./shutdown/supervisor.js");
Object.defineProperty(exports, "__internalMaybeRunShutdownSupervisorFromArgv", { enumerable: true, get: function () { return supervisor_js_2.maybeRunShutdownSupervisorFromArgv; } });
const StagehandDefault = {
    ...PublicApi,
    V3: v3_js_1.V3,
    Stagehand: v3_js_1.V3,
    AnnotatedScreenshotText: LLMClient_js_1.AnnotatedScreenshotText,
    LLMClient: LLMClient_js_1.LLMClient,
    AgentProvider: AgentProvider_js_1.AgentProvider,
    modelToAgentProviderMap: AgentProvider_js_1.modelToAgentProviderMap,
    validateZodSchema: utils_js_1.validateZodSchema,
    isRunningInBun: utils_js_1.isRunningInBun,
    toGeminiSchema: utils_js_1.toGeminiSchema,
    getZodType: utils_js_1.getZodType,
    transformSchema: utils_js_1.transformSchema,
    injectUrls: utils_js_1.injectUrls,
    providerEnvVarMap: utils_js_1.providerEnvVarMap,
    loadApiKeyFromEnv: utils_js_1.loadApiKeyFromEnv,
    trimTrailingTextNode: utils_js_1.trimTrailingTextNode,
    jsonSchemaToZod: utils_js_1.jsonSchemaToZod,
    isZod4Schema: zodCompat_js_1.isZod4Schema,
    isZod3Schema: zodCompat_js_1.isZod3Schema,
    toJsonSchema: zodCompat_js_1.toJsonSchema,
    connectToMCPServer: connection_js_1.connectToMCPServer,
    V3Evaluator: v3Evaluator_js_1.V3Evaluator,
    tool: ai_1.tool,
    getAISDKLanguageModel: LLMProvider_js_1.getAISDKLanguageModel,
    __internalCreateInMemoryAgentCacheHandle: serverAgentCache_js_1.__internalCreateInMemoryAgentCacheHandle,
    __internalMaybeRunShutdownSupervisorFromArgv: supervisor_js_1.maybeRunShutdownSupervisorFromArgv,
};
exports.default = StagehandDefault;
