// @browserbasehq/stagehand/dist/cjs/lib/v3/runtimePaths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMainModule = exports.createRequireFromCaller = exports.getPackageRootDir = exports.getRepoRootDir = exports.getCurrentDirPath = exports.getCurrentFilePath = void 0;
/**
 * Keep this file in sync with:
 * - /packages/core/lib/v3/runtimePaths.ts
 * - /packages/server-v3/scripts/runtimePaths.ts
 * - /packages/server-v4/scripts/runtimePaths.ts
 * - /packages/evals/runtimePaths.ts
 * - /packages/docs/scripts/runtimePaths.js
 */
const node_path_1 = __importDefault(require("node:path"));
const node_url_1 = require("node:url");
const node_module_1 = require("node:module");
const PACKAGE_SEGMENT = "/packages/core/";
const EVAL_FRAMES = new Set(["[eval]", "[eval]-wrapper"]);
const INTERNAL_FRAME_NAMES = new Set([
    "readCallsites",
    "readCallsitePath",
    "resolveCallerFilePath",
    "getCurrentFilePath",
    "getCurrentDirPath",
    "getRepoRootDir",
    "getPackageRootDir",
    "createRequireFromCaller",
    "isMainModule",
]);
const normalizePath = (value) => {
    const input = value.startsWith("file://") ? (0, node_url_1.fileURLToPath)(value) : value;
    return node_path_1.default.resolve(input).replaceAll("\\", "/");
};
const readCallsites = () => {
    const previousPrepare = Error.prepareStackTrace;
    try {
        Error.prepareStackTrace = (_, stack) => stack;
        return (new Error().stack ?? []);
    }
    finally {
        Error.prepareStackTrace = previousPrepare;
    }
};
const readCallsitePath = (callsite) => {
    const callsiteWithScript = callsite;
    const rawPath = callsite.getFileName() ?? callsiteWithScript.getScriptNameOrSourceURL?.();
    if (!rawPath)
        return null;
    if (rawPath.startsWith("node:"))
        return null;
    if (EVAL_FRAMES.has(rawPath))
        return null;
    return normalizePath(rawPath);
};
const isInternalCallsite = (callsite) => {
    const functionName = callsite.getFunctionName();
    if (functionName && INTERNAL_FRAME_NAMES.has(functionName))
        return true;
    const methodName = callsite.getMethodName();
    if (methodName && INTERNAL_FRAME_NAMES.has(methodName))
        return true;
    const callsiteString = callsite.toString();
    for (const frameName of INTERNAL_FRAME_NAMES) {
        if (callsiteString.includes(`${frameName} (`))
            return true;
        if (callsiteString.includes(`.${frameName} (`))
            return true;
    }
    return false;
};
const resolveCallerFilePath = () => {
    const packageCandidates = [];
    const fallbackCandidates = [];
    for (const callsite of readCallsites()) {
        const filePath = readCallsitePath(callsite);
        if (!filePath)
            continue;
        if (isInternalCallsite(callsite))
            continue;
        if (filePath.includes(PACKAGE_SEGMENT)) {
            packageCandidates.push(filePath);
            continue;
        }
        fallbackCandidates.push(filePath);
    }
    const packageCandidate = packageCandidates[0];
    if (packageCandidate)
        return packageCandidate;
    const fallbackCandidate = fallbackCandidates[0];
    if (fallbackCandidate)
        return fallbackCandidate;
    throw new Error("Unable to resolve caller file path.");
};
const getCurrentFilePath = () => resolveCallerFilePath();
exports.getCurrentFilePath = getCurrentFilePath;
const getCurrentDirPath = () => node_path_1.default.dirname((0, exports.getCurrentFilePath)());
exports.getCurrentDirPath = getCurrentDirPath;
const getRepoRootDir = () => {
    const currentFilePath = (0, exports.getCurrentFilePath)();
    const index = currentFilePath.lastIndexOf(PACKAGE_SEGMENT);
    if (index === -1) {
        throw new Error(`Unable to determine repo root from ${currentFilePath} (missing ${PACKAGE_SEGMENT}).`);
    }
    return currentFilePath.slice(0, index);
};
exports.getRepoRootDir = getRepoRootDir;
const getPackageRootDir = () => `${(0, exports.getRepoRootDir)()}${PACKAGE_SEGMENT.slice(0, -1)}`;
exports.getPackageRootDir = getPackageRootDir;
const createRequireFromCaller = () => (0, node_module_1.createRequire)((0, exports.getCurrentFilePath)());
exports.createRequireFromCaller = createRequireFromCaller;
const isMainModule = () => {
    const entryScript = process.argv.at(1);
    if (!entryScript)
        return false;
    return normalizePath(entryScript) === (0, exports.getCurrentFilePath)();
};
exports.isMainModule = isMainModule;
