// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/a11y/snapshot/activeElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeActiveElementXpath = computeActiveElementXpath;
const executionContextRegistry_js_1 = require("../../executionContextRegistry.js");
const a11yInvocation_js_1 = require("../../a11yInvocation.js");
const a11yScripts_generated_js_1 = require("../../../dom/build/a11yScripts.generated.js");
const xpathUtils_js_1 = require("./xpathUtils.js");
/**
 * Compute the absolute XPath for the currently focused element.
 * - Detects which frame has focus via document.hasFocus().
 * - Finds the deepest activeElement (dives into shadow DOM).
 * - Builds an absolute, cross-frame XPath by prefixing iframe hosts.
 */
async function computeActiveElementXpath(page) {
    const tree = page.getFullFrameTree();
    const parentByFrame = new Map();
    (function index(n, parent) {
        parentByFrame.set(n.frame.id, parent);
        for (const c of n.childFrames ?? [])
            index(c, n.frame.id);
    })(tree, null);
    const frames = page.listAllFrameIds();
    let focusedFrameId = null;
    for (const fid of frames) {
        const sess = page.getSessionForFrame(fid);
        try {
            await sess.send("Runtime.enable").catch(() => { });
            const ctxId = await executionContextRegistry_js_1.executionContexts
                .waitForMainWorld(sess, fid, 1000)
                .catch(() => { });
            const hasFocusExpr = (0, a11yInvocation_js_1.buildA11yInvocation)("documentHasFocusStrict", []);
            const evalParams = ctxId
                ? {
                    contextId: ctxId,
                    expression: hasFocusExpr,
                    returnByValue: true,
                }
                : { expression: hasFocusExpr, returnByValue: true };
            const { result } = await sess.send("Runtime.evaluate", evalParams);
            if (result?.value === true) {
                focusedFrameId = fid;
                break;
            }
        }
        catch {
            //
        }
    }
    if (!focusedFrameId)
        focusedFrameId = page.mainFrameId();
    const focusedSession = page.getSessionForFrame(focusedFrameId);
    let objectId;
    try {
        await focusedSession.send("Runtime.enable").catch(() => { });
        const ctxId = await executionContextRegistry_js_1.executionContexts
            .waitForMainWorld(focusedSession, focusedFrameId, 1000)
            .catch(() => { });
        const activeExpr = (0, a11yInvocation_js_1.buildA11yInvocation)("resolveDeepActiveElement", []);
        const evalParams = ctxId
            ? {
                contextId: ctxId,
                expression: activeExpr,
                returnByValue: false,
            }
            : { expression: activeExpr, returnByValue: false };
        const { result } = await focusedSession.send("Runtime.evaluate", evalParams);
        objectId = result?.objectId;
    }
    catch {
        objectId = undefined;
    }
    if (!objectId)
        return null;
    const leafXPath = await (async () => {
        try {
            const { result } = await focusedSession.send("Runtime.callFunctionOn", {
                objectId,
                functionDeclaration: a11yScripts_generated_js_1.a11yScriptSources.nodeToAbsoluteXPath,
                returnByValue: true,
            });
            try {
                await focusedSession.send("Runtime.releaseObject", { objectId });
            }
            catch {
                //
            }
            const xp = result?.value || "";
            return typeof xp === "string" && xp ? xp : null;
        }
        catch {
            try {
                await focusedSession.send("Runtime.releaseObject", { objectId });
            }
            catch {
                //
            }
            return null;
        }
    })();
    if (!leafXPath)
        return null;
    let prefix = "";
    let cur = focusedFrameId;
    while (cur) {
        const parent = parentByFrame.get(cur) ?? null;
        if (!parent)
            break;
        const parentSess = page.getSessionForFrame(parent);
        try {
            const { backendNodeId } = await parentSess.send("DOM.getFrameOwner", { frameId: cur });
            if (typeof backendNodeId === "number") {
                const xp = await (0, xpathUtils_js_1.absoluteXPathForBackendNode)(parentSess, backendNodeId);
                if (xp)
                    prefix = prefix ? (0, xpathUtils_js_1.prefixXPath)(prefix, xp) : (0, xpathUtils_js_1.normalizeXPath)(xp);
            }
        }
        catch {
            //
        }
        cur = parent;
    }
    return prefix ? (0, xpathUtils_js_1.prefixXPath)(prefix, leafXPath) : (0, xpathUtils_js_1.normalizeXPath)(leafXPath);
}
