// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/a11y/snapshot/coordinateResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveXpathForLocation = resolveXpathForLocation;
const executionContextRegistry_js_1 = require("../../executionContextRegistry.js");
const a11yScripts_generated_js_1 = require("../../../dom/build/a11yScripts.generated.js");
const a11yInvocation_js_1 = require("../../a11yInvocation.js");
const focusSelectors_js_1 = require("./focusSelectors.js");
const xpathUtils_js_1 = require("./xpathUtils.js");
/**
 * Resolve deepest node for a page coordinate and compute its absolute XPath across frames.
 * More efficient than building a full hybrid snapshot when only a single node’s XPath is needed.
 */
async function resolveXpathForLocation(page, x, y) {
    const tree = page.getFullFrameTree();
    const parentByFrame = new Map();
    (function index(n, parent) {
        parentByFrame.set(n.frame.id, parent);
        for (const c of n.childFrames ?? [])
            index(c, n.frame.id);
    })(tree, null);
    const iframeChain = [];
    let curFrameId = page.mainFrameId();
    let curSession = page.getSessionForFrame(curFrameId);
    let curX = x;
    let curY = y;
    for (let depth = 0; depth < 8; depth++) {
        try {
            await curSession.send("DOM.enable").catch(() => { });
            let sx = 0;
            let sy = 0;
            try {
                await curSession.send("Runtime.enable").catch(() => { });
                const ctxId = await executionContextRegistry_js_1.executionContexts
                    .waitForMainWorld(curSession, curFrameId)
                    .catch(() => { });
                const scrollExpr = (0, a11yInvocation_js_1.buildA11yInvocation)("getScrollOffsets", []);
                const evalParams = ctxId
                    ? {
                        contextId: ctxId,
                        expression: scrollExpr,
                        returnByValue: true,
                    }
                    : { expression: scrollExpr, returnByValue: true };
                const { result } = await curSession.send("Runtime.evaluate", evalParams);
                sx = Number(result?.value?.sx ?? 0);
                sy = Number(result?.value?.sy ?? 0);
            }
            catch {
                //
            }
            const xi = Math.max(0, Math.floor(curX + sx));
            const yi = Math.max(0, Math.floor(curY + sy));
            let res;
            try {
                res = await curSession.send("DOM.getNodeForLocation", {
                    x: xi,
                    y: yi,
                    includeUserAgentShadowDOM: false,
                    ignorePointerEventsNone: false,
                });
            }
            catch {
                return null;
            }
            const be = res?.backendNodeId;
            const reportedFrameId = res?.frameId;
            if (typeof be === "number" &&
                reportedFrameId &&
                reportedFrameId !== curFrameId) {
                const abs = await (0, xpathUtils_js_1.buildAbsoluteXPathFromChain)(iframeChain, curSession, be);
                return abs
                    ? { frameId: reportedFrameId, backendNodeId: be, absoluteXPath: abs }
                    : null;
            }
            if (typeof be !== "number")
                return null;
            let matchedChild;
            for (const fid of (0, focusSelectors_js_1.listChildrenOf)(parentByFrame, curFrameId)) {
                try {
                    const { backendNodeId } = await curSession.send("DOM.getFrameOwner", { frameId: fid });
                    if (backendNodeId === be) {
                        matchedChild = fid;
                        break;
                    }
                }
                catch {
                    continue;
                }
            }
            if (!matchedChild) {
                const abs = await (0, xpathUtils_js_1.buildAbsoluteXPathFromChain)(iframeChain, curSession, be);
                return abs
                    ? { frameId: curFrameId, backendNodeId: be, absoluteXPath: abs }
                    : null;
            }
            iframeChain.push({
                parentSession: curSession,
                iframeBackendNodeId: be,
            });
            let left = 0;
            let top = 0;
            try {
                const { object } = await curSession.send("DOM.resolveNode", { backendNodeId: be });
                const objectId = object?.objectId;
                if (objectId) {
                    const { result } = await curSession.send("Runtime.callFunctionOn", {
                        objectId,
                        functionDeclaration: a11yScripts_generated_js_1.a11yScriptSources.getBoundingRectLite,
                        returnByValue: true,
                    });
                    left = Number(result?.value?.left ?? 0);
                    top = Number(result?.value?.top ?? 0);
                    await curSession
                        .send("Runtime.releaseObject", { objectId })
                        .catch(() => { });
                }
            }
            catch {
                //
            }
            curX = Math.max(0, curX - left);
            curY = Math.max(0, curY - top);
            curFrameId = matchedChild;
            curSession = page.getSessionForFrame(curFrameId);
        }
        catch {
            return null;
        }
    }
    return null;
}
