// @browserbasehq/stagehand/dist/cjs/lib/v3/understudy/a11y/snapshot/focusSelectors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IFRAME_STEP_RE = void 0;
exports.parseXPathToSteps = parseXPathToSteps;
exports.buildXPathFromSteps = buildXPathFromSteps;
exports.resolveFocusFrameAndTail = resolveFocusFrameAndTail;
exports.resolveCssFocusFrameAndTail = resolveCssFocusFrameAndTail;
exports.resolveObjectIdForXPath = resolveObjectIdForXPath;
exports.resolveObjectIdForCss = resolveObjectIdForCss;
exports.listChildrenOf = listChildrenOf;
const executionContextRegistry_js_1 = require("../../executionContextRegistry.js");
const locatorInvocation_js_1 = require("../../locatorInvocation.js");
const sdkErrors_js_1 = require("../../../types/public/sdkErrors.js");
const xpathUtils_js_1 = require("./xpathUtils.js");
/**
 * Parse a cross-frame XPath into discrete steps. Each step tracks whether it
 * represents a descendant hop (“//”) or a single-child hop (“/”).
 */
function parseXPathToSteps(path) {
    const s = path.trim();
    let i = 0;
    const steps = [];
    while (i < s.length) {
        let axis = "child";
        if (s.startsWith("//", i)) {
            axis = "desc";
            i += 2;
        }
        else if (s[i] === "/") {
            axis = "child";
            i += 1;
        }
        const start = i;
        while (i < s.length && s[i] !== "/")
            i++;
        const raw = s.slice(start, i).trim();
        if (!raw)
            continue;
        const name = raw.replace(/\[\d+\]\s*$/u, "").toLowerCase();
        steps.push({ axis, raw, name });
    }
    return steps;
}
/** Rebuild an XPath string from parsed steps. */
function buildXPathFromSteps(steps) {
    let out = "";
    for (const st of steps) {
        out += st.axis === "desc" ? "//" : "/";
        out += st.raw;
    }
    return out || "/";
}
exports.IFRAME_STEP_RE = /^i?frame(?:\[\d+])?$/i;
/**
 * Given a cross-frame XPath, walk iframe steps to resolve:
 * - the target frameId (last iframe hop)
 * - the tail XPath (within the target frame)
 * - the absolute XPath prefix up to the iframe element hosting that frame
 */
async function resolveFocusFrameAndTail(page, absoluteXPath, parentByFrame, rootId) {
    const steps = parseXPathToSteps(absoluteXPath);
    let ctxFrameId = rootId;
    let buf = [];
    let absPrefix = "";
    const flushIntoChild = async () => {
        if (!buf.length)
            return;
        const selectorForIframe = buildXPathFromSteps(buf);
        const parentSess = page.getSessionForFrame(ctxFrameId);
        const objectId = await resolveObjectIdForXPath(parentSess, selectorForIframe, ctxFrameId);
        if (!objectId)
            throw new sdkErrors_js_1.StagehandIframeError(selectorForIframe, "Failed to resolve iframe element by XPath");
        try {
            await parentSess.send("DOM.enable").catch(() => { });
            const desc = await parentSess.send("DOM.describeNode", { objectId });
            const iframeBackendNodeId = desc.node.backendNodeId;
            let childFrameId;
            for (const fid of listChildrenOf(parentByFrame, ctxFrameId)) {
                try {
                    const { backendNodeId } = await parentSess.send("DOM.getFrameOwner", { frameId: fid });
                    if (backendNodeId === iframeBackendNodeId) {
                        childFrameId = fid;
                        break;
                    }
                }
                catch {
                    continue;
                }
            }
            if (!childFrameId)
                throw new sdkErrors_js_1.StagehandIframeError(selectorForIframe, "Could not map iframe to child frameId");
            absPrefix = (0, xpathUtils_js_1.prefixXPath)(absPrefix || "/", selectorForIframe);
            ctxFrameId = childFrameId;
        }
        finally {
            await parentSess
                .send("Runtime.releaseObject", { objectId })
                .catch(() => { });
        }
        buf = [];
    };
    for (const st of steps) {
        buf.push(st);
        if (exports.IFRAME_STEP_RE.test(st.name)) {
            await flushIntoChild();
        }
    }
    const tailXPath = buildXPathFromSteps(buf);
    return { targetFrameId: ctxFrameId, tailXPath, absPrefix };
}
/** Resolve focus frame and tail CSS selector using '>>' to hop iframes. */
async function resolveCssFocusFrameAndTail(page, rawSelector, parentByFrame, rootId) {
    const parts = rawSelector
        .split(">>")
        .map((s) => s.trim())
        .filter(Boolean);
    let ctxFrameId = rootId;
    const absPrefix = "";
    for (let i = 0; i < Math.max(0, parts.length - 1); i++) {
        const parentSess = page.getSessionForFrame(ctxFrameId);
        const objectId = await resolveObjectIdForCss(parentSess, parts[i], ctxFrameId);
        if (!objectId)
            throw new sdkErrors_js_1.StagehandIframeError(parts[i], "Failed to resolve iframe via CSS hop");
        try {
            await parentSess.send("DOM.enable").catch(() => { });
            const desc = await parentSess.send("DOM.describeNode", { objectId });
            const iframeBackendNodeId = desc.node.backendNodeId;
            let childFrameId;
            for (const fid of listChildrenOf(parentByFrame, ctxFrameId)) {
                try {
                    const { backendNodeId } = await parentSess.send("DOM.getFrameOwner", { frameId: fid });
                    if (backendNodeId === iframeBackendNodeId) {
                        childFrameId = fid;
                        break;
                    }
                }
                catch {
                    continue;
                }
            }
            if (!childFrameId)
                throw new sdkErrors_js_1.StagehandIframeError(parts[i], "Could not map CSS iframe hop to child frameId");
            ctxFrameId = childFrameId;
        }
        finally {
            await parentSess
                .send("Runtime.releaseObject", { objectId })
                .catch(() => { });
        }
    }
    const tailSelector = parts[parts.length - 1] ?? "*";
    return { targetFrameId: ctxFrameId, tailSelector, absPrefix };
}
/** Resolve an XPath to a Runtime remoteObjectId in the given CDP session. */
async function resolveObjectIdForXPath(session, xpath, frameId) {
    let contextId;
    try {
        if (frameId) {
            contextId = await executionContextRegistry_js_1.executionContexts
                .waitForMainWorld(session, frameId, 800)
                .catch(() => executionContextRegistry_js_1.executionContexts.getMainWorld(session, frameId) ?? undefined);
        }
    }
    catch {
        contextId = undefined;
    }
    const expr = (0, locatorInvocation_js_1.buildLocatorInvocation)("resolveXPathMainWorld", [
        JSON.stringify(xpath),
        "0",
    ]);
    const { result, exceptionDetails } = await session.send("Runtime.evaluate", {
        expression: expr,
        returnByValue: false,
        contextId,
        awaitPromise: true,
    });
    if (exceptionDetails)
        return null;
    return result?.objectId ?? null;
}
/** Resolve a CSS selector (supports '>>' within the same frame only) to a Runtime objectId. */
async function resolveObjectIdForCss(session, selector, frameId) {
    let contextId;
    try {
        if (frameId) {
            contextId = await executionContextRegistry_js_1.executionContexts
                .waitForMainWorld(session, frameId, 800)
                .catch(() => executionContextRegistry_js_1.executionContexts.getMainWorld(session, frameId) ?? undefined);
        }
    }
    catch {
        contextId = undefined;
    }
    const primaryExpr = (0, locatorInvocation_js_1.buildLocatorInvocation)("resolveCssSelector", [
        JSON.stringify(selector),
        "0",
    ]);
    const fallbackExpr = (0, locatorInvocation_js_1.buildLocatorInvocation)("resolveCssSelectorPierce", [
        JSON.stringify(selector),
        "0",
    ]);
    const evaluate = async (expression) => {
        const { result, exceptionDetails } = await session.send("Runtime.evaluate", {
            expression,
            returnByValue: false,
            contextId,
            awaitPromise: true,
        });
        if (exceptionDetails)
            return null;
        return result?.objectId ?? null;
    };
    const primary = await evaluate(primaryExpr);
    if (primary)
        return primary;
    return evaluate(fallbackExpr);
}
function listChildrenOf(parentByFrame, parentId) {
    const out = [];
    for (const [fid, p] of parentByFrame.entries()) {
        if (p === parentId)
            out.push(fid);
    }
    return out;
}
