// @browserbasehq/stagehand/dist/esm/lib/v3/understudy/cdp.js
// lib/v3/understudy/cdp.ts
import WebSocket from "ws";
import { STAGEHAND_VERSION } from "../../version.js";
import { FlowLogger, } from "../flowlogger/FlowLogger.js";
import { CdpConnectionClosedError, PageNotFoundError, } from "../types/public/sdkErrors.js";
export class CdpConnection {
    ws;
    nextId = 1;
    inflight = new Map(); // Outstanding request records; `_sendViaSession()` inserts and `onMessage()` removes/resolves them.
    latestCdpCallEvent = new Map();
    eventHandlers = new Map();
    sessions = new Map();
    /** Maps sessionId -> targetId (1:1 mapping) */
    sessionToTarget = new Map();
    sessionDispatchWaiters = new Set();
    id = null; // root
    transportCloseHandlers = new Set();
    flowLoggerContext; // Instance-owned fallback flow context; V3 sets this once and later sends/callbacks re-enter it when ALS is absent.
    onTransportClosed(handler) {
        this.transportCloseHandlers.add(handler);
    }
    offTransportClosed(handler) {
        this.transportCloseHandlers.delete(handler);
    }
    emitTransportClosed(why) {
        for (const h of this.transportCloseHandlers) {
            try {
                h(why);
            }
            catch {
                //
            }
        }
    }
    constructor(ws) {
        this.ws = ws;
        this.ws.on("close", (code, reason) => {
            // Reason is a Buffer in ws; stringify defensively
            const why = `socket-close code=${code} reason=${String(reason || "")}`;
            this.rejectAllInflight(why);
            this.emitTransportClosed(why);
        });
        this.ws.on("error", (err) => {
            const why = `socket-error ${err?.message ?? String(err)}`;
            this.rejectAllInflight(why);
            this.emitTransportClosed(why);
        });
        this.ws.on("message", (data) => this.onMessage(data.toString()));
    }
    static async connect(wsUrl, options) {
        // Include User-Agent header for server-side observability and version tracking
        // Merge user-provided headers, letting them override defaults
        const headers = {
            "User-Agent": `Stagehand/${STAGEHAND_VERSION}`,
            ...options?.headers,
        };
        const ws = new WebSocket(wsUrl, { headers });
        await new Promise((resolve, reject) => {
            ws.once("open", () => resolve());
            ws.once("error", (e) => reject(e));
        });
        return new CdpConnection(ws);
    }
    async enableAutoAttach() {
        await this.send("Target.setAutoAttach", {
            autoAttach: true,
            flatten: true,
            waitForDebuggerOnStart: true,
        });
        await this.send("Target.setDiscoverTargets", { discover: true });
    }
    async send(method, params) {
        const id = this.nextId++;
        const payload = { id, method, params };
        const stack = new Error().stack?.split("\n").slice(1, 4).join("\n");
        const flowLoggerContext = FlowLogger.resolveContext(this.flowLoggerContext);
        const cdpCallEvent = flowLoggerContext
            ? FlowLogger.logCdpCallEvent(flowLoggerContext, {
                method,
                params,
                targetId: null,
            })
            : null;
        if (flowLoggerContext && cdpCallEvent) {
            this.latestCdpCallEvent.set(null, {
                flowLoggerContext,
                cdpCallEvent,
            });
        }
        const p = new Promise((resolve, reject) => {
            this.inflight.set(id, {
                resolve,
                reject,
                sessionId: null,
                method,
                params,
                stack,
                ts: Date.now(),
                flowLoggerContext,
                cdpCallEvent,
            });
        });
        // Prevent unhandledRejection if a session detaches before the caller awaits.
        void p.catch(() => { });
        this.ws.send(JSON.stringify(payload));
        return p;
    }
    on(event, handler) {
        const set = this.eventHandlers.get(event) ?? new Set();
        set.add(handler);
        this.eventHandlers.set(event, set);
    }
    off(event, handler) {
        const set = this.eventHandlers.get(event);
        if (set)
            set.delete(handler);
    }
    async close() {
        if (this.ws.readyState === WebSocket.CLOSED)
            return;
        await new Promise((resolve) => {
            this.ws.once("close", () => resolve());
            this.ws.close();
        });
    }
    rejectAllInflight(why) {
        for (const [id, entry] of this.inflight.entries()) {
            entry.reject(new CdpConnectionClosedError(why));
            this.inflight.delete(id);
        }
        this.latestCdpCallEvent.clear();
        for (const waiter of Array.from(this.sessionDispatchWaiters)) {
            waiter.reject(new CdpConnectionClosedError(why));
        }
    }
    getSession(sessionId) {
        return this.sessions.get(sessionId);
    }
    waitForSessionDispatch(sessionId, method, match) {
        return new Promise((resolve, reject) => {
            const waiter = {
                sessionId,
                method,
                match,
                resolve: () => {
                    this.sessionDispatchWaiters.delete(waiter);
                    resolve();
                },
                reject: (error) => {
                    this.sessionDispatchWaiters.delete(waiter);
                    reject(error);
                },
            };
            this.sessionDispatchWaiters.add(waiter);
        });
    }
    async attachToTarget(targetId) {
        const { sessionId } = (await this.send("Target.attachToTarget", { targetId, flatten: true }));
        let session = this.sessions.get(sessionId);
        if (!session) {
            session = new CdpSession(this, sessionId);
            this.sessions.set(sessionId, session);
        }
        this.sessionToTarget.set(sessionId, targetId);
        return session;
    }
    async getTargets() {
        const res = await this.send("Target.getTargets");
        return res.targetInfos;
    }
    onMessage(json) {
        const msg = JSON.parse(json);
        if ("id" in msg) {
            const rec = this.inflight.get(msg.id);
            if (!rec)
                return;
            this.inflight.delete(msg.id);
            if ("error" in msg && msg.error) {
                // Response/error events only make sense if the original send captured
                // both a flow context to re-enter and the emitted CdpCallEvent to hang
                // the terminal edge under.
                if (rec.flowLoggerContext && rec.cdpCallEvent) {
                    let targetId;
                    if (rec.sessionId) {
                        const mappedTargetId = this.sessionToTarget.get(rec.sessionId);
                        if (mappedTargetId) {
                            targetId = mappedTargetId;
                        }
                        else {
                            targetId = rec.sessionId;
                        }
                    }
                    else {
                        targetId = null;
                    }
                    FlowLogger.logCdpResponseEvent(rec.flowLoggerContext, rec.cdpCallEvent, {
                        method: rec.method,
                        error: `${msg.error.code} ${msg.error.message}`,
                        targetId,
                    });
                }
                rec.reject(new Error(`${msg.error.code} ${msg.error.message}`));
            }
            else {
                // Successful responses reuse the same cached call context so the
                // response lands under the exact CdpCallEvent emitted at send time.
                if (rec.flowLoggerContext && rec.cdpCallEvent) {
                    let targetId;
                    if (rec.sessionId) {
                        const mappedTargetId = this.sessionToTarget.get(rec.sessionId);
                        if (mappedTargetId) {
                            targetId = mappedTargetId;
                        }
                        else {
                            targetId = rec.sessionId;
                        }
                    }
                    else {
                        targetId = null;
                    }
                    FlowLogger.logCdpResponseEvent(rec.flowLoggerContext, rec.cdpCallEvent, {
                        method: rec.method,
                        result: msg.result,
                        targetId,
                    });
                }
                rec.resolve(msg.result);
            }
            return;
        }
        if ("method" in msg) {
            if (msg.method === "Target.attachedToTarget") {
                const p = msg
                    .params;
                if (!this.sessions.has(p.sessionId)) {
                    this.sessions.set(p.sessionId, new CdpSession(this, p.sessionId));
                }
                this.sessionToTarget.set(p.sessionId, p.targetInfo.targetId);
            }
            else if (msg.method === "Target.detachedFromTarget") {
                const p = msg
                    .params;
                for (const [id, entry] of this.inflight.entries()) {
                    if (entry.sessionId === p.sessionId) {
                        entry.reject(new PageNotFoundError(`target closed before CDP response (sessionId=${p.sessionId}, targetId=${p.targetId})`));
                        this.inflight.delete(id);
                    }
                }
                for (const waiter of Array.from(this.sessionDispatchWaiters)) {
                    if (waiter.sessionId === p.sessionId) {
                        waiter.reject(new PageNotFoundError(`target closed before CDP send (sessionId=${p.sessionId}, targetId=${p.targetId})`));
                    }
                }
                this.sessions.delete(p.sessionId);
                this.sessionToTarget.delete(p.sessionId);
                this.latestCdpCallEvent.delete(p.sessionId);
            }
            else if (msg.method === "Target.targetDestroyed") {
                const p = msg.params;
                // Remove any session mapping for this target
                for (const [sessionId, targetId] of this.sessionToTarget.entries()) {
                    if (targetId === p.targetId) {
                        this.sessionToTarget.delete(sessionId);
                        this.latestCdpCallEvent.delete(sessionId);
                        break;
                    }
                }
            }
            const { method, params, sessionId } = msg;
            const latestCdpCallEvent = this.latestCdpCallEvent.get(sessionId ?? null) ??
                (sessionId ? this.latestCdpCallEvent.get(null) : null);
            let targetId;
            if (sessionId) {
                const mappedTargetId = this.sessionToTarget.get(sessionId);
                if (mappedTargetId) {
                    targetId = mappedTargetId;
                }
                else {
                    targetId = sessionId;
                }
            }
            else {
                targetId = null;
            }
            // Unsolicited protocol messages are attached under the most recent call on
            // that session/root when one is known, so later callbacks still show up
            // in the same flow subtree.
            if (latestCdpCallEvent) {
                FlowLogger.logCdpMessageEvent(latestCdpCallEvent.flowLoggerContext, latestCdpCallEvent.cdpCallEvent, {
                    method,
                    params,
                    targetId,
                });
            }
            const dispatch = () => {
                if (sessionId) {
                    const session = this.sessions.get(sessionId);
                    session?.dispatch(method, params);
                    // Forward target lifecycle events to root listeners as well.
                    // Some browsers emit these via a parent session rather than the root
                    // connection; fan-out keeps target tracking consistent.
                    if (method.startsWith("Target.")) {
                        const handlers = this.eventHandlers.get(method);
                        if (handlers)
                            for (const h of handlers)
                                h(params);
                    }
                    return;
                }
                const handlers = this.eventHandlers.get(method);
                if (handlers)
                    for (const h of handlers)
                        h(params);
            };
            if (latestCdpCallEvent) {
                FlowLogger.withContext(latestCdpCallEvent.flowLoggerContext, dispatch);
            }
            else {
                dispatch();
            }
        }
    }
    _sendViaSession(sessionId, method, params) {
        const id = this.nextId++;
        const payload = { id, method, params, sessionId };
        const stack = new Error().stack?.split("\n").slice(1, 4).join("\n");
        const flowLoggerContext = FlowLogger.resolveContext(this.flowLoggerContext);
        let targetId;
        const mappedTargetId = this.sessionToTarget.get(sessionId);
        if (mappedTargetId) {
            targetId = mappedTargetId;
        }
        else {
            targetId = null;
        }
        const cdpCallEvent = flowLoggerContext
            ? FlowLogger.logCdpCallEvent(flowLoggerContext, {
                method,
                params,
                targetId,
            })
            : null;
        if (flowLoggerContext && cdpCallEvent) {
            this.latestCdpCallEvent.set(sessionId, {
                flowLoggerContext,
                cdpCallEvent,
            });
        }
        const p = new Promise((resolve, reject) => {
            this.inflight.set(id, {
                resolve,
                reject,
                sessionId,
                method,
                params,
                stack,
                ts: Date.now(),
                flowLoggerContext,
                cdpCallEvent,
            });
        });
        // Prevent unhandledRejection if a session detaches before the caller awaits.
        void p.catch(() => { });
        for (const waiter of Array.from(this.sessionDispatchWaiters)) {
            if (waiter.sessionId !== sessionId)
                continue;
            if (waiter.method !== method)
                continue;
            if (waiter.match && !waiter.match(params))
                continue;
            waiter.resolve();
            break;
        }
        this.ws.send(JSON.stringify(payload));
        return p;
    }
    _onSessionEvent(sessionId, event, handler) {
        const key = `${sessionId}:${event}`;
        const set = this.eventHandlers.get(key) ?? new Set();
        set.add(handler);
        this.eventHandlers.set(key, set);
    }
    _offSessionEvent(sessionId, event, handler) {
        const key = `${sessionId}:${event}`;
        const set = this.eventHandlers.get(key);
        if (set)
            set.delete(handler);
    }
    _dispatchToSession(sessionId, event, params) {
        const key = `${sessionId}:${event}`;
        const handlers = this.eventHandlers.get(key);
        if (handlers)
            for (const h of handlers)
                h(params);
    }
}
export class CdpSession {
    root;
    id;
    constructor(root, id) {
        this.root = root;
        this.id = id;
    }
    send(method, params) {
        return this.root._sendViaSession(this.id, method, params);
    }
    on(event, handler) {
        this.root._onSessionEvent(this.id, event, handler);
    }
    off(event, handler) {
        this.root._offSessionEvent(this.id, event, handler);
    }
    async close() {
        await this.root.send("Target.detachFromTarget", {
            sessionId: this.id,
        });
    }
    dispatch(event, params) {
        this.root._dispatchToSession(this.id, event, params);
    }
}
