// @browserbasehq/stagehand/dist/esm/lib/v3/understudy/consoleMessage.js
function formatRemoteObject(obj) {
    if (!obj)
        return "";
    if ("value" in obj) {
        const value = obj.value;
        if (value === undefined)
            return "";
        if (typeof value === "string")
            return value;
        try {
            return JSON.stringify(value);
        }
        catch {
            return String(value);
        }
    }
    if (obj.unserializableValue)
        return obj.unserializableValue;
    if (obj.description)
        return obj.description;
    return obj.type ?? "";
}
export class ConsoleMessage {
    event;
    pageRef;
    constructor(event, pageRef) {
        this.event = event;
        this.pageRef = pageRef;
    }
    type() {
        return this.event.type;
    }
    text() {
        const args = this.args();
        if (!args.length)
            return "";
        return args
            .map((arg) => formatRemoteObject(arg))
            .filter((chunk) => chunk.length > 0)
            .join(" ");
    }
    args() {
        return this.event.args ? [...this.event.args] : [];
    }
    location() {
        const frame = this.event.stackTrace?.callFrames?.[0];
        return {
            url: frame?.url,
            lineNumber: frame?.lineNumber,
            columnNumber: frame?.columnNumber,
        };
    }
    page() {
        return this.pageRef;
    }
    timestamp() {
        return this.event.timestamp;
    }
    raw() {
        return this.event;
    }
    toString() {
        return this.text();
    }
}
