// @browserbasehq/stagehand/dist/esm/lib/v3/understudy/frameRegistry.js
/** Minimal “shell” CDP frame used when we haven’t yet seen a real Frame from events. */
function shellFrame(id) {
    return {
        id,
        loaderId: "",
        url: "",
        domainAndRegistry: "",
        securityOrigin: "",
        mimeType: "text/html",
        secureContextType: "InsecureScheme",
        crossOriginIsolatedContextType: "NotIsolated",
        gatedAPIFeatures: [],
    };
}
export class FrameRegistry {
    /** Owner target id (top-level target); informational only */
    ownerTargetId;
    /** Current main/root frame id (changes on root swaps) */
    rootFrameId;
    /** frameId → FrameInfo */
    frames = new Map();
    /** sessionId → Set<frameId> (inverse map for diagnostics/fast membership checks) */
    framesBySession = new Map();
    constructor(ownerTargetId, mainFrameId) {
        this.ownerTargetId = ownerTargetId;
        this.rootFrameId = mainFrameId;
        this.ensureNode(mainFrameId);
    }
    // ---------------------- Mutators (called by Context/Page bridges) ----------------------
    /**
     * Record that a frame attached. If `parentId` is null and `frameId` differs from the current
     * root, this is a root swap and we rename the root id.
     *
     * IMPORTANT: The emitter's `sessionId` is the **owner** for the new/attached frame.
     */
    onFrameAttached(frameId, parentId, sessionId) {
        // Root swap (parentId === null for main frames).
        if (!parentId && frameId !== this.rootFrameId) {
            this.renameNodeId(this.rootFrameId, frameId);
            this.rootFrameId = frameId;
            // ownership moves to this session as well
            this.setOwnerSessionIdInternal(frameId, sessionId);
            return;
        }
        // Normal attach
        this.ensureNode(frameId);
        if (parentId)
            this.ensureNode(parentId);
        const info = this.frames.get(frameId);
        info.parentId = parentId ?? null;
        if (parentId) {
            this.frames.get(parentId).children.add(frameId);
        }
        // Ownership: the session that emitted frameAttached owns this frame.
        this.setOwnerSessionIdInternal(frameId, sessionId);
    }
    /**
     * Record a navigation with the full CDP `Frame`. Also updates ownership based on the emitting
     * session id. Handles root swap if the navigated frame is the new main (no parentId).
     */
    onFrameNavigated(frame, sessionId) {
        this.ensureNode(frame.id);
        const info = this.frames.get(frame.id);
        info.lastSeen = frame;
        // Ownership follows the session that reported the navigation
        this.setOwnerSessionIdInternal(frame.id, sessionId);
        // If this frame has no parent, it might be the (new) main/root
        if (!("parentId" in frame) || !frame.parentId) {
            if (frame.id !== this.rootFrameId) {
                // carry ordinal semantics by renaming the root id
                this.renameNodeId(this.rootFrameId, frame.id);
                this.rootFrameId = frame.id;
            }
        }
    }
    onNavigatedWithinDocument(frameId, url, sessionId) {
        this.ensureNode(frameId);
        const info = this.frames.get(frameId);
        const lastSeen = info.lastSeen ?? shellFrame(frameId);
        info.lastSeen = { ...lastSeen, url };
        this.setOwnerSessionIdInternal(frameId, sessionId);
    }
    /**
     * Record that a frame detached. If `reason !== "swap"`, remove the subtree from the graph,
     * and clean the inverse maps. For “swap” we keep the node to preserve continuity.
     */
    onFrameDetached(frameId, reason = "remove") {
        if (reason === "swap")
            return;
        // Collect subtree starting from frameId.
        const toRemove = [];
        const collect = (fid) => {
            toRemove.push(fid);
            const kids = this.frames.get(fid)?.children ?? new Set();
            for (const k of kids)
                collect(k);
        };
        collect(frameId);
        // Remove nodes, fix parents and inverse maps
        for (const fid of toRemove) {
            const info = this.frames.get(fid);
            if (!info)
                continue;
            // unlink from parent
            if (info.parentId) {
                const p = this.frames.get(info.parentId);
                p?.children.delete(fid);
            }
            // unlink inverse session map
            if (info.ownerSessionId) {
                const bag = this.framesBySession.get(info.ownerSessionId);
                bag?.delete(fid);
                if (bag && bag.size === 0)
                    this.framesBySession.delete(info.ownerSessionId);
            }
            this.frames.delete(fid);
        }
        // Guard root if we removed it; assign a placeholder root if needed
        if (!this.frames.has(this.rootFrameId)) {
            // Choose an arbitrary remaining node as root
            const iter = this.frames.keys().next();
            if (!iter.done)
                this.rootFrameId = iter.value;
        }
    }
    /**
     * An adopted OOPIF child session was created whose **main** frame id equals the parent iframe’s frameId.
     * We mark the entire child subtree as owned by `childSessionId`.
     * (Topology edges remain aligned by the parent session’s `frameAttached` events.)
     */
    adoptChildSession(childSessionId, childMainFrameId) {
        // The child session will emit its own navigations/attachments; as a seed,
        // mark the root frame as owned by the child session.
        this.setOwnerSessionIdInternal(childMainFrameId, childSessionId);
    }
    /**
     * Seed topology and ownership from an existing `Page.getFrameTree` snapshot, typically right after
     * a session is attached. This is a best-effort: we record frames and set the provided `sessionId`
     * as owner for the subtree **if** an owner isn't already set.
     */
    seedFromFrameTree(sessionId, frameTree) {
        const walk = (tree, parent) => {
            this.ensureNode(tree.frame.id);
            // topology
            this.frames.get(tree.frame.id).parentId = parent;
            if (parent)
                this.frames.get(parent).children.add(tree.frame.id);
            // last-seen frame
            this.frames.get(tree.frame.id).lastSeen = tree.frame;
            // ownership (only if unknown)
            if (!this.frames.get(tree.frame.id).ownerSessionId) {
                this.setOwnerSessionIdInternal(tree.frame.id, sessionId);
            }
            for (const c of tree.childFrames ?? [])
                walk(c, tree.frame.id);
        };
        walk(frameTree, null);
    }
    /**
     * Set the backendNodeId of the `<iframe>` element for a child frame **as seen from its parent**.
     * This is useful for building absolute XPath prefixes later (from the parent document).
     */
    setOwnerBackendNodeId(childFrameId, backendNodeId) {
        this.ensureNode(childFrameId);
        this.frames.get(childFrameId).ownerBackendNodeId = backendNodeId;
    }
    // ---------------------- Readers (consumed by Page/snapshot/locators) ----------------------
    mainFrameId() {
        return this.rootFrameId;
    }
    /**
     * Return the owner session id for this frame. If unknown, returns `undefined`.
     */
    getOwnerSessionId(frameId) {
        return this.frames.get(frameId)?.ownerSessionId;
    }
    /**
     * Return the owner backendNodeId (iframe element) if recorded.
     * This is in the **parent** document; pair it with `getParent`.
     */
    getOwnerBackendNodeId(frameId) {
        return this.frames.get(frameId)?.ownerBackendNodeId;
    }
    /**
     * Return the parent frame id, or null for root/unknown.
     */
    getParent(frameId) {
        return this.frames.get(frameId)?.parentId ?? null;
    }
    /**
     * List frame ids in root-first DFS order (same shape as CDP’s FrameTree traversal).
     */
    listAllFrames() {
        const out = [];
        const dfs = (fid) => {
            out.push(fid);
            const kids = this.frames.get(fid)?.children ?? new Set();
            for (const k of kids)
                dfs(k);
        };
        if (this.frames.has(this.rootFrameId))
            dfs(this.rootFrameId);
        return out;
    }
    /**
     * Serialize to `Protocol.Page.FrameTree` starting at the given root id (typically mainFrameId()).
     */
    asProtocolFrameTree(rootId) {
        const build = (fid) => {
            const info = this.frames.get(fid);
            const frame = info?.lastSeen ?? shellFrame(fid);
            const kids = info?.children ?? new Set();
            const childFrames = kids.size
                ? [...kids].map((k) => build(k))
                : undefined;
            return childFrames ? { frame, childFrames } : { frame };
        };
        return build(rootId);
    }
    /**
     * For diagnostics: return the current owner sessions for a frame id (0..n),
     * usually 0 or 1, but helpful to see potential inconsistencies during wiring.
     */
    sessionsForFrame(frameId) {
        const info = this.frames.get(frameId);
        return info?.ownerSessionId ? [info.ownerSessionId] : [];
    }
    /**
     * For diagnostics: return current frame set per session.
     */
    framesForSession(sessionId) {
        return [...(this.framesBySession.get(sessionId) ?? new Set())];
    }
    // ---------------------- Internal helpers ----------------------
    ensureNode(fid) {
        if (this.frames.has(fid))
            return;
        this.frames.set(fid, {
            parentId: null,
            children: new Set(),
            lastSeen: shellFrame(fid),
            ownerSessionId: undefined,
            ownerBackendNodeId: undefined,
        });
    }
    renameNodeId(oldId, newId) {
        if (oldId === newId)
            return;
        this.ensureNode(oldId);
        const info = this.frames.get(oldId);
        // Move info under new id
        this.frames.delete(oldId);
        this.frames.set(newId, { ...info });
        // Fix parent’s children set
        if (info.parentId) {
            const p = this.frames.get(info.parentId);
            if (p) {
                p.children.delete(oldId);
                p.children.add(newId);
            }
        }
        // Fix children’s parent pointers
        for (const c of info.children) {
            const ci = this.frames.get(c);
            if (ci)
                ci.parentId = newId;
        }
        // Fix inverse map (session -> frames)
        if (info.ownerSessionId) {
            const bag = this.framesBySession.get(info.ownerSessionId);
            if (bag) {
                bag.delete(oldId);
                bag.add(newId);
            }
        }
        // If root moved, keep the root id updated is handled by caller
    }
    setOwnerSessionIdInternal(frameId, sessionId) {
        this.ensureNode(frameId);
        const info = this.frames.get(frameId);
        // If the owner is unchanged, do nothing
        if (info.ownerSessionId === sessionId)
            return;
        // Remove from previous owner bag
        if (info.ownerSessionId) {
            const prev = this.framesBySession.get(info.ownerSessionId);
            prev?.delete(frameId);
            if (prev && prev.size === 0)
                this.framesBySession.delete(info.ownerSessionId);
        }
        // Set new owner and update bag
        info.ownerSessionId = sessionId;
        const bag = this.framesBySession.get(sessionId) ?? new Set();
        bag.add(frameId);
        this.framesBySession.set(sessionId, bag);
    }
}
